{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.725470Z",
     "start_time": "2025-02-26T02:56:03.135842Z"
    }
   },
   "source": [
    "from collections import Counter\n",
    "from collections import defaultdict\n",
    "import csv\n",
    "import json\n",
    "import pprint\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "from oaklib import get_adapter\n",
    "from oaklib.datamodels.vocabulary import IS_A\n",
    "\n",
    "sys.path.append('../..')  # Assuming notebooks/ is at the root level\n",
    "import core # appears as if there's a problem in PyCharm Jupyter, but it still works\n"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "at what point would i want to do an aggregation query instead of an $in query?",
   "id": "7a8ed2537401e30e"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.731805Z",
     "start_time": "2025-02-26T02:56:05.728984Z"
    }
   },
   "cell_type": "code",
   "source": [
    "NCBI_METADATA = \"ncbi_metadata\"\n",
    "GOLD_METADATA = \"gold_metadata_studies_with_samples\""
   ],
   "id": "749f5472f5f3ca3e",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.792334Z",
     "start_time": "2025-02-26T02:56:05.789751Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# For Ontology Access Kit (OAK)\n",
    "envo_adapter_string = \"sqlite:obo:envo\""
   ],
   "id": "4d26638d044585ea",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.837764Z",
     "start_time": "2025-02-26T02:56:05.834465Z"
    }
   },
   "cell_type": "code",
   "source": [
    "BIOME = \"ENVO:00000428\"\n",
    "ABP = \"ENVO:01000813\"\n",
    "ENV_MAT = \"ENVO:00010483\"\n",
    "\n",
    "ANIMAL_MANURE = \"ENVO:00003031\"\n",
    "BODILY_FLUID_MATERIAL = \"ENVO:02000019\"\n",
    "FECAL_MATERIAL =\"ENVO:00002003\"\n",
    "# FECES = \"UBERON:0001988\"\n",
    "LIQ_WATER = \"ENVO:00002006\"\n",
    "PETROLEUM = \"ENVO:00002984\"\n",
    "SEDIMENT = \"ENVO:00002007\"\n",
    "SLUDGE = \"ENVO:00002044\"\n",
    "SOIL = \"ENVO:00001998\""
   ],
   "id": "fd055c51ba4c1bf1",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.884864Z",
     "start_time": "2025-02-26T02:56:05.881601Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Configurable match length cutoff\n",
    "MATCH_LENGTH_CUTOFF = 3  # Exclude matches with length < 3 from needs_review_list"
   ],
   "id": "fecbf84a0ea9bc4",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.930986Z",
     "start_time": "2025-02-26T02:56:05.928398Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Define the target harmonized names\n",
    "target_harmonized_names = {\"env_broad_scale\", \"env_local_scale\", \"env_medium\"}"
   ],
   "id": "f970de1dff55b2c8",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:05.979573Z",
     "start_time": "2025-02-26T02:56:05.975170Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#inputs\n",
    "per_slot_value_curations_tsv = \"inputs/emp500_undefined_triad_slot_value_counts_curated.tsv\"\n",
    "macrae_crerar_biosample_clusters_with_coordinates_tsv = \"MacRae-Crerar/macrae_crerar_attributes_curated.tsv\"\n",
    "myrold_attributes_with_land_cover_tsv= \"myrold/myrold_attributes_with_land_cover.tsv\"\n"
   ],
   "id": "565af96d2d0af654",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:06.028894Z",
     "start_time": "2025-02-26T02:56:06.024176Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#outputs\n",
    "detailed_tsv = \"outputs/emp500_term_legitimacy.tsv\"\n",
    "emp500_mixs_style_triads_plus_gold_ids_tsv = \"outputs/emp500_matches_and_curations_plus_gold_ids.tsv\"\n",
    "missing_terms_tsv = \"outputs/emp500_annotations_for_review.tsv\"\n",
    "needs_review_tsv = \"outputs/emp500_env_context_vocabulary_for_review.tsv\"\n",
    "problematic_summary_tsv = \"outputs/emp500_undefined_triad_slot_value_counts.tsv\"\n",
    "\n",
    "summary_json_file = \"outputs/emp500_summary.json\""
   ],
   "id": "23caf7253005a6f8",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:06.129395Z",
     "start_time": "2025-02-26T02:56:06.075942Z"
    }
   },
   "cell_type": "code",
   "source": "envo_adapter = get_adapter(envo_adapter_string)",
   "id": "220b4a2ec76333d2",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:08.306298Z",
     "start_time": "2025-02-26T02:56:06.136709Z"
    }
   },
   "cell_type": "code",
   "source": [
    "envo_entities_list = list(envo_adapter.entities())\n",
    "\n",
    "# Create a dictionary mapping CURIEs to labels\n",
    "envo_curies_to_label = {\n",
    "    curie: envo_adapter.label(curie) for curie in envo_entities_list\n",
    "}\n"
   ],
   "id": "73a9b6aa223c0697",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:08.720165Z",
     "start_time": "2025-02-26T02:56:08.314568Z"
    }
   },
   "cell_type": "code",
   "source": [
    "abp_curies = list(envo_adapter.descendants(ABP, predicates=[IS_A]))\n",
    "biome_curies = list(envo_adapter.descendants(BIOME, predicates=[IS_A]))\n",
    "env_mat_curies = list(envo_adapter.descendants(ENV_MAT, predicates=[IS_A]))\n",
    "\n",
    "animal_manure_curies = list(envo_adapter.descendants(ANIMAL_MANURE, predicates=[IS_A]))\n",
    "bodily_fluid_material_curies = list(envo_adapter.descendants(BODILY_FLUID_MATERIAL, predicates=[IS_A]))\n",
    "fecal_material_curies = list(envo_adapter.descendants(FECAL_MATERIAL, predicates=[IS_A]))\n",
    "# feces_curies = list(envo_adapter.descendants(FECES, predicates=[IS_A]))\n",
    "liq_water_curies = list(envo_adapter.descendants(LIQ_WATER, predicates=[IS_A]))\n",
    "petroleum_curies = list(envo_adapter.descendants(PETROLEUM, predicates=[IS_A]))\n",
    "sediment_curies = list(envo_adapter.descendants(SEDIMENT, predicates=[IS_A]))\n",
    "sludge_curies = list(envo_adapter.descendants(SLUDGE, predicates=[IS_A]))\n",
    "soil_curies = list(envo_adapter.descendants(SOIL, predicates=[IS_A]))"
   ],
   "id": "8984357796b3d468",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:08.732188Z",
     "start_time": "2025-02-26T02:56:08.729405Z"
    }
   },
   "cell_type": "code",
   "source": "non_biome_non_env_mat_abp_curies = set(abp_curies) - set(biome_curies) - set(env_mat_curies)",
   "id": "5c78b76c2d4d63d5",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:08.912865Z",
     "start_time": "2025-02-26T02:56:08.780599Z"
    }
   },
   "cell_type": "code",
   "source": [
    "obsoletes_curies_labelled = list()\n",
    "\n",
    "obsoletes_curies_envo = set(envo_adapter.obsoletes())\n",
    "for curie in obsoletes_curies_envo:\n",
    "    temp_dict = dict()\n",
    "    temp_dict[\"curie\"] = curie\n",
    "    temp_dict[\"label\"] = envo_adapter.label(curie)\n",
    "    obsoletes_curies_labelled.append(temp_dict)"
   ],
   "id": "f25954adcd7ec69d",
   "outputs": [],
   "execution_count": 13
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:08.929053Z",
     "start_time": "2025-02-26T02:56:08.922127Z"
    }
   },
   "cell_type": "code",
   "source": "client = core.get_mongo_client()",
   "id": "9420b4fcbfa8574e",
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:08.975155Z",
     "start_time": "2025-02-26T02:56:08.972459Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Search for BioProjects\n",
    "bioprojects_cursor = core.fetch_mongodb_records_by_text(\n",
    "    client,\n",
    "    NCBI_METADATA,\n",
    "    \"bioprojects\",\n",
    "    \"emp500\"\n",
    ")\n"
   ],
   "id": "461e3e6c5f010745",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.024405Z",
     "start_time": "2025-02-26T02:56:09.020147Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# illustration or removing _id from the dictionaries if desired\n",
    "bioprojects_list = [\n",
    "    {key: value for key, value in document.items() if key!= '_id'}\n",
    "    for document in bioprojects_cursor\n",
    "]"
   ],
   "id": "5c4b0af79f7a6d64",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.079941Z",
     "start_time": "2025-02-26T02:56:09.065468Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(\"BioProjects:\")\n",
    "pprint.pprint(bioprojects_list)"
   ],
   "id": "5c0a5af3ebdef730",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BioProjects:\n",
      "[{'ProjectDescr': {'Description': 'The Earth Microbiome Project Multi-omics '\n",
      "                                  'component (EMP500) involves amplicon and '\n",
      "                                  'shotgun metagenomic sequencing and '\n",
      "                                  'metabolomic profiling of over five hundred '\n",
      "                                  'microbial communities from diverse '\n",
      "                                  'environments on our planet. We developed '\n",
      "                                  'new protocols for shotgun metagenomic '\n",
      "                                  'sequencing and assembly, with the goal of '\n",
      "                                  'applying this workflow to a range of '\n",
      "                                  'environmental samples, combined with '\n",
      "                                  'metabolomic profiling. 16S, 18S, and ITS '\n",
      "                                  'amplicon sequencing was done in addition. '\n",
      "                                  'We acquired a set of >500 fresh '\n",
      "                                  'environmental samples across a range of '\n",
      "                                  'habitats, with the help of the EMP network '\n",
      "                                  'of collaborators. A biobank of frozen '\n",
      "                                  'aliquots for all samples is maintained at '\n",
      "                                  'UCSD (soil samples stored at the Pacific '\n",
      "                                  'Northwest National Laboratory) for future '\n",
      "                                  'methods testing and analysis.',\n",
      "                   'Title': 'Earth Microbiome Project Multi-omics (EMP500)'},\n",
      "  'ProjectID': {'ArchiveID': {'accession': 'PRJEB42019',\n",
      "                              'archive': 'EBI',\n",
      "                              'id': '686344'},\n",
      "                'CenterID': {'center': 'EBI', 'id': '0', 'text': 'ERP125879'},\n",
      "                'LocalID': {'submission_id': 'qiita_sid_13114',\n",
      "                            'text': 'University of California San Diego '\n",
      "                                    'Microbiome Initiative'}},\n",
      "  'ProjectType': {'ProjectTypeSubmission': {'Method': {'method_type': 'eSequencing'},\n",
      "                                            'Objectives': {'Data': {'data_type': 'eOther',\n",
      "                                                                    'text': 'Unspecified '\n",
      "                                                                            'Objective'}},\n",
      "                                            'ProjectDataTypeSet': {'DataType': 'Other'},\n",
      "                                            'Target': {'capture': 'eWhole',\n",
      "                                                       'material': 'eGenome',\n",
      "                                                       'sample_scope': 'eMonoisolate'}}},\n",
      "  'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance'},\n",
      " {'ProjectDescr': {'Description': 'The Third Party Annotation (TPA)  assembly '\n",
      "                                  'was derived from the primary whole genome '\n",
      "                                  'shotgun (WGS) data set PRJEB42019, and was '\n",
      "                                  'assembled with metaspades v3.15.3. This '\n",
      "                                  'project includes samples from the following '\n",
      "                                  'biomes: '\n",
      "                                  'root:Environmental:Aquatic:Marine:Sediment.',\n",
      "                   'Keyword': 'TPA:assembly',\n",
      "                   'Publication': {'DbType': 'ePubmed',\n",
      "                                   'Reference': None,\n",
      "                                   'StructuredCitation': {'AuthorSet': {'Author': [{'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Alex '\n",
      "                                                                                                      'L',\n",
      "                                                                                             'Last': 'Mitchell'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Alexandre',\n",
      "                                                                                             'Last': 'Almeida'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Martin',\n",
      "                                                                                             'Last': 'Beracochea'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Miguel',\n",
      "                                                                                             'Last': 'Boland'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Josephine',\n",
      "                                                                                             'Last': 'Burgin'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Guy',\n",
      "                                                                                             'Last': 'Cochrane'}},\n",
      "                                                                                   {'Consortium': 'Common '\n",
      "                                                                                                  'Workflow '\n",
      "                                                                                                  'Language, '\n",
      "                                                                                                  'a '\n",
      "                                                                                                  'project '\n",
      "                                                                                                  'of '\n",
      "                                                                                                  'the '\n",
      "                                                                                                  'Software '\n",
      "                                                                                                  'Freedom '\n",
      "                                                                                                  'Conservancy, '\n",
      "                                                                                                  'Inc. '\n",
      "                                                                                                  '137 '\n",
      "                                                                                                  'Montague '\n",
      "                                                                                                  'Street, '\n",
      "                                                                                                  'Suite '\n",
      "                                                                                                  '380, '\n",
      "                                                                                                  'Brooklyn, '\n",
      "                                                                                                  'NY '\n",
      "                                                                                                  '11201-3548, '\n",
      "                                                                                                  'USA.',\n",
      "                                                                                    'Name': {'First': 'Michael '\n",
      "                                                                                                      'R',\n",
      "                                                                                             'Last': 'Crusoe'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Varsha',\n",
      "                                                                                             'Last': 'Kale'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Simon '\n",
      "                                                                                                      'C',\n",
      "                                                                                             'Last': 'Potter'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Lorna '\n",
      "                                                                                                      'J',\n",
      "                                                                                             'Last': 'Richardson'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Ekaterina',\n",
      "                                                                                             'Last': 'Sakharova'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Maxim',\n",
      "                                                                                             'Last': 'Scheremetjew'}},\n",
      "                                                                                   {'Consortium': 'Center '\n",
      "                                                                                                  'for '\n",
      "                                                                                                  'Algorithmic '\n",
      "                                                                                                  'Biotechnologies, '\n",
      "                                                                                                  'Saint '\n",
      "                                                                                                  'Petersburg '\n",
      "                                                                                                  'State '\n",
      "                                                                                                  'University, '\n",
      "                                                                                                  'Russia.',\n",
      "                                                                                    'Name': {'First': 'Anton',\n",
      "                                                                                             'Last': 'Korobeynikov'}},\n",
      "                                                                                   {'Consortium': 'Center '\n",
      "                                                                                                  'for '\n",
      "                                                                                                  'Algorithmic '\n",
      "                                                                                                  'Biotechnologies, '\n",
      "                                                                                                  'Saint '\n",
      "                                                                                                  'Petersburg '\n",
      "                                                                                                  'State '\n",
      "                                                                                                  'University, '\n",
      "                                                                                                  'Russia.',\n",
      "                                                                                    'Name': {'First': 'Alex',\n",
      "                                                                                             'Last': 'Shlemov'}},\n",
      "                                                                                   {'Consortium': 'Center '\n",
      "                                                                                                  'for '\n",
      "                                                                                                  'Algorithmic '\n",
      "                                                                                                  'Biotechnologies, '\n",
      "                                                                                                  'Saint '\n",
      "                                                                                                  'Petersburg '\n",
      "                                                                                                  'State '\n",
      "                                                                                                  'University, '\n",
      "                                                                                                  'Russia.',\n",
      "                                                                                    'Name': {'First': 'Olga',\n",
      "                                                                                             'Last': 'Kunyavskaya'}},\n",
      "                                                                                   {'Consortium': 'Center '\n",
      "                                                                                                  'for '\n",
      "                                                                                                  'Algorithmic '\n",
      "                                                                                                  'Biotechnologies, '\n",
      "                                                                                                  'Saint '\n",
      "                                                                                                  'Petersburg '\n",
      "                                                                                                  'State '\n",
      "                                                                                                  'University, '\n",
      "                                                                                                  'Russia.',\n",
      "                                                                                    'Name': {'First': 'Alla',\n",
      "                                                                                             'Last': 'Lapidus'}},\n",
      "                                                                                   {'Consortium': 'European '\n",
      "                                                                                                  'Molecular '\n",
      "                                                                                                  'Biology '\n",
      "                                                                                                  'Laboratory, '\n",
      "                                                                                                  'European '\n",
      "                                                                                                  'Bioinformatics '\n",
      "                                                                                                  'Institute '\n",
      "                                                                                                  '(EMBL-EBI), '\n",
      "                                                                                                  'Wellcome '\n",
      "                                                                                                  'Genome '\n",
      "                                                                                                  'Campus, '\n",
      "                                                                                                  'Hinxton, '\n",
      "                                                                                                  'Cambridge '\n",
      "                                                                                                  'CB10 '\n",
      "                                                                                                  '1SD, '\n",
      "                                                                                                  'UK.',\n",
      "                                                                                    'Name': {'First': 'Robert '\n",
      "                                                                                                      'D',\n",
      "                                                                                             'Last': 'Finn'}}]},\n",
      "                                                          'Journal': {'Issue': 'D1',\n",
      "                                                                      'JournalTitle': 'Nucleic '\n",
      "                                                                                      'acids '\n",
      "                                                                                      'research',\n",
      "                                                                      'PagesFrom': 'D570',\n",
      "                                                                      'PagesTo': 'D578',\n",
      "                                                                      'Volume': '48',\n",
      "                                                                      'Year': '2020'},\n",
      "                                                          'Title': 'MGnify: '\n",
      "                                                                   'the '\n",
      "                                                                   'microbiome '\n",
      "                                                                   'analysis '\n",
      "                                                                   'resource '\n",
      "                                                                   'in 2020.'},\n",
      "                                   'date': '2020-01-08T00:00:00Z',\n",
      "                                   'id': '31696235',\n",
      "                                   'status': 'ePublished'},\n",
      "                   'Title': 'EMG produced TPA metagenomics assembly of '\n",
      "                            'PRJEB42019 data set (Earth Microbiome Project '\n",
      "                            'Multi-omics (EMP500)).'},\n",
      "  'ProjectID': {'ArchiveID': {'accession': 'PRJEB76305',\n",
      "                              'archive': 'EBI',\n",
      "                              'id': '1121787'},\n",
      "                'CenterID': {'center': 'EBI', 'id': '0', 'text': 'ERP160863'},\n",
      "                'LocalID': {'submission_id': 'PRJEB42019_20240604174618',\n",
      "                            'text': 'EMG'}},\n",
      "  'ProjectType': {'ProjectTypeSubmission': {'Method': {'method_type': 'eSequencing'},\n",
      "                                            'Objectives': {'Data': {'data_type': 'eOther',\n",
      "                                                                    'text': 'Unspecified '\n",
      "                                                                            'Objective'}},\n",
      "                                            'ProjectDataTypeSet': {'DataType': 'Other'},\n",
      "                                            'Target': {'capture': 'eWhole',\n",
      "                                                       'material': 'eGenome',\n",
      "                                                       'sample_scope': 'eMonoisolate'}}},\n",
      "  'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance'},\n",
      " {'ProjectDescr': {'Description': 'Materials and Methods</p><p>Sediment '\n",
      "                                  'Collection</p><p>Samples were collected in '\n",
      "                                  'two-day windows in January 2016 and '\n",
      "                                  'February 2018 (see Figure 2.1 for locations '\n",
      "                                  'of collection). Each sediment sample was '\n",
      "                                  'taken from at least three different '\n",
      "                                  'locations within a small area in the creek '\n",
      "                                  'or river, then homogenized in sterile 50mL '\n",
      "                                  'tubes for subsequent DNA extraction and '\n",
      "                                  'quantification of metals. Biofilm samples '\n",
      "                                  'were collected when observed in the creek '\n",
      "                                  'or river bed. Sediment and biofilm samples '\n",
      "                                  'were stored on dry ice until arriving at '\n",
      "                                  'the lab, where they were kept at -80C until '\n",
      "                                  'extraction. Temperature, salinity, pH, and '\n",
      "                                  'redox potential were measured with a YSI '\n",
      "                                  'Pro Plus multiparameter meter at the time '\n",
      "                                  'of sampling. Samples collected in 2016 were '\n",
      "                                  'also sent to the Earth Microbiome Project '\n",
      "                                  '500 (EMP500) for 16S, 18S, and ITS amplicon '\n",
      "                                  'sequencing, shotgun metagenomic library '\n",
      "                                  'sequencing, and non-targeted LC-MS/MS '\n",
      "                                  'metabolomic analysis (data not included '\n",
      "                                  'herein).</p><p>DNA Extraction</p><p>Genomic '\n",
      "                                  'DNA was extracted from re-homogenized '\n",
      "                                  'samples using the PowerSoil DNA Isolation '\n",
      "                                  'Kit (MoBio Laboratories, Inc., Carlsbad, '\n",
      "                                  \"CA, USA) following manufacturer's protocol, \"\n",
      "                                  'with the following changes: solutions C1 '\n",
      "                                  'and C6 were heated to 65C. Additionally, we '\n",
      "                                  'allowed solution C6 to sit on the filter '\n",
      "                                  'membrane for at least one minute before '\n",
      "                                  'centrifugation and repeated the C6 '\n",
      "                                  'step.</p><p>16S rRNA Amplicon Diversity '\n",
      "                                  'Sequencing</p><p>Samples were amplified for '\n",
      "                                  'bacterial 16S DNA amplicon sequencing at '\n",
      "                                  'RTLGenomics (Lubbock, TX). In the first '\n",
      "                                  'round of amplification for all samples '\n",
      "                                  '(2016 &amp; 2018), the V1-V3 of the 16S '\n",
      "                                  'gene was targeted for sequencing using a '\n",
      "                                  'combination of the 28F and Illumina i5 '\n",
      "                                  'sequencing primers and the Illumina i7 '\n",
      "                                  'sequencing primer with the 519R primer. '\n",
      "                                  'Amplifications were performed in 25 ul '\n",
      "                                  'reactions with Qiagen HotStarTaq master mix '\n",
      "                                  '(Qiagen Inc, Valencia, CA, USA), 1ul of '\n",
      "                                  'each 5uM primer, and 1ul of template. '\n",
      "                                  'Reactions were performed on ABI Veriti '\n",
      "                                  'thermocyclers (Applied Biosytems, Carlsbad, '\n",
      "                                  'CA, USA) under the following thermal '\n",
      "                                  'profile: 95C for 5 min, then 25 cycles of '\n",
      "                                  '94C for 30 sec, 54C for 40 sec, 72C for 1 '\n",
      "                                  'min, followed by one cycle of 72C for 10 '\n",
      "                                  'min and 4C hold. Amplicons from the first '\n",
      "                                  'stage of PCR reactions were added to a '\n",
      "                                  'second amplification step based on '\n",
      "                                  'qualitatively determine concentrations '\n",
      "                                  'using ethidium bromide stained gels. '\n",
      "                                  'Primers for the second PCR were designed '\n",
      "                                  'based on the Illumina Nextera PCR primers. '\n",
      "                                  'The second stage amplification utilized the '\n",
      "                                  'same cycling parameters as the first round, '\n",
      "                                  'except it only ran for 10 cycles. Products '\n",
      "                                  'were then pooled in equimolar '\n",
      "                                  'concentrations and each pool was size '\n",
      "                                  'selected in two rounds using SPRIselect '\n",
      "                                  '(BeckmanCoulter, Indianapolis, IN, USA) in '\n",
      "                                  'a 0.7 ratio for both rounds. Size selected '\n",
      "                                  'pools were then quantified using the Quibit '\n",
      "                                  '2.0 fluorometer (Life Technologies) and '\n",
      "                                  'loaded on an Illumina MiSeq (Illumina, Inc. '\n",
      "                                  'San Diego, CA, USA) 2x300 flow cell at 10pM '\n",
      "                                  'and sequenced at RTLGenomics.',\n",
      "                   'Relevance': {'Environmental': 'yes'},\n",
      "                   'Title': 'EXAMINING THE IMPACT OF ACID ROCK DRAINAGE ON THE '\n",
      "                            'SEDIMENT-BORNE MICROBIOMES IN THE ANIMAS RIVER '\n",
      "                            'WATERSHED'},\n",
      "  'ProjectID': {'ArchiveID': {'accession': 'PRJNA1170297',\n",
      "                              'archive': 'NCBI',\n",
      "                              'id': '1170297'},\n",
      "                'LocalID': ['bp0', 'bp0']},\n",
      "  'ProjectType': {'ProjectTypeSubmission': {'IntendedDataTypeSet': {'DataType': 'raw '\n",
      "                                                                                'sequence '\n",
      "                                                                                'reads'},\n",
      "                                            'Method': {'method_type': 'eSequencing'},\n",
      "                                            'Objectives': {'Data': {'data_type': 'eRawSequenceReads'}},\n",
      "                                            'ProjectDataTypeSet': {'DataType': 'raw '\n",
      "                                                                               'sequence '\n",
      "                                                                               'reads'},\n",
      "                                            'Target': {'capture': 'eWhole',\n",
      "                                                       'material': 'eGenome',\n",
      "                                                       'sample_scope': 'eMultispecies'}}},\n",
      "  'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance'},\n",
      " {'ProjectDescr': {'Description': 'Materials and Methods</p><p>Sediment '\n",
      "                                  'Collection</p><p>Samples were collected in '\n",
      "                                  'two-day windows in January 2016 and '\n",
      "                                  'February 2018 (see Figure 2.1 for locations '\n",
      "                                  'of collection). Each sediment sample was '\n",
      "                                  'taken from at least three different '\n",
      "                                  'locations within a small area in the creek '\n",
      "                                  'or river, then homogenized in sterile 50mL '\n",
      "                                  'tubes for subsequent DNA extraction and '\n",
      "                                  'quantification of metals. Biofilm samples '\n",
      "                                  'were collected when observed in the creek '\n",
      "                                  'or river bed. Sediment and biofilm samples '\n",
      "                                  'were stored on dry ice until arriving at '\n",
      "                                  'the lab, where they were kept at -80C until '\n",
      "                                  'extraction. Temperature, salinity, pH, and '\n",
      "                                  'redox potential were measured with a YSI '\n",
      "                                  'Pro Plus multiparameter meter at the time '\n",
      "                                  'of sampling. Samples collected in 2016 were '\n",
      "                                  'also sent to the Earth Microbiome Project '\n",
      "                                  '500 (EMP500) for 16S, 18S, and ITS amplicon '\n",
      "                                  'sequencing, shotgun metagenomic library '\n",
      "                                  'sequencing, and non-targeted LC-MS/MS '\n",
      "                                  'metabolomic analysis (data not included '\n",
      "                                  'herein).</p><p>DNA Extraction</p><p>Genomic '\n",
      "                                  'DNA was extracted from re-homogenized '\n",
      "                                  'samples using the PowerSoil DNA Isolation '\n",
      "                                  'Kit (MoBio Laboratories, Inc., Carlsbad, '\n",
      "                                  \"CA, USA) following manufacturer's protocol, \"\n",
      "                                  'with the following changes: solutions C1 '\n",
      "                                  'and C6 were heated to 65C. Additionally, we '\n",
      "                                  'allowed solution C6 to sit on the filter '\n",
      "                                  'membrane for at least one minute before '\n",
      "                                  'centrifugation and repeated the C6 '\n",
      "                                  'step.</p><p>16S rRNA Amplicon Diversity '\n",
      "                                  'Sequencing</p><p>Samples were amplified for '\n",
      "                                  'bacterial 16S DNA amplicon sequencing at '\n",
      "                                  'RTLGenomics (Lubbock, TX). In the first '\n",
      "                                  'round of amplification for all samples '\n",
      "                                  '(2016 &amp; 2018), the V1-V3 of the 16S '\n",
      "                                  'gene was targeted for sequencing using a '\n",
      "                                  'combination of the 28F and Illumina i5 '\n",
      "                                  'sequencing primers and the Illumina i7 '\n",
      "                                  'sequencing primer with the 519R primer. '\n",
      "                                  'Amplifications were performed in 25 ul '\n",
      "                                  'reactions with Qiagen HotStarTaq master mix '\n",
      "                                  '(Qiagen Inc, Valencia, CA, USA), 1ul of '\n",
      "                                  'each 5uM primer, and 1ul of template. '\n",
      "                                  'Reactions were performed on ABI Veriti '\n",
      "                                  'thermocyclers (Applied Biosytems, Carlsbad, '\n",
      "                                  'CA, USA) under the following thermal '\n",
      "                                  'profile: 95C for 5 min, then 25 cycles of '\n",
      "                                  '94C for 30 sec, 54C for 40 sec, 72C for 1 '\n",
      "                                  'min, followed by one cycle of 72C for 10 '\n",
      "                                  'min and 4C hold. Amplicons from the first '\n",
      "                                  'stage of PCR reactions were added to a '\n",
      "                                  'second amplification step based on '\n",
      "                                  'qualitatively determine concentrations '\n",
      "                                  'using ethidium bromide stained gels. '\n",
      "                                  'Primers for the second PCR were designed '\n",
      "                                  'based on the Illumina Nextera PCR primers. '\n",
      "                                  'The second stage amplification utilized the '\n",
      "                                  'same cycling parameters as the first round, '\n",
      "                                  'except it only ran for 10 cycles. Products '\n",
      "                                  'were then pooled in equimolar '\n",
      "                                  'concentrations and each pool was size '\n",
      "                                  'selected in two rounds using SPRIselect '\n",
      "                                  '(BeckmanCoulter, Indianapolis, IN, USA) in '\n",
      "                                  'a 0.7 ratio for both rounds. Size selected '\n",
      "                                  'pools were then quantified using the Quibit '\n",
      "                                  '2.0 fluorometer (Life Technologies) and '\n",
      "                                  'loaded on an Illumina MiSeq (Illumina, Inc. '\n",
      "                                  'San Diego, CA, USA) 2x300 flow cell at 10pM '\n",
      "                                  'and sequenced at RTLGenomics.',\n",
      "                   'Name': 'Acid Mine Sediment samples',\n",
      "                   'Relevance': {'Environmental': 'yes'},\n",
      "                   'Title': 'EXAMINING THE IMPACT OF ACID ROCK DRAINAGE ON THE '\n",
      "                            'SEDIMENT-BORNE MICROBIOMES IN THE ANIMAS RIVER '\n",
      "                            'WATERSHED'},\n",
      "  'ProjectID': {'ArchiveID': {'accession': 'PRJNA1167286',\n",
      "                              'archive': 'NCBI',\n",
      "                              'id': '1167286'},\n",
      "                'LocalID': ['bp0', 'bp0']},\n",
      "  'ProjectType': {'ProjectTypeSubmission': {'IntendedDataTypeSet': {'DataType': 'raw '\n",
      "                                                                                'sequence '\n",
      "                                                                                'reads'},\n",
      "                                            'Method': {'method_type': 'eSequencing'},\n",
      "                                            'Objectives': {'Data': {'data_type': 'eRawSequenceReads'}},\n",
      "                                            'ProjectDataTypeSet': {'DataType': 'raw '\n",
      "                                                                               'sequence '\n",
      "                                                                               'reads'},\n",
      "                                            'Target': {'Organism': {'OrganismName': 'sediment '\n",
      "                                                                                    'metagenome',\n",
      "                                                                    'Supergroup': 'eOther',\n",
      "                                                                    'species': '749907',\n",
      "                                                                    'taxID': '749907'},\n",
      "                                                       'capture': 'eWhole',\n",
      "                                                       'material': 'eGenome',\n",
      "                                                       'sample_scope': 'eEnvironment'}}},\n",
      "  'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance'}]\n"
     ]
    }
   ],
   "execution_count": 17
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.100511Z",
     "start_time": "2025-02-26T02:56:09.096579Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Search for GOLD studies\n",
    "gold_studies_cursor = core.fetch_mongodb_records_by_text(\n",
    "    client,\n",
    "    GOLD_METADATA,\n",
    "    \"studies\",\n",
    "    \"emp500\"\n",
    ")\n",
    "\n",
    "gold_studies_list = list(gold_studies_cursor)\n"
   ],
   "id": "654d8138e507d7b0",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.154026Z",
     "start_time": "2025-02-26T02:56:09.142760Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(\"GOLD studies:\")\n",
    "pprint.pprint(gold_studies_list)"
   ],
   "id": "af10774d5941ea72",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GOLD studies:\n",
      "[{'_id': ObjectId('67ae0a020614a9099996848c'),\n",
      "  'addDate': '2021-07-24',\n",
      "  'biosamples': ['Gb0239921',\n",
      "                 'Gb0239928',\n",
      "                 'Gb0239989',\n",
      "                 'Gb0240012',\n",
      "                 'Gb0240069',\n",
      "                 'Gb0240099',\n",
      "                 'Gb0240127',\n",
      "                 'Gb0239958',\n",
      "                 'Gb0239985',\n",
      "                 'Gb0239997',\n",
      "                 'Gb0240033',\n",
      "                 'Gb0240039',\n",
      "                 'Gb0240051',\n",
      "                 'Gb0240056',\n",
      "                 'Gb0240068',\n",
      "                 'Gb0240083',\n",
      "                 'Gb0240106',\n",
      "                 'Gb0239927',\n",
      "                 'Gb0297473',\n",
      "                 'Gb0297480',\n",
      "                 'Gb0297366',\n",
      "                 'Gb0297316',\n",
      "                 'Gb0297394',\n",
      "                 'Gb0297436',\n",
      "                 'Gb0297349',\n",
      "                 'Gb0297333',\n",
      "                 'Gb0297352',\n",
      "                 'Gb0297353',\n",
      "                 'Gb0297430',\n",
      "                 'Gb0297433',\n",
      "                 'Gb0297326',\n",
      "                 'Gb0297339',\n",
      "                 'Gb0297346',\n",
      "                 'Gb0297360',\n",
      "                 'Gb0297354',\n",
      "                 'Gb0297444',\n",
      "                 'Gb0297470',\n",
      "                 'Gb0297391',\n",
      "                 'Gb0297403',\n",
      "                 'Gb0297408',\n",
      "                 'Gb0297409',\n",
      "                 'Gb0297476',\n",
      "                 'Gb0297481',\n",
      "                 'Gb0315915',\n",
      "                 'Gb0315920',\n",
      "                 'Gb0315965',\n",
      "                 'Gb0315968',\n",
      "                 'Gb0315975',\n",
      "                 'Gb0315649',\n",
      "                 'Gb0315652',\n",
      "                 'Gb0315893',\n",
      "                 'Gb0315895',\n",
      "                 'Gb0315762',\n",
      "                 'Gb0315764',\n",
      "                 'Gb0315800',\n",
      "                 'Gb0315983',\n",
      "                 'Gb0315943',\n",
      "                 'Gb0315944',\n",
      "                 'Gb0315945',\n",
      "                 'Gb0315950',\n",
      "                 'Gb0315476',\n",
      "                 'Gb0315613',\n",
      "                 'Gb0315644',\n",
      "                 'Gb0315776',\n",
      "                 'Gb0315672',\n",
      "                 'Gb0315787',\n",
      "                 'Gb0315795',\n",
      "                 'Gb0315447',\n",
      "                 'Gb0315852',\n",
      "                 'Gb0315871',\n",
      "                 'Gb0315903',\n",
      "                 'Gb0315914',\n",
      "                 'Gb0315806',\n",
      "                 'Gb0315956',\n",
      "                 'Gb0315805',\n",
      "                 'Gb0315807',\n",
      "                 'Gb0315887',\n",
      "                 'Gb0315891',\n",
      "                 'Gb0315999',\n",
      "                 'Gb0316011',\n",
      "                 'Gb0316015',\n",
      "                 'Gb0316016',\n",
      "                 'Gb0315985',\n",
      "                 'Gb0316029',\n",
      "                 'Gb0316018',\n",
      "                 'Gb0316039',\n",
      "                 'Gb0315746',\n",
      "                 'Gb0315693',\n",
      "                 'Gb0315869',\n",
      "                 'Gb0315422',\n",
      "                 'Gb0315959',\n",
      "                 'Gb0315889',\n",
      "                 'Gb0315522',\n",
      "                 'Gb0315523',\n",
      "                 'Gb0315690',\n",
      "                 'Gb0315692',\n",
      "                 'Gb0315494',\n",
      "                 'Gb0315767',\n",
      "                 'Gb0315484',\n",
      "                 'Gb0315624',\n",
      "                 'Gb0315659',\n",
      "                 'Gb0315554',\n",
      "                 'Gb0315664',\n",
      "                 'Gb0315460',\n",
      "                 'Gb0315988',\n",
      "                 'Gb0315998',\n",
      "                 'Gb0316002',\n",
      "                 'Gb0316004',\n",
      "                 'Gb0316006',\n",
      "                 'Gb0316022',\n",
      "                 'Gb0316023',\n",
      "                 'Gb0315610',\n",
      "                 'Gb0315614',\n",
      "                 'Gb0315678',\n",
      "                 'Gb0315573',\n",
      "                 'Gb0315432',\n",
      "                 'Gb0315710',\n",
      "                 'Gb0315581',\n",
      "                 'Gb0240129',\n",
      "                 'Gb0239924',\n",
      "                 'Gb0239963',\n",
      "                 'Gb0239982',\n",
      "                 'Gb0239993',\n",
      "                 'Gb0239995',\n",
      "                 'Gb0240006',\n",
      "                 'Gb0240020',\n",
      "                 'Gb0240066',\n",
      "                 'Gb0240116',\n",
      "                 'Gb0240121',\n",
      "                 'Gb0239933',\n",
      "                 'Gb0239935',\n",
      "                 'Gb0239959',\n",
      "                 'Gb0239983',\n",
      "                 'Gb0239988',\n",
      "                 'Gb0239999',\n",
      "                 'Gb0240000',\n",
      "                 'Gb0240003',\n",
      "                 'Gb0240018',\n",
      "                 'Gb0240019',\n",
      "                 'Gb0240042',\n",
      "                 'Gb0240055',\n",
      "                 'Gb0240072',\n",
      "                 'Gb0240080',\n",
      "                 'Gb0240082',\n",
      "                 'Gb0240085',\n",
      "                 'Gb0240091',\n",
      "                 'Gb0240094',\n",
      "                 'Gb0240108',\n",
      "                 'Gb0240109',\n",
      "                 'Gb0240117',\n",
      "                 'Gb0297482',\n",
      "                 'Gb0297455',\n",
      "                 'Gb0297463',\n",
      "                 'Gb0297344',\n",
      "                 'Gb0297372',\n",
      "                 'Gb0297317',\n",
      "                 'Gb0297447',\n",
      "                 'Gb0297388',\n",
      "                 'Gb0297397',\n",
      "                 'Gb0297416',\n",
      "                 'Gb0297426',\n",
      "                 'Gb0297446',\n",
      "                 'Gb0297379',\n",
      "                 'Gb0297358',\n",
      "                 'Gb0297362',\n",
      "                 'Gb0297437',\n",
      "                 'Gb0297386',\n",
      "                 'Gb0297466',\n",
      "                 'Gb0297442',\n",
      "                 'Gb0297464',\n",
      "                 'Gb0297469',\n",
      "                 'Gb0297478',\n",
      "                 'Gb0297488',\n",
      "                 'Gb0315922',\n",
      "                 'Gb0315728',\n",
      "                 'Gb0315729',\n",
      "                 'Gb0315732',\n",
      "                 'Gb0315737',\n",
      "                 'Gb0315972',\n",
      "                 'Gb0315976',\n",
      "                 'Gb0315759',\n",
      "                 'Gb0315938',\n",
      "                 'Gb0315946',\n",
      "                 'Gb0315466',\n",
      "                 'Gb0315425',\n",
      "                 'Gb0315594',\n",
      "                 'Gb0315601',\n",
      "                 'Gb0315590',\n",
      "                 'Gb0315619',\n",
      "                 'Gb0315835',\n",
      "                 'Gb0315842',\n",
      "                 'Gb0315435',\n",
      "                 'Gb0315863',\n",
      "                 'Gb0315796',\n",
      "                 'Gb0315689',\n",
      "                 'Gb0315714',\n",
      "                 'Gb0315770',\n",
      "                 'Gb0315868',\n",
      "                 'Gb0315881',\n",
      "                 'Gb0315905',\n",
      "                 'Gb0315809',\n",
      "                 'Gb0315836',\n",
      "                 'Gb0315960',\n",
      "                 'Gb0315804',\n",
      "                 'Gb0315885',\n",
      "                 'Gb0316000',\n",
      "                 'Gb0315810',\n",
      "                 'Gb0316027',\n",
      "                 'Gb0315826',\n",
      "                 'Gb0316037',\n",
      "                 'Gb0316043',\n",
      "                 'Gb0315841',\n",
      "                 'Gb0315454',\n",
      "                 'Gb0315751',\n",
      "                 'Gb0315456',\n",
      "                 'Gb0315721',\n",
      "                 'Gb0315722',\n",
      "                 'Gb0315511',\n",
      "                 'Gb0315777',\n",
      "                 'Gb0315419',\n",
      "                 'Gb0315896',\n",
      "                 'Gb0315436',\n",
      "                 'Gb0315783',\n",
      "                 'Gb0315900',\n",
      "                 'Gb0315530',\n",
      "                 'Gb0315481',\n",
      "                 'Gb0315772',\n",
      "                 'Gb0315634',\n",
      "                 'Gb0315551',\n",
      "                 'Gb0315552',\n",
      "                 'Gb0315553',\n",
      "                 'Gb0316001',\n",
      "                 'Gb0316003',\n",
      "                 'Gb0316028',\n",
      "                 'Gb0315597',\n",
      "                 'Gb0315617',\n",
      "                 'Gb0315638',\n",
      "                 'Gb0315639',\n",
      "                 'Gb0315542',\n",
      "                 'Gb0315574',\n",
      "                 'Gb0315576',\n",
      "                 'Gb0315708',\n",
      "                 'Gb0315616',\n",
      "                 'Gb0315707',\n",
      "                 'Gb0239918',\n",
      "                 'Gb0239922',\n",
      "                 'Gb0239925',\n",
      "                 'Gb0239940',\n",
      "                 'Gb0239981',\n",
      "                 'Gb0240002',\n",
      "                 'Gb0240004',\n",
      "                 'Gb0240070',\n",
      "                 'Gb0240093',\n",
      "                 'Gb0240113',\n",
      "                 'Gb0239938',\n",
      "                 'Gb0239942',\n",
      "                 'Gb0239955',\n",
      "                 'Gb0239967',\n",
      "                 'Gb0239970',\n",
      "                 'Gb0239972',\n",
      "                 'Gb0239975',\n",
      "                 'Gb0239980',\n",
      "                 'Gb0239996',\n",
      "                 'Gb0240009',\n",
      "                 'Gb0240034',\n",
      "                 'Gb0240043',\n",
      "                 'Gb0240048',\n",
      "                 'Gb0240063',\n",
      "                 'Gb0240084',\n",
      "                 'Gb0240088',\n",
      "                 'Gb0240102',\n",
      "                 'Gb0240112',\n",
      "                 'Gb0240114',\n",
      "                 'Gb0240124',\n",
      "                 'Gb0240125',\n",
      "                 'Gb0297479',\n",
      "                 'Gb0297401',\n",
      "                 'Gb0297406',\n",
      "                 'Gb0297467',\n",
      "                 'Gb0297474',\n",
      "                 'Gb0297396',\n",
      "                 'Gb0297377',\n",
      "                 'Gb0297398',\n",
      "                 'Gb0297383',\n",
      "                 'Gb0297319',\n",
      "                 'Gb0297334',\n",
      "                 'Gb0297336',\n",
      "                 'Gb0297395',\n",
      "                 'Gb0297374',\n",
      "                 'Gb0297327',\n",
      "                 'Gb0297384',\n",
      "                 'Gb0297400',\n",
      "                 'Gb0297457',\n",
      "                 'Gb0297449',\n",
      "                 'Gb0297461',\n",
      "                 'Gb0297367',\n",
      "                 'Gb0315918',\n",
      "                 'Gb0315735',\n",
      "                 'Gb0315967',\n",
      "                 'Gb0315969',\n",
      "                 'Gb0315883',\n",
      "                 'Gb0315758',\n",
      "                 'Gb0315760',\n",
      "                 'Gb0316050',\n",
      "                 'Gb0315931',\n",
      "                 'Gb0315936',\n",
      "                 'Gb0315947',\n",
      "                 'Gb0315413',\n",
      "                 'Gb0315434',\n",
      "                 'Gb0315473',\n",
      "                 'Gb0315600',\n",
      "                 'Gb0315606',\n",
      "                 'Gb0315489',\n",
      "                 'Gb0315580',\n",
      "                 'Gb0315761',\n",
      "                 'Gb0315773',\n",
      "                 'Gb0315643',\n",
      "                 'Gb0315648',\n",
      "                 'Gb0315797',\n",
      "                 'Gb0315827',\n",
      "                 'Gb0315716',\n",
      "                 'Gb0315756',\n",
      "                 'Gb0315901',\n",
      "                 'Gb0315906',\n",
      "                 'Gb0315702',\n",
      "                 'Gb0315839',\n",
      "                 'Gb0315857',\n",
      "                 'Gb0315858',\n",
      "                 'Gb0315861',\n",
      "                 'Gb0315953',\n",
      "                 'Gb0315991',\n",
      "                 'Gb0316012',\n",
      "                 'Gb0315984',\n",
      "                 'Gb0316021',\n",
      "                 'Gb0315838',\n",
      "                 'Gb0315791',\n",
      "                 'Gb0315712',\n",
      "                 'Gb0315718',\n",
      "                 'Gb0315752',\n",
      "                 'Gb0315646',\n",
      "                 'Gb0315875',\n",
      "                 'Gb0315720',\n",
      "                 'Gb0315525',\n",
      "                 'Gb0315477',\n",
      "                 'Gb0315479',\n",
      "                 'Gb0315533',\n",
      "                 'Gb0316045',\n",
      "                 'Gb0315864',\n",
      "                 'Gb0315755',\n",
      "                 'Gb0315417',\n",
      "                 'Gb0315911',\n",
      "                 'Gb0315771',\n",
      "                 'Gb0315655',\n",
      "                 'Gb0315962',\n",
      "                 'Gb0315964',\n",
      "                 'Gb0315443',\n",
      "                 'Gb0315496',\n",
      "                 'Gb0315640',\n",
      "                 'Gb0315548',\n",
      "                 'Gb0315662',\n",
      "                 'Gb0315698',\n",
      "                 'Gb0315705',\n",
      "                 'Gb0315706',\n",
      "                 'Gb0315442',\n",
      "                 'Gb0316033',\n",
      "                 'Gb0315588',\n",
      "                 'Gb0315696',\n",
      "                 'Gb0315544',\n",
      "                 'Gb0315546',\n",
      "                 'Gb0315441',\n",
      "                 'Gb0315575',\n",
      "                 'Gb0315582',\n",
      "                 'Gb0315591',\n",
      "                 'Gb0315611',\n",
      "                 'Gb0315462',\n",
      "                 'Gb0315604',\n",
      "                 'Gb0239937',\n",
      "                 'Gb0240059',\n",
      "                 'Gb0240065',\n",
      "                 'Gb0240111',\n",
      "                 'Gb0240126',\n",
      "                 'Gb0240110',\n",
      "                 'Gb0239934',\n",
      "                 'Gb0239964',\n",
      "                 'Gb0239965',\n",
      "                 'Gb0239986',\n",
      "                 'Gb0240010',\n",
      "                 'Gb0240015',\n",
      "                 'Gb0240026',\n",
      "                 'Gb0240027',\n",
      "                 'Gb0240041',\n",
      "                 'Gb0240045',\n",
      "                 'Gb0240050',\n",
      "                 'Gb0240071',\n",
      "                 'Gb0240079',\n",
      "                 'Gb0240090',\n",
      "                 'Gb0240097',\n",
      "                 'Gb0240119',\n",
      "                 'Gb0240120',\n",
      "                 'Gb0239930',\n",
      "                 'Gb0297485',\n",
      "                 'Gb0297320',\n",
      "                 'Gb0297347',\n",
      "                 'Gb0297351',\n",
      "                 'Gb0297389',\n",
      "                 'Gb0297424',\n",
      "                 'Gb0297343',\n",
      "                 'Gb0297328',\n",
      "                 'Gb0297350',\n",
      "                 'Gb0297368',\n",
      "                 'Gb0297387',\n",
      "                 'Gb0297373',\n",
      "                 'Gb0297392',\n",
      "                 'Gb0297378',\n",
      "                 'Gb0297418',\n",
      "                 'Gb0297420',\n",
      "                 'Gb0297432',\n",
      "                 'Gb0297468',\n",
      "                 'Gb0297341',\n",
      "                 'Gb0297361',\n",
      "                 'Gb0297443',\n",
      "                 'Gb0297484',\n",
      "                 'Gb0297330',\n",
      "                 'Gb0315917',\n",
      "                 'Gb0315926',\n",
      "                 'Gb0315727',\n",
      "                 'Gb0315730',\n",
      "                 'Gb0316042',\n",
      "                 'Gb0315872',\n",
      "                 'Gb0315763',\n",
      "                 'Gb0315939',\n",
      "                 'Gb0315940',\n",
      "                 'Gb0315798',\n",
      "                 'Gb0315799',\n",
      "                 'Gb0315979',\n",
      "                 'Gb0315428',\n",
      "                 'Gb0315453',\n",
      "                 'Gb0315426',\n",
      "                 'Gb0315568',\n",
      "                 'Gb0315599',\n",
      "                 'Gb0315603',\n",
      "                 'Gb0315526',\n",
      "                 'Gb0315538',\n",
      "                 'Gb0315584',\n",
      "                 'Gb0315608',\n",
      "                 'Gb0315691',\n",
      "                 'Gb0315811',\n",
      "                 'Gb0315828',\n",
      "                 'Gb0315754',\n",
      "                 'Gb0315865',\n",
      "                 'Gb0315711',\n",
      "                 'Gb0315902',\n",
      "                 'Gb0315817',\n",
      "                 'Gb0315808',\n",
      "                 'Gb0315813',\n",
      "                 'Gb0316024',\n",
      "                 'Gb0316020',\n",
      "                 'Gb0316032',\n",
      "                 'Gb0316047',\n",
      "                 'Gb0316040',\n",
      "                 'Gb0316044',\n",
      "                 'Gb0315844',\n",
      "                 'Gb0315845',\n",
      "                 'Gb0315860',\n",
      "                 'Gb0315717',\n",
      "                 'Gb0315418',\n",
      "                 'Gb0315681',\n",
      "                 'Gb0315874',\n",
      "                 'Gb0315508',\n",
      "                 'Gb0315510',\n",
      "                 'Gb0315781',\n",
      "                 'Gb0315880',\n",
      "                 'Gb0315515',\n",
      "                 'Gb0315520',\n",
      "                 'Gb0315747',\n",
      "                 'Gb0315478',\n",
      "                 'Gb0315529',\n",
      "                 'Gb0315531',\n",
      "                 'Gb0315532',\n",
      "                 'Gb0315534',\n",
      "                 'Gb0316048',\n",
      "                 'Gb0315471',\n",
      "                 'Gb0315474',\n",
      "                 'Gb0315540',\n",
      "                 'Gb0315742',\n",
      "                 'Gb0315743',\n",
      "                 'Gb0315656',\n",
      "                 'Gb0315963',\n",
      "                 'Gb0315866',\n",
      "                 'Gb0315488',\n",
      "                 'Gb0315500',\n",
      "                 'Gb0315633',\n",
      "                 'Gb0315636',\n",
      "                 'Gb0315641',\n",
      "                 'Gb0315549',\n",
      "                 'Gb0315578',\n",
      "                 'Gb0315663',\n",
      "                 'Gb0315667',\n",
      "                 'Gb0315684',\n",
      "                 'Gb0315856',\n",
      "                 'Gb0315987',\n",
      "                 'Gb0315996',\n",
      "                 'Gb0316008',\n",
      "                 'Gb0316009',\n",
      "                 'Gb0316041',\n",
      "                 'Gb0315596',\n",
      "                 'Gb0315657',\n",
      "                 'Gb0315661',\n",
      "                 'Gb0315615',\n",
      "                 'Gb0315621',\n",
      "                 'Gb0315463',\n",
      "                 'Gb0315589',\n",
      "                 'Gb0239923',\n",
      "                 'Gb0239943',\n",
      "                 'Gb0239944',\n",
      "                 'Gb0239947',\n",
      "                 'Gb0239951',\n",
      "                 'Gb0239977',\n",
      "                 'Gb0240057',\n",
      "                 'Gb0240089',\n",
      "                 'Gb0239932',\n",
      "                 'Gb0239936',\n",
      "                 'Gb0239941',\n",
      "                 'Gb0239952',\n",
      "                 'Gb0239954',\n",
      "                 'Gb0239962',\n",
      "                 'Gb0239966',\n",
      "                 'Gb0239990',\n",
      "                 'Gb0239991',\n",
      "                 'Gb0240001',\n",
      "                 'Gb0240014',\n",
      "                 'Gb0240017',\n",
      "                 'Gb0240022',\n",
      "                 'Gb0240025',\n",
      "                 'Gb0240028',\n",
      "                 'Gb0240049',\n",
      "                 'Gb0240061',\n",
      "                 'Gb0240073',\n",
      "                 'Gb0240075',\n",
      "                 'Gb0240077',\n",
      "                 'Gb0240087',\n",
      "                 'Gb0240100',\n",
      "                 'Gb0240101',\n",
      "                 'Gb0240118',\n",
      "                 'Gb0240128',\n",
      "                 'Gb0239919',\n",
      "                 'Gb0297407',\n",
      "                 'Gb0297313',\n",
      "                 'Gb0297314',\n",
      "                 'Gb0297338',\n",
      "                 'Gb0297382',\n",
      "                 'Gb0297429',\n",
      "                 'Gb0297471',\n",
      "                 'Gb0297323',\n",
      "                 'Gb0297365',\n",
      "                 'Gb0297431',\n",
      "                 'Gb0297376',\n",
      "                 'Gb0297472',\n",
      "                 'Gb0297355',\n",
      "                 'Gb0315916',\n",
      "                 'Gb0315919',\n",
      "                 'Gb0315921',\n",
      "                 'Gb0315923',\n",
      "                 'Gb0315736',\n",
      "                 'Gb0315739',\n",
      "                 'Gb0315978',\n",
      "                 'Gb0315888',\n",
      "                 'Gb0315502',\n",
      "                 'Gb0315898',\n",
      "                 'Gb0315930',\n",
      "                 'Gb0315933',\n",
      "                 'Gb0315934',\n",
      "                 'Gb0315935',\n",
      "                 'Gb0315793',\n",
      "                 'Gb0315784',\n",
      "                 'Gb0315980',\n",
      "                 'Gb0315412',\n",
      "                 'Gb0315951',\n",
      "                 'Gb0315429',\n",
      "                 'Gb0315595',\n",
      "                 'Gb0315563',\n",
      "                 'Gb0315627',\n",
      "                 'Gb0315703',\n",
      "                 'Gb0315818',\n",
      "                 'Gb0315832',\n",
      "                 'Gb0315748',\n",
      "                 'Gb0315859',\n",
      "                 'Gb0315867',\n",
      "                 'Gb0315750',\n",
      "                 'Gb0315851',\n",
      "                 'Gb0315645',\n",
      "                 'Gb0315846',\n",
      "                 'Gb0315803',\n",
      "                 'Gb0315876',\n",
      "                 'Gb0315878',\n",
      "                 'Gb0315882',\n",
      "                 'Gb0315992',\n",
      "                 'Gb0316025',\n",
      "                 'Gb0316026',\n",
      "                 'Gb0315834',\n",
      "                 'Gb0315843',\n",
      "                 'Gb0315862',\n",
      "                 'Gb0315437',\n",
      "                 'Gb0315713',\n",
      "                 'Gb0315774',\n",
      "                 'Gb0315753',\n",
      "                 'Gb0315725',\n",
      "                 'Gb0315506',\n",
      "                 'Gb0315509',\n",
      "                 'Gb0315884',\n",
      "                 'Gb0315886',\n",
      "                 'Gb0315890',\n",
      "                 'Gb0315518',\n",
      "                 'Gb0315519',\n",
      "                 'Gb0315524',\n",
      "                 'Gb0315749',\n",
      "                 'Gb0315450',\n",
      "                 'Gb0315465',\n",
      "                 'Gb0315492',\n",
      "                 'Gb0315535',\n",
      "                 'Gb0315536',\n",
      "                 'Gb0316046',\n",
      "                 'Gb0315469',\n",
      "                 'Gb0315483',\n",
      "                 'Gb0315904',\n",
      "                 'Gb0315766',\n",
      "                 'Gb0315673',\n",
      "                 'Gb0315486',\n",
      "                 'Gb0315562',\n",
      "                 'Gb0315577',\n",
      "                 'Gb0315847',\n",
      "                 'Gb0315666',\n",
      "                 'Gb0315855',\n",
      "                 'Gb0315464',\n",
      "                 'Gb0315994',\n",
      "                 'Gb0316007',\n",
      "                 'Gb0315658',\n",
      "                 'Gb0315699',\n",
      "                 'Gb0315541',\n",
      "                 'Gb0315543',\n",
      "                 'Gb0315560',\n",
      "                 'Gb0315440',\n",
      "                 'Gb0315620',\n",
      "                 'Gb0315623',\n",
      "                 'Gb0315550',\n",
      "                 'Gb0240054',\n",
      "                 'Gb0240074',\n",
      "                 'Gb0239957',\n",
      "                 'Gb0239864',\n",
      "                 'Gb0239976',\n",
      "                 'Gb0239984',\n",
      "                 'Gb0239994',\n",
      "                 'Gb0240007',\n",
      "                 'Gb0240008',\n",
      "                 'Gb0240013',\n",
      "                 'Gb0240024',\n",
      "                 'Gb0240029',\n",
      "                 'Gb0240030',\n",
      "                 'Gb0240031',\n",
      "                 'Gb0240036',\n",
      "                 'Gb0240058',\n",
      "                 'Gb0240060',\n",
      "                 'Gb0240062',\n",
      "                 'Gb0240104',\n",
      "                 'Gb0240105',\n",
      "                 'Gb0240123',\n",
      "                 'Gb0297487',\n",
      "                 'Gb0297318',\n",
      "                 'Gb0297335',\n",
      "                 'Gb0297337',\n",
      "                 'Gb0297399',\n",
      "                 'Gb0297375',\n",
      "                 'Gb0297465',\n",
      "                 'Gb0297356',\n",
      "                 'Gb0297423',\n",
      "                 'Gb0297440',\n",
      "                 'Gb0297450',\n",
      "                 'Gb0297439',\n",
      "                 'Gb0297454',\n",
      "                 'Gb0297448',\n",
      "                 'Gb0297452',\n",
      "                 'Gb0297453',\n",
      "                 'Gb0297459',\n",
      "                 'Gb0297370',\n",
      "                 'Gb0297315',\n",
      "                 'Gb0297324',\n",
      "                 'Gb0315927',\n",
      "                 'Gb0315738',\n",
      "                 'Gb0315971',\n",
      "                 'Gb0315977',\n",
      "                 'Gb0315651',\n",
      "                 'Gb0315870',\n",
      "                 'Gb0315894',\n",
      "                 'Gb0315897',\n",
      "                 'Gb0315765',\n",
      "                 'Gb0315932',\n",
      "                 'Gb0315941',\n",
      "                 'Gb0315942',\n",
      "                 'Gb0315790',\n",
      "                 'Gb0315411',\n",
      "                 'Gb0315444',\n",
      "                 'Gb0315448',\n",
      "                 'Gb0315487',\n",
      "                 'Gb0315557',\n",
      "                 'Gb0315490',\n",
      "                 'Gb0315527',\n",
      "                 'Gb0315564',\n",
      "                 'Gb0315628',\n",
      "                 'Gb0315769',\n",
      "                 'Gb0315775',\n",
      "                 'Gb0315780',\n",
      "                 'Gb0315609',\n",
      "                 'Gb0315618',\n",
      "                 'Gb0315669',\n",
      "                 'Gb0315677',\n",
      "                 'Gb0315837',\n",
      "                 'Gb0315745',\n",
      "                 'Gb0315715',\n",
      "                 'Gb0315724',\n",
      "                 'Gb0315812',\n",
      "                 'Gb0315820',\n",
      "                 'Gb0315954',\n",
      "                 'Gb0315958',\n",
      "                 'Gb0315952',\n",
      "                 'Gb0315986',\n",
      "                 'Gb0315816',\n",
      "                 'Gb0315997',\n",
      "                 'Gb0316017',\n",
      "                 'Gb0315825',\n",
      "                 'Gb0316030',\n",
      "                 'Gb0315794',\n",
      "                 'Gb0315458',\n",
      "                 'Gb0315955',\n",
      "                 'Gb0315671',\n",
      "                 'Gb0315680',\n",
      "                 'Gb0315688',\n",
      "                 'Gb0315877',\n",
      "                 'Gb0315650',\n",
      "                 'Gb0315514',\n",
      "                 'Gb0315516',\n",
      "                 'Gb0315451',\n",
      "                 'Gb0315539',\n",
      "                 'Gb0315482',\n",
      "                 'Gb0315726',\n",
      "                 'Gb0315668',\n",
      "                 'Gb0315491',\n",
      "                 'Gb0315497',\n",
      "                 'Gb0315498',\n",
      "                 'Gb0315637',\n",
      "                 'Gb0315701',\n",
      "                 'Gb0315853',\n",
      "                 'Gb0315414',\n",
      "                 'Gb0315993',\n",
      "                 'Gb0315583',\n",
      "                 'Gb0315586',\n",
      "                 'Gb0316014',\n",
      "                 'Gb0315660',\n",
      "                 'Gb0315566',\n",
      "                 'Gb0315410',\n",
      "                 'Gb0315709',\n",
      "                 'Gb0315598',\n",
      "                 'Gb0315607',\n",
      "                 'Gb0315558',\n",
      "                 'Gb0315593',\n",
      "                 'Gb0239926',\n",
      "                 'Gb0239960',\n",
      "                 'Gb0239968',\n",
      "                 'Gb0239969',\n",
      "                 'Gb0239979',\n",
      "                 'Gb0240046',\n",
      "                 'Gb0240086',\n",
      "                 'Gb0239931',\n",
      "                 'Gb0239945',\n",
      "                 'Gb0239961',\n",
      "                 'Gb0239974',\n",
      "                 'Gb0239998',\n",
      "                 'Gb0240011',\n",
      "                 'Gb0240016',\n",
      "                 'Gb0240021',\n",
      "                 'Gb0240032',\n",
      "                 'Gb0240037',\n",
      "                 'Gb0240038',\n",
      "                 'Gb0240040',\n",
      "                 'Gb0240044',\n",
      "                 'Gb0240047',\n",
      "                 'Gb0240052',\n",
      "                 'Gb0240053',\n",
      "                 'Gb0240096',\n",
      "                 'Gb0240107',\n",
      "                 'Gb0240115',\n",
      "                 'Gb0240130',\n",
      "                 'Gb0240131',\n",
      "                 'Gb0297475',\n",
      "                 'Gb0297486',\n",
      "                 'Gb0297393',\n",
      "                 'Gb0297402',\n",
      "                 'Gb0297342',\n",
      "                 'Gb0297369',\n",
      "                 'Gb0297427',\n",
      "                 'Gb0297435',\n",
      "                 'Gb0297438',\n",
      "                 'Gb0297325',\n",
      "                 'Gb0297357',\n",
      "                 'Gb0297462',\n",
      "                 'Gb0297340',\n",
      "                 'Gb0297359',\n",
      "                 'Gb0297371',\n",
      "                 'Gb0297419',\n",
      "                 'Gb0297363',\n",
      "                 'Gb0297380',\n",
      "                 'Gb0297381',\n",
      "                 'Gb0297456',\n",
      "                 'Gb0297458',\n",
      "                 'Gb0297445',\n",
      "                 'Gb0297405',\n",
      "                 'Gb0297329',\n",
      "                 'Gb0297451',\n",
      "                 'Gb0297483',\n",
      "                 'Gb0297322',\n",
      "                 'Gb0315924',\n",
      "                 'Gb0315925',\n",
      "                 'Gb0315928',\n",
      "                 'Gb0315731',\n",
      "                 'Gb0315733',\n",
      "                 'Gb0315966',\n",
      "                 'Gb0315973',\n",
      "                 'Gb0315974',\n",
      "                 'Gb0315647',\n",
      "                 'Gb0315929',\n",
      "                 'Gb0315937',\n",
      "                 'Gb0315785',\n",
      "                 'Gb0315786',\n",
      "                 'Gb0315948',\n",
      "                 'Gb0315949',\n",
      "                 'Gb0315433',\n",
      "                 'Gb0315472',\n",
      "                 'Gb0315475',\n",
      "                 'Gb0315427',\n",
      "                 'Gb0315561',\n",
      "                 'Gb0315569',\n",
      "                 'Gb0315507',\n",
      "                 'Gb0315630',\n",
      "                 'Gb0315685',\n",
      "                 'Gb0315829',\n",
      "                 'Gb0316049',\n",
      "                 'Gb0315675',\n",
      "                 'Gb0315676',\n",
      "                 'Gb0315682',\n",
      "                 'Gb0315445',\n",
      "                 'Gb0315849',\n",
      "                 'Gb0315957',\n",
      "                 'Gb0315879',\n",
      "                 'Gb0315892',\n",
      "                 'Gb0315907',\n",
      "                 'Gb0315814',\n",
      "                 'Gb0316038',\n",
      "                 'Gb0315821',\n",
      "                 'Gb0315830',\n",
      "                 'Gb0315831',\n",
      "                 'Gb0315833',\n",
      "                 'Gb0315840',\n",
      "                 'Gb0315457',\n",
      "                 'Gb0315788',\n",
      "                 'Gb0315674',\n",
      "                 'Gb0315503',\n",
      "                 'Gb0315779',\n",
      "                 'Gb0315873',\n",
      "                 'Gb0315452',\n",
      "                 'Gb0315459',\n",
      "                 'Gb0315517',\n",
      "                 'Gb0315449',\n",
      "                 'Gb0315470',\n",
      "                 'Gb0315495',\n",
      "                 'Gb0315467',\n",
      "                 'Gb0315468',\n",
      "                 'Gb0315480',\n",
      "                 'Gb0315909',\n",
      "                 'Gb0315741',\n",
      "                 'Gb0315744',\n",
      "                 'Gb0315438',\n",
      "                 'Gb0315653',\n",
      "                 'Gb0315485',\n",
      "                 'Gb0315501',\n",
      "                 'Gb0315625',\n",
      "                 'Gb0315547',\n",
      "                 'Gb0315679',\n",
      "                 'Gb0315695',\n",
      "                 'Gb0315415',\n",
      "                 'Gb0316005',\n",
      "                 'Gb0316031',\n",
      "                 'Gb0315612',\n",
      "                 'Gb0315629',\n",
      "                 'Gb0315571',\n",
      "                 'Gb0315572',\n",
      "                 'Gb0315592',\n",
      "                 'Gb0315602',\n",
      "                 'Gb0315642',\n",
      "                 'Gb0315556',\n",
      "                 'Gb0239939',\n",
      "                 'Gb0239956',\n",
      "                 'Gb0239949',\n",
      "                 'Gb0239973',\n",
      "                 'Gb0239992',\n",
      "                 'Gb0240078',\n",
      "                 'Gb0239946',\n",
      "                 'Gb0239948',\n",
      "                 'Gb0239950',\n",
      "                 'Gb0239953',\n",
      "                 'Gb0239971',\n",
      "                 'Gb0239978',\n",
      "                 'Gb0239987',\n",
      "                 'Gb0240005',\n",
      "                 'Gb0240023',\n",
      "                 'Gb0240035',\n",
      "                 'Gb0240064',\n",
      "                 'Gb0240067',\n",
      "                 'Gb0240076',\n",
      "                 'Gb0240081',\n",
      "                 'Gb0240092',\n",
      "                 'Gb0240095',\n",
      "                 'Gb0240098',\n",
      "                 'Gb0240103',\n",
      "                 'Gb0240122',\n",
      "                 'Gb0239920',\n",
      "                 'Gb0239929',\n",
      "                 'Gb0297390',\n",
      "                 'Gb0297421',\n",
      "                 'Gb0297441',\n",
      "                 'Gb0297345',\n",
      "                 'Gb0297348',\n",
      "                 'Gb0297404',\n",
      "                 'Gb0297417',\n",
      "                 'Gb0297425',\n",
      "                 'Gb0297428',\n",
      "                 'Gb0297434',\n",
      "                 'Gb0297364',\n",
      "                 'Gb0297422',\n",
      "                 'Gb0297385',\n",
      "                 'Gb0297460',\n",
      "                 'Gb0297477',\n",
      "                 'Gb0297321',\n",
      "                 'Gb0315734',\n",
      "                 'Gb0315740',\n",
      "                 'Gb0315970',\n",
      "                 'Gb0315899',\n",
      "                 'Gb0315981',\n",
      "                 'Gb0315982',\n",
      "                 'Gb0315421',\n",
      "                 'Gb0315424',\n",
      "                 'Gb0315420',\n",
      "                 'Gb0315513',\n",
      "                 'Gb0315504',\n",
      "                 'Gb0315512',\n",
      "                 'Gb0315528',\n",
      "                 'Gb0315555',\n",
      "                 'Gb0315565',\n",
      "                 'Gb0315631',\n",
      "                 'Gb0315719',\n",
      "                 'Gb0315782',\n",
      "                 'Gb0315635',\n",
      "                 'Gb0315697',\n",
      "                 'Gb0315700',\n",
      "                 'Gb0315801',\n",
      "                 'Gb0315819',\n",
      "                 'Gb0315789',\n",
      "                 'Gb0315723',\n",
      "                 'Gb0315912',\n",
      "                 'Gb0315683',\n",
      "                 'Gb0315686',\n",
      "                 'Gb0315823',\n",
      "                 'Gb0315850',\n",
      "                 'Gb0315961',\n",
      "                 'Gb0315802',\n",
      "                 'Gb0315989',\n",
      "                 'Gb0315990',\n",
      "                 'Gb0315995',\n",
      "                 'Gb0315815',\n",
      "                 'Gb0315822',\n",
      "                 'Gb0315824',\n",
      "                 'Gb0315455',\n",
      "                 'Gb0315694',\n",
      "                 'Gb0315757',\n",
      "                 'Gb0315670',\n",
      "                 'Gb0315687',\n",
      "                 'Gb0315446',\n",
      "                 'Gb0315505',\n",
      "                 'Gb0315778',\n",
      "                 'Gb0315423',\n",
      "                 'Gb0315521',\n",
      "                 'Gb0315493',\n",
      "                 'Gb0315792',\n",
      "                 'Gb0315537',\n",
      "                 'Gb0315908',\n",
      "                 'Gb0315910',\n",
      "                 'Gb0315913',\n",
      "                 'Gb0315768',\n",
      "                 'Gb0315654',\n",
      "                 'Gb0315499',\n",
      "                 'Gb0315626',\n",
      "                 'Gb0315567',\n",
      "                 'Gb0315579',\n",
      "                 'Gb0315665',\n",
      "                 'Gb0315848',\n",
      "                 'Gb0315854',\n",
      "                 'Gb0315416',\n",
      "                 'Gb0316010',\n",
      "                 'Gb0316013',\n",
      "                 'Gb0316019',\n",
      "                 'Gb0316034',\n",
      "                 'Gb0316035',\n",
      "                 'Gb0316036',\n",
      "                 'Gb0315587',\n",
      "                 'Gb0315704',\n",
      "                 'Gb0315545',\n",
      "                 'Gb0315559',\n",
      "                 'Gb0315570',\n",
      "                 'Gb0315430',\n",
      "                 'Gb0315431',\n",
      "                 'Gb0315439',\n",
      "                 'Gb0315605',\n",
      "                 'Gb0315622',\n",
      "                 'Gb0315461',\n",
      "                 'Gb0315585',\n",
      "                 'Gb0315632'],\n",
      "  'contacts': [{'email': 'janet.jansson@pnnl.gov',\n",
      "                'jgiSsoId': '62',\n",
      "                'name': 'Janet Jansson',\n",
      "                'roles': ['other', 'Study PI', 'PI']},\n",
      "               {'email': 'emily.graham@pnnl.gov',\n",
      "                'jgiSsoId': '52018',\n",
      "                'name': 'Emily Graham',\n",
      "                'roles': ['other']}],\n",
      "  'description': 'The Earth Microbiome Project Multi-omics component (EMP500) '\n",
      "                 'involves amplicon and shotgun metagenomic sequencing and '\n",
      "                 'metabolomic profiling of over five hundred microbial '\n",
      "                 'communities from diverse environments on our planet. We '\n",
      "                 'developed new protocols for shotgun metagenomic sequencing '\n",
      "                 'and assembly, with the goal of applying this workflow to a '\n",
      "                 'range of environmental samples, combined with metabolomic '\n",
      "                 'profiling. 16S, 18S, and ITS amplicon sequencing was done in '\n",
      "                 'addition. We acquired a set of >500 fresh environmental '\n",
      "                 'samples across a range of habitats, with the help of the EMP '\n",
      "                 'network of collaborators. A biobank of frozen aliquots for '\n",
      "                 'all samples is maintained at UCSD (soil samples stored at '\n",
      "                 'the Pacific Northwest National Laboratory) for future '\n",
      "                 'methods testing and analysis.',\n",
      "  'modDate': '2022-04-08',\n",
      "  'studyGoldId': 'Gs0154244',\n",
      "  'studyName': 'Earth Microbiome Project Multi-omics (EMP500)'}]\n"
     ]
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.202240Z",
     "start_time": "2025-02-26T02:56:09.199611Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_bioproject_accession = \"PRJEB42019\"\n",
    "emp500_gold_study_id = \"Gs0154244\""
   ],
   "id": "50de93b92c01a583",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.253141Z",
     "start_time": "2025-02-26T02:56:09.245071Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_bioproj_biosample_links_cursor = core.fetch_mongodb_records_by_path(\n",
    "    client,\n",
    "    NCBI_METADATA,\n",
    "    \"sra_biosamples_bioprojects\",\n",
    "    \"bioproject_accession\",\n",
    "    emp500_bioproject_accession\n",
    ")\n",
    "\n",
    "emp500_bioproj_biosample_links = [\n",
    "    {key: value for key, value in document.items() if key!= '_id'}\n",
    "    for document in emp500_bioproj_biosample_links_cursor\n",
    "]"
   ],
   "id": "e9df4cf751056d92",
   "outputs": [],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.300020Z",
     "start_time": "2025-02-26T02:56:09.296609Z"
    }
   },
   "cell_type": "code",
   "source": "emp500_biosample_accessions = [doc[\"biosample_accession\"] for doc in emp500_bioproj_biosample_links]",
   "id": "71dc31be94a21af6",
   "outputs": [],
   "execution_count": 22
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.413680Z",
     "start_time": "2025-02-26T02:56:09.343031Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_ncbi_biosamples_cursor = core.fetch_mongodb_records_by_path_in(\n",
    "    client,\n",
    "    NCBI_METADATA,\n",
    "    \"biosamples\",\n",
    "    \"accession\",\n",
    "    emp500_biosample_accessions\n",
    ")\n",
    "\n",
    "emp500_ncbi_biosamples = [\n",
    "    {key: value for key, value in document.items() if key!= '_id'}\n",
    "    for document in emp500_ncbi_biosamples_cursor\n",
    "]"
   ],
   "id": "ecc40b098893970b",
   "outputs": [],
   "execution_count": 23
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.429303Z",
     "start_time": "2025-02-26T02:56:09.426091Z"
    }
   },
   "cell_type": "code",
   "source": "emp500_ncbi_biosamples = core.lod_to_dod(emp500_ncbi_biosamples, \"accession\")",
   "id": "45430ed50eff1c4a",
   "outputs": [],
   "execution_count": 24
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.486789Z",
     "start_time": "2025-02-26T02:56:09.483527Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(emp500_ncbi_biosamples))",
   "id": "5b398f7b932a79d4",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1024\n"
     ]
    }
   ],
   "execution_count": 25
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.532614Z",
     "start_time": "2025-02-26T02:56:09.529048Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_gold_study_cursor = core.fetch_mongodb_records_by_path(\n",
    "    client,\n",
    "    NCBI_METADATA,\n",
    "    \"sra_biosamples_bioprojects\",\n",
    "    \"bioproject_accession\",\n",
    "    emp500_bioproject_accession\n",
    ")\n",
    "\n",
    "emp500_bioproj_biosample_links = [\n",
    "    {key: value for key, value in document.items() if key!= '_id'}\n",
    "    for document in emp500_bioproj_biosample_links_cursor\n",
    "]"
   ],
   "id": "e1245569efe46cf2",
   "outputs": [],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.580854Z",
     "start_time": "2025-02-26T02:56:09.577118Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Search for GOLD studies\n",
    "emp500_gold_study_cursor = core.fetch_mongodb_records_by_path(\n",
    "    client,\n",
    "    GOLD_METADATA,\n",
    "    \"studies\",\n",
    "    \"studyGoldId\",\n",
    "    emp500_gold_study_id\n",
    ")\n",
    "\n",
    "emp500_gold_study_list = list(emp500_gold_study_cursor)"
   ],
   "id": "39ccc89438af92eb",
   "outputs": [],
   "execution_count": 27
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.626404Z",
     "start_time": "2025-02-26T02:56:09.623626Z"
    }
   },
   "cell_type": "code",
   "source": "emp500_gold_biosamples_list = emp500_gold_study_list[0].get(\"biosamples\", [])",
   "id": "45cbec7d52d4ffd0",
   "outputs": [],
   "execution_count": 28
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.701601Z",
     "start_time": "2025-02-26T02:56:09.669074Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_gold_biosamples_cursor = core.fetch_mongodb_records_by_path_in(\n",
    "    client,\n",
    "    GOLD_METADATA,\n",
    "    \"biosamples\",\n",
    "    \"biosampleGoldId\",\n",
    "    emp500_gold_biosamples_list\n",
    ")\n",
    "\n",
    "emp500_gold_biosamples = [\n",
    "    {key: value for key, value in document.items() if key!= '_id'}\n",
    "    for document in emp500_gold_biosamples_cursor\n",
    "]"
   ],
   "id": "b3aeeb3fe7d86093",
   "outputs": [],
   "execution_count": 29
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.718410Z",
     "start_time": "2025-02-26T02:56:09.715301Z"
    }
   },
   "cell_type": "code",
   "source": "emp500_gold_biosamples = core.lod_to_dod(emp500_gold_biosamples, \"biosampleGoldId\")",
   "id": "237ed8beb5fa2301",
   "outputs": [],
   "execution_count": 30
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:09.765297Z",
     "start_time": "2025-02-26T02:56:09.760953Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(emp500_gold_biosamples))",
   "id": "1ef743f8134eabe7",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1024\n"
     ]
    }
   ],
   "execution_count": 31
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Biosamples have been fetched from NCBI and GOLD at this point",
   "id": "24764c981305714e"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.003418Z",
     "start_time": "2025-02-26T02:56:09.814245Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_gold_seq_projs_cursor = core.fetch_mongodb_records_by_path_in(\n",
    "    client,\n",
    "    GOLD_METADATA,\n",
    "    \"projects\",\n",
    "    \"biosampleGoldId\",\n",
    "    emp500_gold_biosamples_list\n",
    ")\n",
    "\n",
    "emp500_gold_seq_projs = [\n",
    "    {key: value for key, value in document.items() if key!= '_id'}\n",
    "    for document in emp500_gold_seq_projs_cursor\n",
    "]"
   ],
   "id": "da8d466784f36710",
   "outputs": [],
   "execution_count": 32
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.022997Z",
     "start_time": "2025-02-26T02:56:10.020170Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(emp500_gold_seq_projs))",
   "id": "918e3a84dddc99ca",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1836\n"
     ]
    }
   ],
   "execution_count": 33
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.081584Z",
     "start_time": "2025-02-26T02:56:10.072289Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_seq_proj_summaries_by_accession = defaultdict(dict)\n",
    "\n",
    "for record in emp500_gold_seq_projs:\n",
    "    bioproj_accession = record['ncbiBioProjectAccession']\n",
    "    biosample_accession = record['ncbiBioSampleAccession']\n",
    "    biosample_gold_id = record['biosampleGoldId']\n",
    "    sra_ids = record['sraExperimentIds']\n",
    "\n",
    "    if biosample_accession in emp500_seq_proj_summaries_by_accession:\n",
    "        # Check if other fields are the same\n",
    "        existing_record = emp500_seq_proj_summaries_by_accession[biosample_accession]\n",
    "        if (\n",
    "                existing_record['ncbiBioProjectAccession'] == bioproj_accession and\n",
    "                existing_record['biosampleGoldId'] == biosample_gold_id\n",
    "        ):\n",
    "            # Update existing record with new sraExperimentIds\n",
    "            existing_record['sraExperimentIds'].update(sra_ids)\n",
    "        else:\n",
    "            # Handle conflicting data (e.g., raise an error or log a warning)\n",
    "            print(f\"Conflicting data for accession {biosample_accession}: {record} vs {existing_record}\")\n",
    "    else:\n",
    "        # Create a new record\n",
    "        emp500_seq_proj_summaries_by_accession[biosample_accession] = {\n",
    "            'biosampleGoldId': biosample_gold_id,\n",
    "            'ncbiBioProjectAccession': bioproj_accession,\n",
    "            'sraExperimentIds': set(sra_ids)\n",
    "        }"
   ],
   "id": "573b7b28fa1ac9b8",
   "outputs": [],
   "execution_count": 34
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.127664Z",
     "start_time": "2025-02-26T02:56:10.124614Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(emp500_seq_proj_summaries_by_accession))",
   "id": "3aec7dfd83ee4ab2",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1024\n"
     ]
    }
   ],
   "execution_count": 35
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.189036Z",
     "start_time": "2025-02-26T02:56:10.180559Z"
    }
   },
   "cell_type": "code",
   "source": [
    "emp500_gold_integrated_biosamples = {}\n",
    "\n",
    "for accession, seq_proj_summary in emp500_seq_proj_summaries_by_accession.items():\n",
    "    biosample_gold_id = seq_proj_summary.get('biosampleGoldId')  # Get gold ID\n",
    "    if biosample_gold_id:\n",
    "      gold_biosample_data = emp500_gold_biosamples.get(biosample_gold_id)\n",
    "      if gold_biosample_data:\n",
    "        # Merge the dictionaries (you can customize how to handle conflicts)\n",
    "        integrated_data = {**seq_proj_summary, **gold_biosample_data}\n",
    "        emp500_gold_integrated_biosamples[accession] = integrated_data\n"
   ],
   "id": "90dcbf6c14924018",
   "outputs": [],
   "execution_count": 36
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "todo: add label lookup and confirmation and biome, abp, env_medium, soil and water checks for GOLD samples in `emp500_gold_integrated_biosamples`",
   "id": "a30f5edf05eba67"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.260695Z",
     "start_time": "2025-02-26T02:56:10.231302Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for biosample_gold_id, gold_biosample in emp500_gold_integrated_biosamples.items():\n",
    "    ebs_true_curie = gold_biosample['envoBroadScale']['id'].replace(\"_\", \":\")\n",
    "    els_true_curie = gold_biosample['envoLocalScale']['id'].replace(\"_\", \":\")\n",
    "    em_true_curie = gold_biosample['envoMedium']['id'].replace(\"_\", \":\")\n",
    "\n",
    "    gold_biosample['envoBroadScale_label_lu'] = envo_curies_to_label.get(ebs_true_curie)\n",
    "    gold_biosample['envoBroadScale_label_exact_match'] = gold_biosample['envoBroadScale_label_lu'] == \\\n",
    "                                                         gold_biosample['envoBroadScale']['label']\n",
    "    gold_biosample['envoBroadScale_biome'] = ebs_true_curie in biome_curies\n",
    "    gold_biosample['envoLocalScale_label_lu'] = envo_curies_to_label.get(els_true_curie)\n",
    "    gold_biosample['envoLocalScale_label_exact_match'] = gold_biosample['envoLocalScale_label_lu'] == \\\n",
    "                                                         gold_biosample['envoLocalScale']['label']\n",
    "    gold_biosample['envoLocalScale_abp_subset'] = els_true_curie in non_biome_non_env_mat_abp_curies\n",
    "    gold_biosample['envoMedium_label_lu'] = envo_curies_to_label.get(em_true_curie)\n",
    "    gold_biosample['envoMedium_label_exact_match'] = gold_biosample['envoMedium_label_lu'] == \\\n",
    "                                                     gold_biosample['envoMedium']['label']\n",
    "    gold_biosample['envoMedium_env_mat'] = em_true_curie in env_mat_curies\n",
    "    gold_biosample['envoMedium_soil'] = em_true_curie in soil_curies\n",
    "    gold_biosample['envoMedium_liq_water'] = em_true_curie in liq_water_curies\n",
    "    gold_biosample['envoMedium_sediment'] = em_true_curie in sediment_curies\n",
    "    # gold_biosample['envoMedium_feces'] = em_true_curie in feces_curies\n",
    "    gold_biosample['envoMedium_fecal_material'] = em_true_curie in fecal_material_curies\n",
    "    gold_biosample['envoMedium_animal_manure'] = em_true_curie in animal_manure_curies\n",
    "    gold_biosample['envoMedium_bodily_fluid_material'] = em_true_curie in bodily_fluid_material_curies\n",
    "    gold_biosample['envoMedium_sludge'] = em_true_curie in sludge_curies\n",
    "    gold_biosample['envoMedium_petroleum'] = em_true_curie in petroleum_curies\n",
    "\n"
   ],
   "id": "83a7d04922dfab94",
   "outputs": [],
   "execution_count": 37
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.295062Z",
     "start_time": "2025-02-26T02:56:10.291264Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_harmonized_value(ncbi_biosample, harmonized_name):\n",
    "  \"\"\"\n",
    "  Extracts the 'content' from ncbi_biosample['Attributes']['Attribute']\n",
    "  where 'harmonized_name' matches the given value.\n",
    "\n",
    "  Args:\n",
    "    ncbi_biosample: The dictionary containing biosample data.\n",
    "    harmonized_name: The 'harmonized_name' to search for.\n",
    "\n",
    "  Returns:\n",
    "    The 'content' value if a match is found, otherwise None.\n",
    "  \"\"\"\n",
    "  if 'Attributes' in ncbi_biosample and 'Attribute' in ncbi_biosample['Attributes']:\n",
    "    for attribute in ncbi_biosample['Attributes']['Attribute']:\n",
    "      if attribute.get('harmonized_name') == harmonized_name:\n",
    "        return attribute.get('content')\n",
    "  return None"
   ],
   "id": "79c0625988122bcc",
   "outputs": [],
   "execution_count": 38
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.380485Z",
     "start_time": "2025-02-26T02:56:10.344442Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo: is it worth keeping track of any package knowledge from NCBI or GOLD?\n",
    "\n",
    "emp500_integrated_biosample_summaries = {}\n",
    "\n",
    "ncbi_env_content_counter = Counter()  # Store counts of all strings used as env triad values\n",
    "\n",
    "for accession, ncbi_biosample in emp500_ncbi_biosamples.items():\n",
    "    integrated_summary_dict = dict()\n",
    "    gold_biosample = emp500_gold_integrated_biosamples.get(accession)  # Get gold ID\n",
    "    if gold_biosample:\n",
    "        # print(f\"Found matching GOLD data for {gold_biosample['biosampleGoldId']}\")\n",
    "        integrated_summary_dict['accession'] = accession\n",
    "        integrated_summary_dict['env_broad_scale'] = get_harmonized_value(ncbi_biosample, 'env_broad_scale')\n",
    "        ncbi_env_content_counter[integrated_summary_dict['env_broad_scale']] += 1  # Count occurrences\n",
    "        integrated_summary_dict['env_local_scale'] = get_harmonized_value(ncbi_biosample, 'env_local_scale')\n",
    "        ncbi_env_content_counter[integrated_summary_dict['env_local_scale']] += 1  # Count occurrences\n",
    "        integrated_summary_dict['env_medium'] = get_harmonized_value(ncbi_biosample, 'env_medium')\n",
    "        ncbi_env_content_counter[integrated_summary_dict['env_medium']] += 1  # Count occurrences\n",
    "        integrated_summary_dict['biosampleGoldId'] = gold_biosample['biosampleGoldId']\n",
    "        integrated_summary_dict['ncbiBioProjectAccession'] = gold_biosample['ncbiBioProjectAccession']\n",
    "        integrated_summary_dict['sraExperimentIds'] = \"|\".join(sorted(gold_biosample['sraExperimentIds']))\n",
    "        integrated_summary_dict['ecosystemPathId'] = gold_biosample['ecosystemPathId']\n",
    "        integrated_summary_dict['ecosystem'] = gold_biosample['ecosystem']\n",
    "        integrated_summary_dict['ecosystemCategory'] = gold_biosample['ecosystemCategory']\n",
    "        integrated_summary_dict['ecosystemType'] = gold_biosample['ecosystemType']\n",
    "        integrated_summary_dict['ecosystemSubtype'] = gold_biosample['ecosystemSubtype']\n",
    "        integrated_summary_dict['specificEcosystem'] = gold_biosample['specificEcosystem']\n",
    "\n",
    "        integrated_summary_dict['envoBroadScale_id'] = gold_biosample['envoBroadScale']['id'].replace(\"_\", \":\")\n",
    "        integrated_summary_dict['envoBroadScale_label'] = gold_biosample['envoBroadScale']['label']\n",
    "\n",
    "        integrated_summary_dict['envoBroadScale_label_lu'] = gold_biosample['envoBroadScale_label_lu']\n",
    "        integrated_summary_dict['envoBroadScale_label_exact_match'] = gold_biosample['envoBroadScale_label_exact_match']\n",
    "        integrated_summary_dict['envoBroadScale_biome'] = gold_biosample['envoBroadScale_biome']\n",
    "\n",
    "        integrated_summary_dict['envoLocalScale_id'] = gold_biosample['envoLocalScale']['id'].replace(\"_\", \":\")\n",
    "        integrated_summary_dict['envoLocalScale_label'] = gold_biosample['envoLocalScale']['label']\n",
    "\n",
    "        integrated_summary_dict['envoLocalScale_label_lu'] = gold_biosample['envoLocalScale_label_lu']\n",
    "        integrated_summary_dict['envoLocalScale_label_exact_match'] = gold_biosample['envoLocalScale_label_exact_match']\n",
    "        integrated_summary_dict['envoLocalScale_abp_subset'] = gold_biosample['envoLocalScale_abp_subset']\n",
    "\n",
    "        integrated_summary_dict['envoMedium_id'] = gold_biosample['envoMedium']['id'].replace(\"_\", \":\")\n",
    "        integrated_summary_dict['envoMedium_label'] = gold_biosample['envoMedium']['label']\n",
    "        integrated_summary_dict['envoMedium_label_lu'] = gold_biosample['envoMedium_label_lu']\n",
    "        integrated_summary_dict['envoMedium_label_exact_match'] = gold_biosample['envoMedium_label_exact_match']\n",
    "        integrated_summary_dict['envoMedium_env_mat'] = gold_biosample['envoMedium_env_mat']\n",
    "        integrated_summary_dict['envoMedium_soil'] = gold_biosample['envoMedium_soil']\n",
    "        integrated_summary_dict['envoMedium_liq_water'] = gold_biosample['envoMedium_liq_water']\n",
    "        integrated_summary_dict['envoMedium_sediment'] = gold_biosample['envoMedium_sediment']\n",
    "        # integrated_summary_dict['envoMedium_feces'] = gold_biosample['envoMedium_feces']\n",
    "        integrated_summary_dict['envoMedium_fecal_material'] = gold_biosample['envoMedium_fecal_material']\n",
    "        integrated_summary_dict['envoMedium_animal_manure'] = gold_biosample['envoMedium_animal_manure']\n",
    "        integrated_summary_dict['envoMedium_bodily_fluid_material'] = gold_biosample['envoMedium_bodily_fluid_material']\n",
    "        integrated_summary_dict['envoMedium_sludge'] = gold_biosample['envoMedium_sludge']\n",
    "        integrated_summary_dict['envoMedium_petroleum'] = gold_biosample['envoMedium_petroleum']\n",
    "        integrated_summary_dict['envoMedium_accounted_for'] = gold_biosample['envoMedium_soil'] or gold_biosample[\n",
    "            'envoMedium_liq_water'] or gold_biosample['envoMedium_sediment'] or \\\n",
    "                                                              gold_biosample['envoMedium_fecal_material'] or \\\n",
    "                                                              gold_biosample['envoMedium_animal_manure'] or \\\n",
    "                                                              gold_biosample['envoMedium_bodily_fluid_material'] or \\\n",
    "                                                              gold_biosample['envoMedium_sludge'] or gold_biosample[\n",
    "                                                                  'envoMedium_petroleum']\n",
    "\n",
    "        integrated_summary_dict['addDate'] = gold_biosample['addDate']\n",
    "        integrated_summary_dict['modDate'] = gold_biosample['modDate']\n",
    "\n",
    "        emp500_integrated_biosample_summaries[accession] = integrated_summary_dict\n",
    "    else:\n",
    "        print(f\"No matching GOLD data found for {accession}\")"
   ],
   "id": "feb2814ea2c5425e",
   "outputs": [],
   "execution_count": 39
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.425201Z",
     "start_time": "2025-02-26T02:56:10.421747Z"
    }
   },
   "cell_type": "code",
   "source": "print(len(ncbi_env_content_counter))",
   "id": "80769ed3fb15eb7c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "112\n"
     ]
    }
   ],
   "execution_count": 40
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:10.478790Z",
     "start_time": "2025-02-26T02:56:10.469527Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_env_content_counter",
   "id": "22f94c962f0301ad",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'urban biome': 285,\n",
       "         'research facility': 191,\n",
       "         'feces': 184,\n",
       "         'sterile water': 178,\n",
       "         'organic material': 166,\n",
       "         'host-associated habitat': 133,\n",
       "         'soil': 132,\n",
       "         'marine benthic feature': 125,\n",
       "         'temperate mixed forest biome': 112,\n",
       "         'marine benthic biome': 90,\n",
       "         'animal-associated habitat': 84,\n",
       "         'marine coral reef biome': 51,\n",
       "         'sediment': 45,\n",
       "         'subpolar coniferous forest biome': 45,\n",
       "         'marine biome': 42,\n",
       "         'marine salt marsh biome': 40,\n",
       "         'desert biome': 40,\n",
       "         'marine sediment': 38,\n",
       "         'marine reef biome': 30,\n",
       "         'tropical broadleaf forest biome': 28,\n",
       "         'insecta-associated habitat': 28,\n",
       "         'lake sediment': 27,\n",
       "         'montane desert': 26,\n",
       "         'alpine soil': 26,\n",
       "         'kelp forest': 25,\n",
       "         'saline marsh': 24,\n",
       "         'temperate coniferous forest biome': 22,\n",
       "         'dense settlement biome': 22,\n",
       "         'coral reef': 22,\n",
       "         'sea water': 22,\n",
       "         'forest biome': 21,\n",
       "         'volcanic feature': 21,\n",
       "         'volcanic soil': 21,\n",
       "         'forest': 21,\n",
       "         'Small lake biome': 20,\n",
       "         'pond bed': 20,\n",
       "         'ocean': 20,\n",
       "         'neritic epipelagic zone biome': 19,\n",
       "         'tropical desert biome': 19,\n",
       "         'stratum': 18,\n",
       "         'peat soil': 18,\n",
       "         'bioreactor': 17,\n",
       "         'biosolids': 17,\n",
       "         'marine pelagic biome': 17,\n",
       "         'mine': 16,\n",
       "         'ocean water': 16,\n",
       "         'temperate coniferous forest': 16,\n",
       "         'cropland biome': 16,\n",
       "         'tundra biome': 15,\n",
       "         'glacier': 15,\n",
       "         'intestine environment': 15,\n",
       "         'oil reservior': 14,\n",
       "         'petroleum': 14,\n",
       "         'surface soil': 14,\n",
       "         'desert sand': 13,\n",
       "         'rock': 13,\n",
       "         'cell culture': 13,\n",
       "         'coastal water body': 12,\n",
       "         'freshwater biome': 12,\n",
       "         'montane grasslands and shrubland biome': 12,\n",
       "         'steppe soil': 12,\n",
       "         'Small river biome': 10,\n",
       "         'tributary': 10,\n",
       "         'stream sediment': 10,\n",
       "         'gravel field': 10,\n",
       "         'gravel': 10,\n",
       "         'cropland': 10,\n",
       "         'anthropogenic terrestrial biome': 10,\n",
       "         'anthropogenic contamination feature': 10,\n",
       "         'coast': 10,\n",
       "         'anaerobic bioreactor': 9,\n",
       "         'anaerobic sludge': 9,\n",
       "         'coastal water': 9,\n",
       "         'cliff': 9,\n",
       "         'wetland': 9,\n",
       "         'dune': 9,\n",
       "         'grassland biome': 7,\n",
       "         'intermittent pond': 7,\n",
       "         'temperate shrubland': 7,\n",
       "         'water': 7,\n",
       "         'forest soil': 6,\n",
       "         'animal manure': 6,\n",
       "         'groundwater': 6,\n",
       "         'marine subtidal rocky reef biome': 6,\n",
       "         'tropical moist broadleaf forest': 6,\n",
       "         'pitcher plant fluid': 6,\n",
       "         'oil palm plantation': 6,\n",
       "         'marine reef': 5,\n",
       "         'mud': 5,\n",
       "         'rangeland biome': 5,\n",
       "         'rangeland': 5,\n",
       "         'planted forest': 5,\n",
       "         'temperate desert biome': 4,\n",
       "         'temperate grassland biome': 4,\n",
       "         'grassland soil': 4,\n",
       "         'mucus': 4,\n",
       "         'sebum': 3,\n",
       "         'bog': 3,\n",
       "         'banana plantation': 3,\n",
       "         'wood': 2,\n",
       "         'biofilm': 2,\n",
       "         'sand': 2,\n",
       "         'temperate woodland biome': 2,\n",
       "         'temperate woodland': 2,\n",
       "         'mangrove': 1,\n",
       "         'lagoon': 1,\n",
       "         'microbial mat': 1,\n",
       "         'stromatolite mat': 1,\n",
       "         'flooded savanna': 1,\n",
       "         'marine subtidal rocky reef': 1,\n",
       "         'neritic bathypelagic zone biome': 1,\n",
       "         'temperate shrubland biome': 1})"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 41
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:20.863808Z",
     "start_time": "2025-02-26T02:56:10.531940Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# use the OAK annotator to match the submitted env triad values to ENVO terms\n",
    "# Don't panic about red error messages\n",
    "\n",
    "# List to store all annotations\n",
    "ncbi_env_triad_terms_annotations_list = []\n",
    "\n",
    "for content_value, count in ncbi_env_content_counter.items():\n",
    "    try:\n",
    "        annotations = envo_adapter.annotate_text(content_value)\n",
    "        for annotation in annotations:\n",
    "            # Fetch the label for the object_id\n",
    "            object_label = envo_adapter.label(annotation.object_id) if annotation.object_id else \"Unknown\"\n",
    "\n",
    "            annotation_dict = {\n",
    "                \"content_value\": content_value,\n",
    "                \"count\": count,  # Frequency of this content value in the dataset\n",
    "                \"predicate_id\": annotation.predicate_id,\n",
    "                \"object_id\": annotation.object_id,\n",
    "                \"object_label\": object_label,  # Looked up label\n",
    "                \"subject_start\": annotation.subject_start,\n",
    "                \"subject_end\": annotation.subject_end,\n",
    "                \"match_string\": annotation.match_string,\n",
    "                \"matches_whole_text\": annotation.matches_whole_text,\n",
    "            }\n",
    "            ncbi_env_triad_terms_annotations_list.append(annotation_dict)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing '{content_value}': {e}\")\n"
   ],
   "id": "bd14a4de53fc0fb7",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=None,value=Carbonate which is formed as the result of some biological process.,datatype=None,language=None,); ValueError: Carbonate which is formed as the result of some biological process. is not a valid URI or CURIE\n",
      "WARNING:root:Skipping <http://geneontology.org/foo/applies-pattern> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://schema.org/image> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://swrl.stanford.edu/ontologies/3.3/swrla.owl#isRuleEnabled> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2306597> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q525> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q715269> as it is not a valid CURIE\n"
     ]
    }
   ],
   "execution_count": 42
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:20.885836Z",
     "start_time": "2025-02-26T02:56:20.882931Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for annotation in ncbi_env_triad_terms_annotations_list:\n",
    "    annotation[\"match_string_len\"] = len(annotation[\"match_string\"]) if annotation[\"match_string\"] else 0"
   ],
   "id": "a5d82f9c95dd4515",
   "outputs": [],
   "execution_count": 43
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:20.947936Z",
     "start_time": "2025-02-26T02:56:20.944590Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Pretty-print some sample results\n",
    "for annotation in ncbi_env_triad_terms_annotations_list[:3]:  # Show first 10 annotations\n",
    "    pprint.pprint(annotation)"
   ],
   "id": "c31aee26906c9a8e",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'content_value': 'urban biome',\n",
      " 'count': 285,\n",
      " 'match_string': 'biome',\n",
      " 'match_string_len': 5,\n",
      " 'matches_whole_text': False,\n",
      " 'object_id': 'ENVO:00000428',\n",
      " 'object_label': 'biome',\n",
      " 'predicate_id': 'rdfs:label',\n",
      " 'subject_end': 11,\n",
      " 'subject_start': 7}\n",
      "{'content_value': 'urban biome',\n",
      " 'count': 285,\n",
      " 'match_string': 'urban biome',\n",
      " 'match_string_len': 11,\n",
      " 'matches_whole_text': True,\n",
      " 'object_id': 'ENVO:01000249',\n",
      " 'object_label': 'urban biome',\n",
      " 'predicate_id': 'rdfs:label',\n",
      " 'subject_end': 11,\n",
      " 'subject_start': 1}\n",
      "{'content_value': 'anaerobic bioreactor',\n",
      " 'count': 9,\n",
      " 'match_string': 'bioreactor',\n",
      " 'match_string_len': 10,\n",
      " 'matches_whole_text': False,\n",
      " 'object_id': 'ENVO:00002123',\n",
      " 'object_label': 'bioreactor',\n",
      " 'predicate_id': 'rdfs:label',\n",
      " 'subject_end': 20,\n",
      " 'subject_start': 11}\n"
     ]
    }
   ],
   "execution_count": 44
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.005268Z",
     "start_time": "2025-02-26T02:56:20.999343Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Lists for categorization\n",
    "perfect_match_list = []\n",
    "needs_review_list = []\n",
    "\n",
    "# Identify content_values that have at least one perfect match with predicate_id='rdfs:label'\n",
    "all_annotated_content_values = set()  # Track all content_values that were annotated\n",
    "non_perfect_match_content_values = set()  # Track all content_values that didnt make it into perfect matches\n",
    "perfect_match_content_values = set()\n",
    "unmatched_content_values = set()  # Store content_values that are neither in perfect_match_list nor needs_review_list\n",
    "\n",
    "for annotation in ncbi_env_triad_terms_annotations_list:\n",
    "    content_value = annotation[\"content_value\"]\n",
    "    all_annotated_content_values.add(content_value)\n",
    "\n",
    "    if annotation[\"matches_whole_text\"] and annotation[\"predicate_id\"] == \"rdfs:label\":\n",
    "        perfect_match_list.append(annotation)\n",
    "        perfect_match_content_values.add(content_value)\n",
    "\n",
    "# Identify content_values that have no perfect match and meet the length cutoff\n",
    "needs_review_content_values = set()\n",
    "\n",
    "for annotation in ncbi_env_triad_terms_annotations_list:\n",
    "    content_value = annotation[\"content_value\"]\n",
    "    annotation[\"match_string_len\"] = len(annotation[\"match_string\"]) if annotation[\"match_string\"] else 0\n",
    "\n",
    "    if (\n",
    "            content_value not in perfect_match_content_values and\n",
    "            annotation[\"match_string_len\"] >= MATCH_LENGTH_CUTOFF\n",
    "    ):\n",
    "        needs_review_list.append(annotation)\n",
    "        needs_review_content_values.add(content_value)\n",
    "\n",
    "# Identify content_values that are in env_content_counter but not in perfect_match_list or needs_review_list\n",
    "for content_value in ncbi_env_content_counter.keys():\n",
    "    if content_value not in perfect_match_content_values and content_value not in needs_review_content_values:\n",
    "        unmatched_content_values.add(content_value)\n"
   ],
   "id": "808fae7789da5e90",
   "outputs": [],
   "execution_count": 45
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.037715Z",
     "start_time": "2025-02-26T02:56:21.029003Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Helper function: Normalize text (lowercase + remove extra spaces)\n",
    "def normalize_text(text):\n",
    "    return re.sub(r'\\s+', ' ', text).strip().lower()  # Replace multiple spaces & trim\n",
    "\n",
    "\n",
    "# Create a lookup set of normalized obsolete labels\n",
    "obsolete_label_set = {normalize_text(entry[\"label\"]) for entry in obsoletes_curies_labelled}\n",
    "\n",
    "# Prepare result storage\n",
    "exact_obsolete_matches = []\n",
    "\n",
    "# Check if \"obsolete \" + content_value exists in the obsolete labels\n",
    "for content_value in needs_review_content_values | unmatched_content_values:  # Union of both sets\n",
    "    obsolete_label_candidate = normalize_text(f\"obsolete {content_value}\")\n",
    "    if obsolete_label_candidate in obsolete_label_set:\n",
    "        # Find the corresponding CURIE using normalized comparison\n",
    "        matching_entry = next(entry for entry in obsoletes_curies_labelled\n",
    "                              if normalize_text(entry[\"label\"]) == obsolete_label_candidate)\n",
    "        exact_obsolete_matches.append({\n",
    "            \"content_value\": content_value,  # ORIGINAL content_value\n",
    "            \"obsolete_label\": matching_entry[\"label\"],  # ORIGINAL obsolete label\n",
    "            \"obsolete_curie\": matching_entry[\"curie\"]\n",
    "        })\n"
   ],
   "id": "318998e811fbd4e8",
   "outputs": [],
   "execution_count": 46
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.064136Z",
     "start_time": "2025-02-26T02:56:21.060902Z"
    }
   },
   "cell_type": "code",
   "source": "pprint.pprint(exact_obsolete_matches)",
   "id": "1e5db98f15c648b8",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'content_value': 'animal-associated habitat',\n",
      "  'obsolete_curie': 'ENVO:00006776',\n",
      "  'obsolete_label': 'obsolete animal-associated habitat'},\n",
      " {'content_value': 'ocean water',\n",
      "  'obsolete_curie': 'ENVO:00002151',\n",
      "  'obsolete_label': 'obsolete ocean water'},\n",
      " {'content_value': 'insecta-associated habitat',\n",
      "  'obsolete_curie': 'ENVO:00009004',\n",
      "  'obsolete_label': 'obsolete insecta-associated habitat'},\n",
      " {'content_value': 'marine benthic feature',\n",
      "  'obsolete_curie': 'ENVO:01000105',\n",
      "  'obsolete_label': 'obsolete marine benthic feature'},\n",
      " {'content_value': 'montane grasslands and shrubland biome',\n",
      "  'obsolete_curie': 'ENVO:00000882',\n",
      "  'obsolete_label': 'obsolete Montane grasslands and shrubland biome'}]\n"
     ]
    }
   ],
   "execution_count": 47
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.157812Z",
     "start_time": "2025-02-26T02:56:21.089518Z"
    }
   },
   "cell_type": "code",
   "source": [
    "obsolete_curies = {entry[\"obsolete_curie\"] for entry in\n",
    "                   exact_obsolete_matches}  # obsolete values that were used (indirectly)\n",
    "# obsolete_curies = {entry[\"curie\"] for entry in obsoletes_curies_labelled} # all obsolete values in EnvO\n",
    "\n",
    "obsoletes_metadata_list = []\n",
    "\n",
    "for curie in obsolete_curies:\n",
    "    entity_metadata = envo_adapter.entity_metadata_map(curie)  # Fetch metadata\n",
    "    if entity_metadata:\n",
    "        obsoletes_metadata_list.append(entity_metadata)\n"
   ],
   "id": "ce0327da0b7ca6e6",
   "outputs": [],
   "execution_count": 48
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.185768Z",
     "start_time": "2025-02-26T02:56:21.181642Z"
    }
   },
   "cell_type": "code",
   "source": "pprint.pprint(obsoletes_metadata_list[:3])  # Show first 3 obsolete entities",
   "id": "d8241b3cd3be40c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'IAO:0100001': ['ENVO:00002149'],\n",
      "  'id': ['ENVO:00002151'],\n",
      "  'oio:hasDbXref': ['https://en.wikipedia.org/wiki/Ocean_water'],\n",
      "  'oio:hasOBONamespace': ['ENVO'],\n",
      "  'oio:id': ['ENVO:00002151'],\n",
      "  'owl:deprecated': [True],\n",
      "  'rdfs:isDefinedBy': ['http://purl.obolibrary.org/obo/envo.owl'],\n",
      "  'rdfs:label': ['obsolete ocean water'],\n",
      "  'schema:url': ['http://purl.obolibrary.org/obo/ENVO_00002151'],\n",
      "  'sh:prefix': ['ENVO']},\n",
      " {'id': ['ENVO:00000882'],\n",
      "  'oio:consider': ['ENVO:01000194', 'ENVO:01000216'],\n",
      "  'oio:hasOBONamespace': ['ENVO'],\n",
      "  'oio:id': ['ENVO:00000882'],\n",
      "  'owl:deprecated': [True],\n",
      "  'rdfs:isDefinedBy': ['http://purl.obolibrary.org/obo/envo.owl'],\n",
      "  'rdfs:label': ['obsolete Montane grasslands and shrubland biome'],\n",
      "  'schema:url': ['http://purl.obolibrary.org/obo/ENVO_00000882'],\n",
      "  'sh:prefix': ['ENVO']},\n",
      " {'IAO:0000115': ['A habitat that is in or on a living insect. Here \"insect\" '\n",
      "                  'denotes an individual of a species that is a sub-taxon of '\n",
      "                  'NCBITaxon:50557.'],\n",
      "  'id': ['ENVO:00009004'],\n",
      "  'oio:hasOBONamespace': ['ENVO'],\n",
      "  'oio:id': ['ENVO:00009004'],\n",
      "  'owl:deprecated': [True],\n",
      "  'rdfs:isDefinedBy': ['http://purl.obolibrary.org/obo/envo.owl'],\n",
      "  'rdfs:label': ['obsolete insecta-associated habitat'],\n",
      "  'schema:url': ['http://purl.obolibrary.org/obo/ENVO_00009004'],\n",
      "  'sh:prefix': ['ENVO']}]\n"
     ]
    }
   ],
   "execution_count": 49
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.213982Z",
     "start_time": "2025-02-26T02:56:21.209863Z"
    }
   },
   "cell_type": "code",
   "source": [
    "obsoletes_predicate_counter = Counter()\n",
    "\n",
    "for metadata in obsoletes_metadata_list:\n",
    "    if isinstance(metadata, dict):  # Ensure it's a dictionary\n",
    "        obsoletes_predicate_counter.update(metadata.keys())\n"
   ],
   "id": "298ad4944669f71f",
   "outputs": [],
   "execution_count": 50
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.241139Z",
     "start_time": "2025-02-26T02:56:21.237316Z"
    }
   },
   "cell_type": "code",
   "source": "obsoletes_predicate_counter",
   "id": "7515ae26443554a2",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'id': 5,\n",
       "         'owl:deprecated': 5,\n",
       "         'rdfs:label': 5,\n",
       "         'sh:prefix': 5,\n",
       "         'schema:url': 5,\n",
       "         'rdfs:isDefinedBy': 5,\n",
       "         'oio:hasOBONamespace': 4,\n",
       "         'oio:id': 4,\n",
       "         'IAO:0000115': 3,\n",
       "         'oio:consider': 2,\n",
       "         'IAO:0100001': 1,\n",
       "         'oio:hasDbXref': 1,\n",
       "         'oio:hasExactSynonym': 1,\n",
       "         'oio:inSubset': 1,\n",
       "         'oio:created_by': 1,\n",
       "         'oio:creation_date': 1,\n",
       "         'rdfs:comment': 1})"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 51
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.270389Z",
     "start_time": "2025-02-26T02:56:21.264926Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Compute high-level summary statistics\n",
    "summary = {\n",
    "    \"total_unique_terms\": len(ncbi_env_content_counter),  # Unique terms submitted\n",
    "    \"perfect_matches\": len(perfect_match_content_values),  # Unique perfect matches\n",
    "    \"matches_need_review\": len(needs_review_content_values),  # Unique terms needing review\n",
    "    \"no_valid_oak_match\": len(unmatched_content_values),  # Terms that didn't get a valid OAK annotation\n",
    "    \"exact_obsolete_matches\": len(exact_obsolete_matches),  # Unique terms that exactly matched an obsolete label\n",
    "    \"obsolete_terms_with_replacement\": sum(\n",
    "        1 for m in obsoletes_metadata_list if \"IAO:0100001\" in m\n",
    "    ),  # Obsolete terms with a clear replacement\n",
    "    \"obsolete_terms_with_consider\": sum(\n",
    "        1 for m in obsoletes_metadata_list if \"oio:consider\" in m\n",
    "    ),  # Obsolete terms with \"consider\" alternatives\n",
    "}\n",
    "\n",
    "# Consistency check\n",
    "if summary[\"perfect_matches\"] + summary[\"matches_need_review\"] + summary[\"no_valid_oak_match\"] != summary[\n",
    "    \"total_unique_terms\"]:\n",
    "    print(\n",
    "        f\"WARNING: perfect_matches ({summary['perfect_matches']}) + matches_need_review ({summary['matches_need_review']}) \"\n",
    "        f\"+ no_valid_oak_match ({summary['no_valid_oak_match']})  total_unique_terms ({summary['total_unique_terms']})\"\n",
    "    )\n",
    "\n",
    "# Save to JSON\n",
    "\n",
    "with open(summary_json_file, \"w\", encoding=\"utf-8\") as json_file:\n",
    "    json.dump(summary, json_file, indent=4)\n",
    "\n",
    "print(f\"Summary saved to {summary_json_file}\")\n"
   ],
   "id": "a2939481744c47d6",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary saved to outputs/emp500_summary.json\n"
     ]
    }
   ],
   "execution_count": 52
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.306501Z",
     "start_time": "2025-02-26T02:56:21.294406Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Initialize an empty list to store term legitimacy data\n",
    "term_legitimacy_data = []\n",
    "\n",
    "for content_value, emp500_usages in ncbi_env_content_counter.items():\n",
    "    # Determine perfect match status\n",
    "    is_perfect_match = content_value in perfect_match_content_values\n",
    "\n",
    "    # Get matched_label and matched_curie from perfect_match_list\n",
    "    perfect_match_entry = next(\n",
    "        (entry for entry in perfect_match_list if entry[\"content_value\"] == content_value),\n",
    "        None\n",
    "    )\n",
    "    matched_label = perfect_match_entry[\"object_label\"] if perfect_match_entry else \"\"\n",
    "    matched_curie = perfect_match_entry[\"object_id\"] if perfect_match_entry else \"\"\n",
    "\n",
    "    # Determine obsolete status\n",
    "    obsolete_entry = next((e for e in exact_obsolete_matches if e[\"content_value\"] == content_value), None)\n",
    "    is_obsolete = bool(obsolete_entry)\n",
    "    obsolete_label = obsolete_entry[\"obsolete_label\"] if obsolete_entry else \"\"\n",
    "    obsolete_curie = \"\"\n",
    "\n",
    "    # Get obsolete CURIE from obsoletes_curies_labelled\n",
    "    if obsolete_entry:\n",
    "        obsolete_curie_entry = next(\n",
    "            (o for o in obsoletes_curies_labelled if obsolete_label in o[\"label\"]),\n",
    "            None\n",
    "        )\n",
    "        obsolete_curie = obsolete_curie_entry[\"curie\"] if obsolete_curie_entry else \"\"\n",
    "\n",
    "    # Determine replacement term (IAO:0100001)\n",
    "    replacement_curie = \"\"\n",
    "    replacement_label = \"\"\n",
    "    consider_replacements = \"\"\n",
    "\n",
    "    if obsolete_curie:\n",
    "        metadata = next((m for m in obsoletes_metadata_list if \"id\" in m and obsolete_curie in m[\"id\"]), None)\n",
    "\n",
    "        # Process replacement_curie\n",
    "        if metadata and \"IAO:0100001\" in metadata:\n",
    "            replacement_curie = metadata[\"IAO:0100001\"][0]  # Extract replacement CURIE\n",
    "            replacement_label = envo_adapter.label(replacement_curie) if replacement_curie else \"\"\n",
    "\n",
    "        # Process oio:consider replacements\n",
    "        if metadata and \"oio:consider\" in metadata:\n",
    "            consider_list = metadata[\"oio:consider\"]\n",
    "            consider_replacements = \" | \".join(\n",
    "                f\"{curie}/{envo_adapter.label(curie)}\" for curie in consider_list\n",
    "            )\n",
    "\n",
    "    # Append data to the list\n",
    "    term_legitimacy_data.append({\n",
    "        \"content_value\": content_value,\n",
    "        \"emp500_usages\": emp500_usages,\n",
    "        \"perfect_match\": is_perfect_match,\n",
    "        \"matched_label\": matched_label,\n",
    "        \"matched_curie\": matched_curie,\n",
    "        \"is_obsolete\": is_obsolete,\n",
    "        \"obsolete_label\": obsolete_label,\n",
    "        \"obsolete_curie\": obsolete_curie,\n",
    "        \"replacement_curie\": replacement_curie,\n",
    "        \"replacement_label\": replacement_label,\n",
    "        \"consider_replacements\": consider_replacements\n",
    "    })\n",
    "\n",
    "\n",
    "\n",
    "# Write collected data to TSV using csv.DictWriter\n",
    "with open(detailed_tsv, mode=\"w\", newline=\"\", encoding=\"utf-8\") as file:\n",
    "    fieldnames = [\n",
    "        \"content_value\", \"emp500_usages\", \"perfect_match\", \"matched_label\", \"matched_curie\",\n",
    "        \"is_obsolete\", \"obsolete_label\", \"obsolete_curie\", \"replacement_curie\", \"replacement_label\",\n",
    "        \"consider_replacements\"\n",
    "    ]\n",
    "    writer = csv.DictWriter(file, delimiter=\"\\t\", fieldnames=fieldnames)\n",
    "\n",
    "    # Write header\n",
    "    writer.writeheader()\n",
    "\n",
    "    # Write data rows\n",
    "    writer.writerows(term_legitimacy_data)\n"
   ],
   "id": "e810413ac26d2f64",
   "outputs": [],
   "execution_count": 53
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.338006Z",
     "start_time": "2025-02-26T02:56:21.329976Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Extract `content_value` entries from term_legitimacy_data where `is_obsolete` is True\n",
    "obsolete_content_values = {\n",
    "    entry[\"content_value\"] for entry in term_legitimacy_data if entry[\"is_obsolete\"]\n",
    "}\n",
    "\n",
    "# Create a unified list to hold all formatted entries for writing\n",
    "needs_review_data = []\n",
    "\n",
    "# Process `needs_review_list` while filtering out obsolete content values\n",
    "for entry in needs_review_list:\n",
    "    if entry[\"content_value\"] not in obsolete_content_values:\n",
    "        formatted_entry = {\n",
    "            \"content_value\": entry[\"content_value\"],\n",
    "            \"emp500_usages\": entry.pop(\"count\", 0),  # Rename \"count\" to \"emp500_usages\"\n",
    "            \"predicate_id\": entry.get(\"predicate_id\", \"\"),\n",
    "            \"object_id\": entry.get(\"object_id\", \"\"),\n",
    "            \"object_label\": entry.get(\"object_label\", \"\"),\n",
    "            \"subject_start\": entry.get(\"subject_start\", \"\"),\n",
    "            \"subject_end\": entry.get(\"subject_end\", \"\"),\n",
    "            \"match_string\": entry.get(\"match_string\", \"\"),\n",
    "            \"match_string_len\": entry.get(\"match_string_len\", \"\"),\n",
    "            \"matches_whole_text\": entry.get(\"matches_whole_text\", \"\")\n",
    "        }\n",
    "        needs_review_data.append(formatted_entry)\n",
    "\n",
    "# Process unmatched content values (excluding obsolete terms)\n",
    "filtered_unmatched_content_values = unmatched_content_values - obsolete_content_values\n",
    "\n",
    "for content_value in filtered_unmatched_content_values:\n",
    "    formatted_entry = {\n",
    "        \"content_value\": content_value,\n",
    "        \"emp500_usages\": ncbi_env_content_counter.get(content_value, 0),\n",
    "        \"predicate_id\": \"\",\n",
    "        \"object_id\": \"\",\n",
    "        \"object_label\": \"\",\n",
    "        \"subject_start\": \"\",\n",
    "        \"subject_end\": \"\",\n",
    "        \"match_string\": \"\",\n",
    "        \"match_string_len\": \"\",\n",
    "        \"matches_whole_text\": \"\"\n",
    "    }\n",
    "    needs_review_data.append(formatted_entry)\n",
    "\n",
    "# Define fieldnames based on actual expected columns\n",
    "fieldnames = [\n",
    "    \"content_value\", \"emp500_usages\", \"predicate_id\", \"object_id\", \"object_label\",\n",
    "    \"subject_start\", \"subject_end\", \"match_string\", \"match_string_len\", \"matches_whole_text\"\n",
    "]\n",
    "\n",
    "# Write collected data to TSV using csv.DictWriter\n",
    "with open(needs_review_tsv, mode=\"w\", newline=\"\", encoding=\"utf-8\") as file:\n",
    "    writer = csv.DictWriter(file, delimiter=\"\\t\", fieldnames=fieldnames)\n",
    "    writer.writeheader()\n",
    "    writer.writerows(needs_review_data)\n",
    "\n",
    "print(f\"Needs Review TSV saved to {needs_review_tsv} (Filtered & ensures obsolete terms do NOT reappear)\")\n"
   ],
   "id": "b8da2145cb9feaf",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Needs Review TSV saved to outputs/emp500_env_context_vocabulary_for_review.tsv (Filtered & ensures obsolete terms do NOT reappear)\n"
     ]
    }
   ],
   "execution_count": 54
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.366194Z",
     "start_time": "2025-02-26T02:56:21.362680Z"
    }
   },
   "cell_type": "code",
   "source": [
    "per_slot_value_curations_list = []\n",
    "with open(per_slot_value_curations_tsv, 'r', encoding='utf-8') as file:\n",
    "    reader = csv.DictReader(file, delimiter='\\t')\n",
    "    for row in reader:\n",
    "      per_slot_value_curations_list.append(row)"
   ],
   "id": "9dcfa1d393eee81d",
   "outputs": [],
   "execution_count": 55
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.393127Z",
     "start_time": "2025-02-26T02:56:21.390191Z"
    }
   },
   "cell_type": "code",
   "source": [
    "per_slot_value_curations_dict = {}\n",
    "for row in per_slot_value_curations_list:\n",
    "    key = (row['problematic_key'], row['problematic_value'])\n",
    "    per_slot_value_curations_dict[key] = row\n"
   ],
   "id": "ca3c6375e535b5be",
   "outputs": [],
   "execution_count": 56
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.424800Z",
     "start_time": "2025-02-26T02:56:21.416940Z"
    }
   },
   "cell_type": "code",
   "source": "pprint.pprint(per_slot_value_curations_dict)",
   "id": "6e7c059b6bc43da0",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{('env_broad_scale', 'Small lake biome'): {'count': '20',\n",
      "                                           'done': 'true',\n",
      "                                           'notes': '',\n",
      "                                           'problematic_key': 'env_broad_scale',\n",
      "                                           'problematic_value': 'Small lake '\n",
      "                                                                'biome',\n",
      "                                           'replacement_curie': 'ENVO:00000892',\n",
      "                                           'replacement_label': 'small '\n",
      "                                                                'freshwater '\n",
      "                                                                'lake biome'},\n",
      " ('env_broad_scale', 'flooded savanna'): {'count': '1',\n",
      "                                          'done': 'true',\n",
      "                                          'notes': 'SAMEA7723866 Seasonal '\n",
      "                                                   'succession influences the '\n",
      "                                                   'functioning of a '\n",
      "                                                   'host-associated microbial '\n",
      "                                                   'food web (pitcher plant '\n",
      "                                                   'Darlingtonia california)',\n",
      "                                          'problematic_key': 'env_broad_scale',\n",
      "                                          'problematic_value': 'flooded '\n",
      "                                                               'savanna',\n",
      "                                          'replacement_curie': 'ENVO:01000190',\n",
      "                                          'replacement_label': 'flooded '\n",
      "                                                               'savanna biome'},\n",
      " ('env_broad_scale', 'montane grasslands and shrubland biome'): {'count': '12',\n",
      "                                                                 'done': 'true',\n",
      "                                                                 'notes': 'SAMEA7724107 '\n",
      "                                                                          '- '\n",
      "                                                                          'SAMEA7724118. '\n",
      "                                                                          'MacRae-Crerar',\n",
      "                                                                 'problematic_key': 'env_broad_scale',\n",
      "                                                                 'problematic_value': 'montane '\n",
      "                                                                                      'grasslands '\n",
      "                                                                                      'and '\n",
      "                                                                                      'shrubland '\n",
      "                                                                                      'biome',\n",
      "                                                                 'replacement_curie': '',\n",
      "                                                                 'replacement_label': ''},\n",
      " ('env_broad_scale', 'neritic bathypelagic zone biome'): {'count': '1',\n",
      "                                                          'done': 'true',\n",
      "                                                          'notes': 'SAMEA7723671 '\n",
      "                                                                   ' '\n",
      "                                                                   'SAMEA7723943, '\n",
      "                                                                   'esp. '\n",
      "                                                                   'SAMEA7723939. '\n",
      "                                                                   'Palenik '\n",
      "                                                                   'San Diego '\n",
      "                                                                   'seawater: '\n",
      "                                                                   'Neritic '\n",
      "                                                                   'names a '\n",
      "                                                                   'superficial '\n",
      "                                                                   'horizontal '\n",
      "                                                                   'region, '\n",
      "                                                                   'Bathypelagic '\n",
      "                                                                   'names a '\n",
      "                                                                   'deep '\n",
      "                                                                   'vertical '\n",
      "                                                                   'zone. '\n",
      "                                                                   'reported '\n",
      "                                                                   'depth: '\n",
      "                                                                   '\"3000\". '\n",
      "                                                                   'GEBCO '\n",
      "                                                                   'confirms '\n",
      "                                                                   'ocean '\n",
      "                                                                   'floor is ~ '\n",
      "                                                                   '3800 m '\n",
      "                                                                   'deep '\n",
      "                                                                   'around '\n",
      "                                                                   '33.35 N, '\n",
      "                                                                   '122.12 W '\n",
      "                                                                   '-> '\n",
      "                                                                   'bathypelagic',\n",
      "                                                          'problematic_key': 'env_broad_scale',\n",
      "                                                          'problematic_value': 'neritic '\n",
      "                                                                               'bathypelagic '\n",
      "                                                                               'zone '\n",
      "                                                                               'biome',\n",
      "                                                          'replacement_curie': 'ENVO:01000037',\n",
      "                                                          'replacement_label': 'oceanic '\n",
      "                                                                               'bathypelagic '\n",
      "                                                                               'zone '\n",
      "                                                                               'biome'},\n",
      " ('env_broad_scale', 'temperate shrubland'): {'count': '6',\n",
      "                                              'done': 'true',\n",
      "                                              'notes': '',\n",
      "                                              'problematic_key': 'env_broad_scale',\n",
      "                                              'problematic_value': 'temperate '\n",
      "                                                                   'shrubland',\n",
      "                                              'replacement_curie': 'ENVO:01000215',\n",
      "                                              'replacement_label': 'temperate '\n",
      "                                                                   'shrubland '\n",
      "                                                                   'biome'},\n",
      " ('env_local_scale', 'animal-associated habitat'): {'count': '84',\n",
      "                                                    'done': 'true',\n",
      "                                                    'notes': '',\n",
      "                                                    'problematic_key': 'env_local_scale',\n",
      "                                                    'problematic_value': 'animal-associated '\n",
      "                                                                         'habitat',\n",
      "                                                    'replacement_curie': 'ENVO:01001002',\n",
      "                                                    'replacement_label': 'animal-associated '\n",
      "                                                                         'environment'},\n",
      " ('env_local_scale', 'bog'): {'count': '3',\n",
      "                              'done': 'false',\n",
      "                              'notes': 'SAMEA7723551. subpolar coniferous '\n",
      "                                       'forest biome/soil. via related synonym',\n",
      "                              'problematic_key': 'env_local_scale',\n",
      "                              'problematic_value': 'bog',\n",
      "                              'replacement_curie': 'ENVO:00000044',\n",
      "                              'replacement_label': 'peatland'},\n",
      " ('env_local_scale', 'cropland'): {'count': '10',\n",
      "                                   'done': 'false',\n",
      "                                   'notes': 'SAMEA7724195 - SAMEA7724228. '\n",
      "                                            'Myrold',\n",
      "                                   'problematic_key': 'env_local_scale',\n",
      "                                   'problematic_value': 'cropland',\n",
      "                                   'replacement_curie': '',\n",
      "                                   'replacement_label': ''},\n",
      " ('env_local_scale', 'forest'): {'count': '21',\n",
      "                                 'done': 'true',\n",
      "                                 'notes': '',\n",
      "                                 'problematic_key': 'env_local_scale',\n",
      "                                 'problematic_value': 'forest',\n",
      "                                 'replacement_curie': 'ENVO:01001243',\n",
      "                                 'replacement_label': 'forest'},\n",
      " ('env_local_scale', 'host-associated habitat'): {'count': '133',\n",
      "                                                  'done': 'true',\n",
      "                                                  'notes': '',\n",
      "                                                  'problematic_key': 'env_local_scale',\n",
      "                                                  'problematic_value': 'host-associated '\n",
      "                                                                       'habitat',\n",
      "                                                  'replacement_curie': 'ENVO:01001002',\n",
      "                                                  'replacement_label': 'animal-associated '\n",
      "                                                                       'environment'},\n",
      " ('env_local_scale', 'insecta-associated habitat'): {'count': '28',\n",
      "                                                     'done': 'true',\n",
      "                                                     'notes': 'SAMEA7723460. '\n",
      "                                                              'tropical '\n",
      "                                                              'broadleaf '\n",
      "                                                              'forest '\n",
      "                                                              'biome/organic '\n",
      "                                                              'material. the '\n",
      "                                                              'fact that it '\n",
      "                                                              'came from an '\n",
      "                                                              'insect should '\n",
      "                                                              'be captured in '\n",
      "                                                              'a host slot',\n",
      "                                                     'problematic_key': 'env_local_scale',\n",
      "                                                     'problematic_value': 'insecta-associated '\n",
      "                                                                          'habitat',\n",
      "                                                     'replacement_curie': 'ENVO:01001002',\n",
      "                                                     'replacement_label': 'animal-associated '\n",
      "                                                                          'environment'},\n",
      " ('env_local_scale', 'intermittent pond'): {'count': '7',\n",
      "                                            'done': 'true',\n",
      "                                            'notes': '',\n",
      "                                            'problematic_key': 'env_local_scale',\n",
      "                                            'problematic_value': 'intermittent '\n",
      "                                                                 'pond',\n",
      "                                            'replacement_curie': 'ENVO:00000504',\n",
      "                                            'replacement_label': 'container of '\n",
      "                                                                 'an '\n",
      "                                                                 'intermittent '\n",
      "                                                                 'pond'},\n",
      " ('env_local_scale', 'mangrove'): {'count': '1',\n",
      "                                   'done': 'true',\n",
      "                                   'notes': 'SAMEA7723419. used in combination '\n",
      "                                            'with marine benthic biome/marine '\n",
      "                                            'sediment. Twin Cayes, Belize '\n",
      "                                            'Barrier Reef System. Most depths '\n",
      "                                            '< 30 m, but some 300 or 700. '\n",
      "                                            'Might not want to call all '\n",
      "                                            'benthic.',\n",
      "                                   'problematic_key': 'env_local_scale',\n",
      "                                   'problematic_value': 'mangrove',\n",
      "                                   'replacement_curie': 'ENVO:00000103',\n",
      "                                   'replacement_label': 'mangrove island'},\n",
      " ('env_local_scale', 'marine benthic feature'): {'count': '125',\n",
      "                                                 'done': 'false',\n",
      "                                                 'notes': '',\n",
      "                                                 'problematic_key': 'env_local_scale',\n",
      "                                                 'problematic_value': 'marine '\n",
      "                                                                      'benthic '\n",
      "                                                                      'feature',\n",
      "                                                 'replacement_curie': '',\n",
      "                                                 'replacement_label': ''},\n",
      " ('env_local_scale', 'montane desert'): {'count': '26',\n",
      "                                         'done': 'false',\n",
      "                                         'notes': 'SAMEA7724257. desert '\n",
      "                                                  'biome/alpine soil. ... mine '\n",
      "                                                  'other biosample (lat lon?), '\n",
      "                                                  'bioproject or study slots, '\n",
      "                                                  'or journal article',\n",
      "                                         'problematic_key': 'env_local_scale',\n",
      "                                         'problematic_value': 'montane desert',\n",
      "                                         'replacement_curie': 'ENVO:01001903',\n",
      "                                         'replacement_label': 'montane desert '\n",
      "                                                              'ecosystem'},\n",
      " ('env_local_scale', 'oil reservior'): {'count': '14',\n",
      "                                        'done': 'true',\n",
      "                                        'notes': 'SAMEA7723792 - SAMEA7723805 '\n",
      "                                                 'Oil reservoir misspelled in '\n",
      "                                                 'source',\n",
      "                                        'problematic_key': 'env_local_scale',\n",
      "                                        'problematic_value': 'oil reservior',\n",
      "                                        'replacement_curie': 'ENVO:00002185',\n",
      "                                        'replacement_label': 'oil reservoir'},\n",
      " ('env_local_scale', 'rangeland'): {'count': '5',\n",
      "                                    'done': 'false',\n",
      "                                    'notes': 'SAMEA7724195 - SAMEA7724228. '\n",
      "                                             'Myrold',\n",
      "                                    'problematic_key': 'env_local_scale',\n",
      "                                    'problematic_value': 'rangeland',\n",
      "                                    'replacement_curie': '',\n",
      "                                    'replacement_label': ''},\n",
      " ('env_local_scale', 'stratum'): {'count': '18',\n",
      "                                  'done': 'false',\n",
      "                                  'notes': 'SAMEA7724332: forest floor '\n",
      "                                           '[ENVO:01001582] OR active '\n",
      "                                           'permafrost layer [ENVO:04000009] ?',\n",
      "                                  'problematic_key': 'env_local_scale',\n",
      "                                  'problematic_value': 'stratum',\n",
      "                                  'replacement_curie': '',\n",
      "                                  'replacement_label': ''},\n",
      " ('env_local_scale', 'temperate coniferous forest'): {'count': '16',\n",
      "                                                      'done': 'false',\n",
      "                                                      'notes': 'SAMEA7724195 - '\n",
      "                                                               'SAMEA7724228. '\n",
      "                                                               'Myrold',\n",
      "                                                      'problematic_key': 'env_local_scale',\n",
      "                                                      'problematic_value': 'temperate '\n",
      "                                                                           'coniferous '\n",
      "                                                                           'forest',\n",
      "                                                      'replacement_curie': '',\n",
      "                                                      'replacement_label': ''},\n",
      " ('env_local_scale', 'temperate shrubland'): {'count': '1',\n",
      "                                              'done': 'false',\n",
      "                                              'notes': 'SAMEA7724195 - '\n",
      "                                                       'SAMEA7724228. Myrold',\n",
      "                                              'problematic_key': 'env_local_scale',\n",
      "                                              'problematic_value': 'temperate '\n",
      "                                                                   'shrubland',\n",
      "                                              'replacement_curie': '',\n",
      "                                              'replacement_label': ''},\n",
      " ('env_local_scale', 'temperate woodland'): {'count': '2',\n",
      "                                             'done': 'false',\n",
      "                                             'notes': 'SAMEA7724195 - '\n",
      "                                                      'SAMEA7724228. Myrold',\n",
      "                                             'problematic_key': 'env_local_scale',\n",
      "                                             'problematic_value': 'temperate '\n",
      "                                                                  'woodland',\n",
      "                                             'replacement_curie': '',\n",
      "                                             'replacement_label': ''},\n",
      " ('env_local_scale', 'wetland'): {'count': '9',\n",
      "                                  'done': 'false',\n",
      "                                  'notes': '',\n",
      "                                  'problematic_key': 'env_local_scale',\n",
      "                                  'problematic_value': 'wetland',\n",
      "                                  'replacement_curie': '',\n",
      "                                  'replacement_label': ''},\n",
      " ('env_medium', 'coastal water'): {'count': '9',\n",
      "                                   'done': 'true',\n",
      "                                   'notes': '',\n",
      "                                   'problematic_key': 'env_medium',\n",
      "                                   'problematic_value': 'coastal water',\n",
      "                                   'replacement_curie': 'ENVO:00002006',\n",
      "                                   'replacement_label': 'liquid water'},\n",
      " ('env_medium', 'pitcher plant fluid'): {'count': '6',\n",
      "                                         'done': 'true',\n",
      "                                         'notes': 'SAMEA7723869, SAMEA7723863. '\n",
      "                                                  'ask to have '\n",
      "                                                  'phytotelma/phytotelm or '\n",
      "                                                  'ascidia/ascidium added to '\n",
      "                                                  'EnvO. taxon should go in a '\n",
      "                                                  'different slot. '\n",
      "                                                  'env_local_scale could be '\n",
      "                                                  'leaf [PO:0025034] ?',\n",
      "                                         'problematic_key': 'env_medium',\n",
      "                                         'problematic_value': 'pitcher plant '\n",
      "                                                              'fluid',\n",
      "                                         'replacement_curie': 'ENVO:01000815',\n",
      "                                         'replacement_label': 'liquid '\n",
      "                                                              'environmental '\n",
      "                                                              'material'}}\n"
     ]
    }
   ],
   "execution_count": 57
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.457928Z",
     "start_time": "2025-02-26T02:56:21.450649Z"
    }
   },
   "cell_type": "code",
   "source": "myrold_attributes_with_land_cover_frame = pd.read_csv(myrold_attributes_with_land_cover_tsv, sep=\"\\t\")",
   "id": "1e656ac72f4cd37b",
   "outputs": [],
   "execution_count": 58
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.503836Z",
     "start_time": "2025-02-26T02:56:21.490545Z"
    }
   },
   "cell_type": "code",
   "source": "myrold_attributes_with_land_cover = myrold_attributes_with_land_cover_frame.to_dict(orient='records')",
   "id": "fb0be19d54ef3163",
   "outputs": [],
   "execution_count": 59
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.545556Z",
     "start_time": "2025-02-26T02:56:21.542901Z"
    }
   },
   "cell_type": "code",
   "source": "myrold_attributes_with_land_cover_dict = {i['accession']: i for i in myrold_attributes_with_land_cover }",
   "id": "1b2d433bf6431305",
   "outputs": [],
   "execution_count": 60
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.639651Z",
     "start_time": "2025-02-26T02:56:21.610086Z"
    }
   },
   "cell_type": "code",
   "source": "myrold_attributes_with_land_cover_dict",
   "id": "4cefbbc3c723ae9e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SAMEA7724195': {'accession': 'SAMEA7724195',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724195',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s001',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 719.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 504.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 255.0,\n",
       "  'collection_timestamp': '12/15/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Douglas-fir',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 286,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Tillamook County',\n",
       "  'latitude': 45.046708,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -123.895233,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s001',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 1,\n",
       "  'tier': nan,\n",
       "  'water content': 1.2109,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 43,\n",
       "  'envo_curie': 'ENVO:01000855',\n",
       "  'envo_label': 'area of mixed forest'},\n",
       " 'SAMEA7724196': {'accession': 'SAMEA7724196',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724196',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s002',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 861.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 493.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 252.0,\n",
       "  'collection_timestamp': '12/15/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Douglas-fir',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 311,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Tillamook County',\n",
       "  'latitude': 45.044492,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -123.893006,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s002',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 2,\n",
       "  'tier': nan,\n",
       "  'water content': 0.9845,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 43,\n",
       "  'envo_curie': 'ENVO:01000855',\n",
       "  'envo_label': 'area of mixed forest'},\n",
       " 'SAMEA7724197': {'accession': 'SAMEA7724197',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724197',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s003',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1451.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 476.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 242.0,\n",
       "  'collection_timestamp': '12/15/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Douglas-fir',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 313,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Tillamook County',\n",
       "  'latitude': 45.0437,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -123.894378,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s003',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 3,\n",
       "  'tier': nan,\n",
       "  'water content': 0.7961,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 43,\n",
       "  'envo_curie': 'ENVO:01000855',\n",
       "  'envo_label': 'area of mixed forest'},\n",
       " 'SAMEA7724198': {'accession': 'SAMEA7724198',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724198',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s004',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 868.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 649.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 348.0,\n",
       "  'collection_timestamp': '12/15/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Red alder',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 313,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Tillamook County',\n",
       "  'latitude': 45.046103,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -123.895811,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s004',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 4,\n",
       "  'tier': nan,\n",
       "  'water content': 2.0964,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 43,\n",
       "  'envo_curie': 'ENVO:01000855',\n",
       "  'envo_label': 'area of mixed forest'},\n",
       " 'SAMEA7724199': {'accession': 'SAMEA7724199',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724199',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s005',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 869.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 626.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 347.0,\n",
       "  'collection_timestamp': '12/15/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Red alder',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 305,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Tillamook County',\n",
       "  'latitude': 45.045375,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -123.893619,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s005',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 5,\n",
       "  'tier': nan,\n",
       "  'water content': 1.704,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 43,\n",
       "  'envo_curie': 'ENVO:01000855',\n",
       "  'envo_label': 'area of mixed forest'},\n",
       " 'SAMEA7724200': {'accession': 'SAMEA7724200',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724200',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s006',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 893.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 555.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 308.0,\n",
       "  'collection_timestamp': '12/15/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Red alder',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 322,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Tillamook County',\n",
       "  'latitude': 45.044433,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -123.894908,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s006',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 6,\n",
       "  'tier': nan,\n",
       "  'water content': 1.5223,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 43,\n",
       "  'envo_curie': 'ENVO:01000855',\n",
       "  'envo_label': 'area of mixed forest'},\n",
       " 'SAMEA7724201': {'accession': 'SAMEA7724201',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724201',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s007',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1076.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 459.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 257.0,\n",
       "  'collection_timestamp': '11/11/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Douglas-fir',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 641,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Lane County',\n",
       "  'latitude': 44.230317,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -122.191222,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s007',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 7,\n",
       "  'tier': nan,\n",
       "  'water content': 0.5692,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724202': {'accession': 'SAMEA7724202',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724202',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s008',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1606.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 445.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 224.0,\n",
       "  'collection_timestamp': '11/11/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Douglas-fir',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 662,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Lane County',\n",
       "  'latitude': 44.229294,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -122.191167,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s008',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 8,\n",
       "  'tier': nan,\n",
       "  'water content': 0.7353,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724203': {'accession': 'SAMEA7724203',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724203',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s009',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 935.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 440.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 238.0,\n",
       "  'collection_timestamp': '11/11/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Douglas-fir',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 760,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Lane County',\n",
       "  'latitude': 44.226658,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -122.163778,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s009',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 9,\n",
       "  'tier': nan,\n",
       "  'water content': 0.6717,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724204': {'accession': 'SAMEA7724204',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724204',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s010',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 816.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 523.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 327.0,\n",
       "  'collection_timestamp': '11/11/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Red alder',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 637,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Lane County',\n",
       "  'latitude': 44.230556,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -122.191464,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s010',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 10,\n",
       "  'tier': nan,\n",
       "  'water content': 0.6048,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724205': {'accession': 'SAMEA7724205',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724205',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:23:59Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s011',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 786.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 477.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 259.0,\n",
       "  'collection_timestamp': '11/11/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Red alder',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 665,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Lane County',\n",
       "  'latitude': 44.229297,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -122.191414,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s011',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 11,\n",
       "  'tier': nan,\n",
       "  'water content': 0.6144,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724206': {'accession': 'SAMEA7724206',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724206',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.5.s012',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 656.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 490.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 291.0,\n",
       "  'collection_timestamp': '11/11/15 0:00',\n",
       "  'cur_land_use': 'forest',\n",
       "  'cur_vegetation': 'Red alder',\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 758,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 5,\n",
       "  'emp500_title': 'Tree-associated soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Two sites (low, high soil N) by two tree species (Douglas-fir, red alder) in three replicate blocks per site',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: State of Oregon,Lane County',\n",
       "  'latitude': 44.226456,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': nan,\n",
       "  'longitude': -122.163825,\n",
       "  'physical_specimen_location': 'UCSD Knight Lab Freezer I and David Myrold Lab',\n",
       "  'project name': 'Alder-Fir Study',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.5.s012',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': nan,\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 12,\n",
       "  'tier': nan,\n",
       "  'water content': 0.531,\n",
       "  'water_content_soil_meth': 'gravimetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724207': {'accession': 'SAMEA7724207',\n",
       "  'ena first public': nan,\n",
       "  'ena last update': nan,\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': '2020-12-17T04:08:06Z',\n",
       "  'ena-last-update': '2020-12-16T01:24:00Z',\n",
       "  'external id': 'SAMEA7724207',\n",
       "  'insdc center alias': nan,\n",
       "  'insdc center name': 'UCSDMI',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s001',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 858.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 554.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 340.0,\n",
       "  'collection_timestamp': '4/8/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 177,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Marion County',\n",
       "  'latitude': 45.2505,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Fine, mixed, active, mesic Xeric Palehumults',\n",
       "  'longitude': -123.07038,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s001',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': 'soil metagenome',\n",
       "  'soil type': 'Jory',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 1,\n",
       "  'tier': 1.0,\n",
       "  'water content': 26.26458316,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 81,\n",
       "  'envo_curie': 'ENVO:01000891',\n",
       "  'envo_label': 'area of pastureland or hayfields'},\n",
       " 'SAMEA7724208': {'accession': 'SAMEA7724208',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724208',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s002',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1111.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 303.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 256.0,\n",
       "  'collection_timestamp': '4/10/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 70,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_2_NRCS0056_032_GC',\n",
       "  'geo loc name': 'USA: Oregon,Benton County',\n",
       "  'latitude': 44.62691,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Fine-silty, mixed, superactive, mesic Argiaquic Xeric Argialbolls',\n",
       "  'longitude': -123.21072,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s002',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Amity',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 2,\n",
       "  'tier': 1.0,\n",
       "  'water content': 40.87406679,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 81,\n",
       "  'envo_curie': 'ENVO:01000891',\n",
       "  'envo_label': 'area of pastureland or hayfields'},\n",
       " 'SAMEA7724209': {'accession': 'SAMEA7724209',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724209',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s003',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 821.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 574.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 357.0,\n",
       "  'collection_timestamp': '4/10/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 69,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_3_NRCS0061_040_GC',\n",
       "  'geo loc name': 'USA: Oregon,Benton County',\n",
       "  'latitude': 44.62484,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Fine-silty, mixed, superactive, mesic Pachic Ultic Argixerolls',\n",
       "  'longitude': -123.21777,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s003',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Willamette',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 3,\n",
       "  'tier': 1.0,\n",
       "  'water content': 32.94649763,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 81,\n",
       "  'envo_curie': 'ENVO:01000891',\n",
       "  'envo_label': 'area of pastureland or hayfields'},\n",
       " 'SAMEA7724210': {'accession': 'SAMEA7724210',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724210',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s004',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1043.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 431.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 348.0,\n",
       "  'collection_timestamp': '4/10/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 72,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_4_NRCS0066_041_GC',\n",
       "  'geo loc name': 'USA: Oregon,Benton County',\n",
       "  'latitude': 44.63881,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Fine-silty, mixed, superactive, mesic Aquultic Argixerolls',\n",
       "  'longitude': -123.18977,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s004',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Woodburn',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 4,\n",
       "  'tier': 1.0,\n",
       "  'water content': 41.85246269,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 81,\n",
       "  'envo_curie': 'ENVO:01000891',\n",
       "  'envo_label': 'area of pastureland or hayfields'},\n",
       " 'SAMEA7724211': {'accession': 'SAMEA7724211',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724211',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s005',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 744.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 611.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 394.0,\n",
       "  'collection_timestamp': '4/15/15 0:00',\n",
       "  'cur_land_use': 'rangeland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1171,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'rangeland biome',\n",
       "  'env feature': 'rangeland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_5_NRCS0096_036_GC',\n",
       "  'geo loc name': 'USA: Oregon,Union County',\n",
       "  'latitude': 45.093472,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Clayey-skeletal, smectitic, mesic Lithic Argixerolls',\n",
       "  'longitude': -117.784667,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s005',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Gwinly',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 5,\n",
       "  'tier': 1.0,\n",
       "  'water content': 25.57745123,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 71,\n",
       "  'envo_curie': 'ENVO:01000888',\n",
       "  'envo_label': 'area of gramanoid or herbaceous vegetation'},\n",
       " 'SAMEA7724212': {'accession': 'SAMEA7724212',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724212',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s006',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 831.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 610.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 375.0,\n",
       "  'collection_timestamp': '4/17/15 0:00',\n",
       "  'cur_land_use': 'temperate coniferous forest',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1473,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Klamath County',\n",
       "  'latitude': 43.228761,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Ashy-pumiceous, glassy Xeric Vitricryands',\n",
       "  'longitude': -121.794902,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s006',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Lapine',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 6,\n",
       "  'tier': 1.0,\n",
       "  'water content': 64.03522285,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724213': {'accession': 'SAMEA7724213',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724213',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s007',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 848.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 740.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 392.0,\n",
       "  'collection_timestamp': '4/17/15 0:00',\n",
       "  'cur_land_use': 'rangeland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1379,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'rangeland biome',\n",
       "  'env feature': 'rangeland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Klamath County',\n",
       "  'latitude': 42.881698,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Diatomaceous, euic Limnic Cryosaprists',\n",
       "  'longitude': -121.738759,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s007',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Wocusbay',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 7,\n",
       "  'tier': 1.0,\n",
       "  'water content': 342.2400884,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 95,\n",
       "  'envo_curie': 'ENVO:01000894',\n",
       "  'envo_label': 'area of emergent herbaceous wetland'},\n",
       " 'SAMEA7724214': {'accession': 'SAMEA7724214',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724214',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s008',\n",
       "  'alpha_shotgun_woltka_min10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': nan,\n",
       "  'collection_timestamp': '4/17/15 0:00',\n",
       "  'cur_land_use': 'rangeland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 773,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'rangeland biome',\n",
       "  'env feature': 'rangeland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_8_NRCS0176_035_GC',\n",
       "  'geo loc name': 'USA: Oregon,Sherman County',\n",
       "  'latitude': 45.24199,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Loamy-skeletal, mixed, superactive, mesic Aridic Lithic Haploxerolls',\n",
       "  'longitude': -120.5791,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s008',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Lickskillet',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 8,\n",
       "  'tier': 1.0,\n",
       "  'water content': 14.95127057,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 71,\n",
       "  'envo_curie': 'ENVO:01000888',\n",
       "  'envo_label': 'area of gramanoid or herbaceous vegetation'},\n",
       " 'SAMEA7724215': {'accession': 'SAMEA7724215',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724215',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s009',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 622.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 574.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 335.0,\n",
       "  'collection_timestamp': '4/16/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 790,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_9_NRCS0181_038_GC',\n",
       "  'geo loc name': 'USA: Oregon,Sherman County',\n",
       "  'latitude': 45.16543,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Fine-silty, mixed, superactive, mesic Typic Haploxerolls',\n",
       "  'longitude': -120.6592,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s009',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Condon',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 9,\n",
       "  'tier': 1.0,\n",
       "  'water content': 13.41290111,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 82,\n",
       "  'envo_curie': 'ENVO:01000892',\n",
       "  'envo_label': 'area of cropland'},\n",
       " 'SAMEA7724216': {'accession': 'SAMEA7724216',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724216',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s010',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 823.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 480.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 277.0,\n",
       "  'collection_timestamp': '4/22/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 457,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Umatilla County',\n",
       "  'latitude': 45.71948,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Coarse-silty, mixed, superactive, mesic Typic Haploxerolls',\n",
       "  'longitude': -118.62585,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s010',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Walla Walla',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 10,\n",
       "  'tier': 1.0,\n",
       "  'water content': 11.08752956,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 82,\n",
       "  'envo_curie': 'ENVO:01000892',\n",
       "  'envo_label': 'area of cropland'},\n",
       " 'SAMEA7724217': {'accession': 'SAMEA7724217',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724217',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s011',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1449.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 726.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 399.0,\n",
       "  'collection_timestamp': '4/21/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 6,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_11_NRCS0221_033_GC',\n",
       "  'geo loc name': 'USA: Oregon,Tillamook County',\n",
       "  'latitude': 45.47756,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Fine-silty, mixed, superactive, isomesic Fluventic Humudepts',\n",
       "  'longitude': -123.84166,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s011',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Nehalem',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 11,\n",
       "  'tier': 1.0,\n",
       "  'water content': 71.79258607,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 81,\n",
       "  'envo_curie': 'ENVO:01000891',\n",
       "  'envo_label': 'area of pastureland or hayfields'},\n",
       " 'SAMEA7724218': {'accession': 'SAMEA7724218',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724218',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s012',\n",
       "  'alpha_shotgun_woltka_min10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': nan,\n",
       "  'collection_timestamp': '4/28/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 321,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_12_NRCS0261_039_GC',\n",
       "  'geo loc name': 'USA: Oregon,Umatilla County',\n",
       "  'latitude': 45.65347,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Coarse-silty, mixed, superactive, mesic Calcidic Haploxerolls',\n",
       "  'longitude': -119.46691,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s012',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Ritzville',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 12,\n",
       "  'tier': 1.0,\n",
       "  'water content': 5.965089674,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 82,\n",
       "  'envo_curie': 'ENVO:01000892',\n",
       "  'envo_label': 'area of cropland'},\n",
       " 'SAMEA7724219': {'accession': 'SAMEA7724219',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724219',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s013',\n",
       "  'alpha_shotgun_woltka_min10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': nan,\n",
       "  'collection_timestamp': '5/4/15 0:00',\n",
       "  'cur_land_use': 'rangeland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1355,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'rangeland biome',\n",
       "  'env feature': 'rangeland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Deschutes County',\n",
       "  'latitude': 43.410408,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Fine-loamy, mixed, superactive, frigid Xeric Argidurids',\n",
       "  'longitude': -119.630923,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s013',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Brace',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 13,\n",
       "  'tier': 1.0,\n",
       "  'water content': 10.44359165,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 52,\n",
       "  'envo_curie': 'ENVO:01000869',\n",
       "  'envo_label': 'area of scrub'},\n",
       " 'SAMEA7724220': {'accession': 'SAMEA7724220',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724220',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s014',\n",
       "  'alpha_shotgun_woltka_min10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': nan,\n",
       "  'collection_timestamp': '5/4/15 0:00',\n",
       "  'cur_land_use': 'temperate coniferous forest',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1496,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_14_NRCS0331_043_GC',\n",
       "  'geo loc name': 'USA: Oregon,Grant County',\n",
       "  'latitude': 44.57589,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Loamy-skeletal, isotic, frigid Vitrandic Argixerolls',\n",
       "  'longitude': -118.29807,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s014',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Klicker',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 14,\n",
       "  'tier': 1.0,\n",
       "  'water content': 10.56014099,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 52,\n",
       "  'envo_curie': 'ENVO:01000869',\n",
       "  'envo_label': 'area of scrub'},\n",
       " 'SAMEA7724221': {'accession': 'SAMEA7724221',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724221',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s015',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 1310.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 790.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 453.0,\n",
       "  'collection_timestamp': '5/8/15 0:00',\n",
       "  'cur_land_use': 'temperate woodland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 67,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate woodland biome',\n",
       "  'env feature': 'temperate woodland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Benton County',\n",
       "  'latitude': 44.60695,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Very-fine, smectitic, mesic Xeric Endoaquerts',\n",
       "  'longitude': -123.238343,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s015',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Bashaw',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 15,\n",
       "  'tier': 1.0,\n",
       "  'water content': 79.61128738,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 95,\n",
       "  'envo_curie': 'ENVO:01000894',\n",
       "  'envo_label': 'area of emergent herbaceous wetland'},\n",
       " 'SAMEA7724222': {'accession': 'SAMEA7724222',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724222',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:00Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s016',\n",
       "  'alpha_shotgun_woltka_min10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': nan,\n",
       "  'collection_timestamp': '5/8/15 0:00',\n",
       "  'cur_land_use': 'rangeland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1422,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'rangeland biome',\n",
       "  'env feature': 'rangeland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_16_NRCS0371_042_GC',\n",
       "  'geo loc name': 'USA: Oregon,Crook County',\n",
       "  'latitude': 43.97827,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Clayey-skeletal, smectitic, frigid Lithic Argixerolls',\n",
       "  'longitude': -119.879203,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s016',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Ateron',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 16,\n",
       "  'tier': 1.0,\n",
       "  'water content': 9.582366246,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 52,\n",
       "  'envo_curie': 'ENVO:01000869',\n",
       "  'envo_label': 'area of scrub'},\n",
       " 'SAMEA7724223': {'accession': 'SAMEA7724223',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724223',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:01Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s017',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 601.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 529.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 297.0,\n",
       "  'collection_timestamp': '5/12/15 0:00',\n",
       "  'cur_land_use': 'temperate woodland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 589,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate woodland biome',\n",
       "  'env feature': 'temperate woodland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Jackson County',\n",
       "  'latitude': 42.311278,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Fine-loamy, mixed, superactive, mesic Mollic Haploxeralfs',\n",
       "  'longitude': -122.983444,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s017',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Vannoy',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 17,\n",
       "  'tier': 1.0,\n",
       "  'water content': 21.12107927,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724224': {'accession': 'SAMEA7724224',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724224',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:01Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s018',\n",
       "  'alpha_shotgun_woltka_min10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': nan,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': nan,\n",
       "  'collection_timestamp': '5/13/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 1242,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_18_NRCS0426_034_GC',\n",
       "  'geo loc name': 'USA: Oregon,Harney County',\n",
       "  'latitude': 42.26074,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Fine, smectitic, mesic Sodic Aquicambids',\n",
       "  'longitude': -118.60881,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s018',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Alvodest',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 18,\n",
       "  'tier': 1.0,\n",
       "  'water content': 16.22584686,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 31,\n",
       "  'envo_curie': 'ENVO:01000752',\n",
       "  'envo_label': 'area of barren land'},\n",
       " 'SAMEA7724225': {'accession': 'SAMEA7724225',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724225',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:01Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s019',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 486.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 383.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 211.0,\n",
       "  'collection_timestamp': '5/13/15 0:00',\n",
       "  'cur_land_use': 'temperate coniferous forest',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 460,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Lincoln County',\n",
       "  'latitude': 44.52175,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Fine-loamy, isotic, mesic Andic Humudepts',\n",
       "  'longitude': -123.7665,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s019',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Bohannon',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 19,\n",
       "  'tier': 1.0,\n",
       "  'water content': 65.25774777,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724226': {'accession': 'SAMEA7724226',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724226',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:01Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s020',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 815.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 686.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 432.0,\n",
       "  'collection_timestamp': '5/15/15 0:00',\n",
       "  'cur_land_use': 'cropland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 104,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'cropland biome',\n",
       "  'env feature': 'cropland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': 'EMP_Myr_22_NRCS0526_037_GC',\n",
       "  'geo loc name': 'USA: Oregon,Morrow County',\n",
       "  'latitude': 45.87244,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Mixed, mesic Xeric Torripsamments',\n",
       "  'longitude': -119.57654,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s020',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Quincy',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 20,\n",
       "  'tier': 1.0,\n",
       "  'water content': 14.34269186,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 82,\n",
       "  'envo_curie': 'ENVO:01000892',\n",
       "  'envo_label': 'area of cropland'},\n",
       " 'SAMEA7724227': {'accession': 'SAMEA7724227',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724227',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:01Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s021',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 603.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 599.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 365.0,\n",
       "  'collection_timestamp': '5/18/15 0:00',\n",
       "  'cur_land_use': 'temperate coniferous forest',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 46,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate coniferous forest biome',\n",
       "  'env feature': 'temperate coniferous forest',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Coos County',\n",
       "  'latitude': 43.2122,\n",
       "  'lcms_batch': nan,\n",
       "  'lcms_extraction_protocol': nan,\n",
       "  'local_class': 'Coarse-loamy, isotic, isomesic, ortstein Typic Haplorthods',\n",
       "  'longitude': -124.35476,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s021',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Bullards',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 21,\n",
       "  'tier': 1.0,\n",
       "  'water content': 17.57797233,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 42,\n",
       "  'envo_curie': 'ENVO:01000843',\n",
       "  'envo_label': 'area of evergreen forest'},\n",
       " 'SAMEA7724228': {'accession': 'SAMEA7724228',\n",
       "  'ena first public': '2020-12-17',\n",
       "  'ena last update': '2020-12-16',\n",
       "  'ena-checklist': 'ERC000011',\n",
       "  'ena-first-public': nan,\n",
       "  'ena-last-update': nan,\n",
       "  'external id': 'SAMEA7724228',\n",
       "  'insdc center alias': 'UCSDMI',\n",
       "  'insdc center name': 'University of California San Diego Microbiome Initiative',\n",
       "  'insdc first public': '2020-12-17T04:08:06Z',\n",
       "  'insdc last update': '2020-12-16T01:24:01Z',\n",
       "  'insdc status': 'public',\n",
       "  'submitter id': 'qiita_sid_13114:13114.myrold.59.s022',\n",
       "  'alpha_shotgun_woltka_min10k_richness': 631.0,\n",
       "  'alpha_shotgun_woltka_rar10k_richness': 574.0,\n",
       "  'alpha_shotgun_woltka_rar3450_richness': 358.0,\n",
       "  'collection_timestamp': '6/3/15 0:00',\n",
       "  'cur_land_use': 'temperate shrubland',\n",
       "  'cur_vegetation': nan,\n",
       "  'depth_sample': 0.15,\n",
       "  'elevation': 548,\n",
       "  'emp500_principal_investigator': 'Myrold',\n",
       "  'emp500_study_id': 59,\n",
       "  'emp500_title': 'Oregon soils',\n",
       "  'empo_2': 'Non-saline',\n",
       "  'empo_3': 'Soil (non-saline)',\n",
       "  'env biome': 'temperate shrubland biome',\n",
       "  'env feature': 'temperate shrubland',\n",
       "  'env package': 'soil',\n",
       "  'env_material': 'soil',\n",
       "  'environmental_package': 'soil',\n",
       "  'experimental_factor': 'Biogeographical study',\n",
       "  'gcms_sample_name': nan,\n",
       "  'geo loc name': 'USA: Oregon,Josephine County',\n",
       "  'latitude': 42.237306,\n",
       "  'lcms_batch': 8.0,\n",
       "  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
       "  'local_class': 'Clayey-skeletal, magnesic, mesic Typic Dystroxerepts',\n",
       "  'longitude': -123.654694,\n",
       "  'physical_specimen_location': 'PNNL and David Myrold Laboratory',\n",
       "  'project name': 'Microbial Biogeography of Oregon Soils',\n",
       "  'sample name': 'qiita_sid_13114:13114.myrold.59.s022',\n",
       "  'sample type': 'soil',\n",
       "  'sampling method': 'bulk',\n",
       "  'scientific_name': nan,\n",
       "  'soil type': 'Eightlar',\n",
       "  'store_cond': -80,\n",
       "  'study_sample_number': 22,\n",
       "  'tier': 1.0,\n",
       "  'water content': 21.57024981,\n",
       "  'water_content_soil_meth': 'gravemetric',\n",
       "  'land_cover': 52,\n",
       "  'envo_curie': 'ENVO:01000869',\n",
       "  'envo_label': 'area of scrub'}}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 61
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.689398Z",
     "start_time": "2025-02-26T02:56:21.685775Z"
    }
   },
   "cell_type": "code",
   "source": [
    "macrae_crerar_biosample_curations_list = []\n",
    "with open(macrae_crerar_biosample_clusters_with_coordinates_tsv, 'r', encoding='utf-8') as file:\n",
    "    reader = csv.DictReader(file, delimiter='\\t')\n",
    "    for row in reader:\n",
    "      macrae_crerar_biosample_curations_list.append(row)"
   ],
   "id": "6d6c602ac38c2180",
   "outputs": [],
   "execution_count": 62
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.730200Z",
     "start_time": "2025-02-26T02:56:21.727441Z"
    }
   },
   "cell_type": "code",
   "source": "macrae_crerar_biosample_curations_dict = {i['accession']: i for i in macrae_crerar_biosample_curations_list }",
   "id": "592f256525dca200",
   "outputs": [],
   "execution_count": 63
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.780721Z",
     "start_time": "2025-02-26T02:56:21.763624Z"
    }
   },
   "cell_type": "code",
   "source": "pprint.pprint(macrae_crerar_biosample_curations_dict)",
   "id": "89fb1012b47d29c5",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'SAMEA7724107': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724107',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s001',\n",
      "                  'accession': 'SAMEA7724107',\n",
      "                  'air_temp': '27.8',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '353',\n",
      "                  'block': 'B09',\n",
      "                  'cluster': '1',\n",
      "                  'collection_timestamp': '7/9/12 15:12',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1741',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_1_008_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000216',\n",
      "                  'inferred_label': 'montane shrubland biome',\n",
      "                  'latitude': '51.01.658',\n",
      "                  'latitude_decimal': '51.1994444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.643',\n",
      "                  'longitude_decimal': '100.928611111111',\n",
      "                  'ph': '7.3',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '51.6',\n",
      "                  'resp_ave': '0.565',\n",
      "                  'respiration1': '0.62',\n",
      "                  'respiration2': '0.51',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s001',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '11.4',\n",
      "                  'soil_moisture2': '11.5',\n",
      "                  'soil_moisture3': '13.9',\n",
      "                  'soil_moisture_ave': '12.26666667',\n",
      "                  'soil_salinity1': '74',\n",
      "                  'soil_salinity2': '76',\n",
      "                  'soil_salinity3': '65',\n",
      "                  'soil_salinity_ave': '71.66666667',\n",
      "                  'soil_temp1': '27.4',\n",
      "                  'soil_temp2': '31.2',\n",
      "                  'soil_temp3': '30.4',\n",
      "                  'soil_temp_ave': '29.66666667',\n",
      "                  'study_sample_number': '1'},\n",
      " 'SAMEA7724108': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724108',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s002',\n",
      "                  'accession': 'SAMEA7724108',\n",
      "                  'air_temp': '24.9',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '869',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '672',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '384',\n",
      "                  'block': 'B09',\n",
      "                  'cluster': '2',\n",
      "                  'collection_timestamp': '7/9/12 15:05',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1741',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_2_005_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000216',\n",
      "                  'inferred_label': 'montane shrubland biome',\n",
      "                  'latitude': '51.01.658',\n",
      "                  'latitude_decimal': '51.1994444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.643',\n",
      "                  'longitude_decimal': '100.928611111111',\n",
      "                  'ph': '7',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '47.1',\n",
      "                  'resp_ave': '0.275',\n",
      "                  'respiration1': '0.29',\n",
      "                  'respiration2': '0.26',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s002',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '8.5',\n",
      "                  'soil_moisture2': '7.1',\n",
      "                  'soil_moisture3': '8.9',\n",
      "                  'soil_moisture_ave': '8.166666667',\n",
      "                  'soil_salinity1': 'not provided',\n",
      "                  'soil_salinity2': 'not provided',\n",
      "                  'soil_salinity3': '76',\n",
      "                  'soil_salinity_ave': '76',\n",
      "                  'soil_temp1': '28.3',\n",
      "                  'soil_temp2': '27.6',\n",
      "                  'soil_temp3': '27.2',\n",
      "                  'soil_temp_ave': '27.7',\n",
      "                  'study_sample_number': '2'},\n",
      " 'SAMEA7724109': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724109',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s003',\n",
      "                  'accession': 'SAMEA7724109',\n",
      "                  'air_temp': '26.2',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '890',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '640',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '389',\n",
      "                  'block': 'B09',\n",
      "                  'cluster': '1',\n",
      "                  'collection_timestamp': '7/9/12 15:17',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1741',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_3_006_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000216',\n",
      "                  'inferred_label': 'montane shrubland biome',\n",
      "                  'latitude': '51.01.658',\n",
      "                  'latitude_decimal': '51.1994444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.643',\n",
      "                  'longitude_decimal': '100.928611111111',\n",
      "                  'ph': '7.1',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '45.5',\n",
      "                  'resp_ave': '0.43',\n",
      "                  'respiration1': '0.55',\n",
      "                  'respiration2': '0.31',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s003',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '9.7',\n",
      "                  'soil_moisture2': '10',\n",
      "                  'soil_moisture3': '11.6',\n",
      "                  'soil_moisture_ave': '10.43333333',\n",
      "                  'soil_salinity1': '82',\n",
      "                  'soil_salinity2': '80',\n",
      "                  'soil_salinity3': '65',\n",
      "                  'soil_salinity_ave': '75.66666667',\n",
      "                  'soil_temp1': '27.8',\n",
      "                  'soil_temp2': '30.3',\n",
      "                  'soil_temp3': '31.9',\n",
      "                  'soil_temp_ave': '30',\n",
      "                  'study_sample_number': '3'},\n",
      " 'SAMEA7724110': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724110',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s004',\n",
      "                  'accession': 'SAMEA7724110',\n",
      "                  'air_temp': '24.6',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '1107',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '620',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '359',\n",
      "                  'block': 'B09',\n",
      "                  'cluster': '1',\n",
      "                  'collection_timestamp': '7/9/12 14:59',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1741',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_4_009_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000216',\n",
      "                  'inferred_label': 'montane shrubland biome',\n",
      "                  'latitude': '51.01.658',\n",
      "                  'latitude_decimal': '51.1994444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.643',\n",
      "                  'longitude_decimal': '100.928611111111',\n",
      "                  'ph': '7.25',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '46.7',\n",
      "                  'resp_ave': '0.71',\n",
      "                  'respiration1': '0.67',\n",
      "                  'respiration2': '0.75',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s004',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '12.1',\n",
      "                  'soil_moisture2': '10.1',\n",
      "                  'soil_moisture3': '12.7',\n",
      "                  'soil_moisture_ave': '11.63333333',\n",
      "                  'soil_salinity1': '66',\n",
      "                  'soil_salinity2': '77',\n",
      "                  'soil_salinity3': '71',\n",
      "                  'soil_salinity_ave': '71.33333333',\n",
      "                  'soil_temp1': '30.7',\n",
      "                  'soil_temp2': '31.6',\n",
      "                  'soil_temp3': '30',\n",
      "                  'soil_temp_ave': '30.76666667',\n",
      "                  'study_sample_number': '4'},\n",
      " 'SAMEA7724111': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724111',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s005',\n",
      "                  'accession': 'SAMEA7724111',\n",
      "                  'air_temp': '35.2',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '1012',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '634',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '354',\n",
      "                  'block': 'B10',\n",
      "                  'cluster': '2',\n",
      "                  'collection_timestamp': '7/9/12 14:15',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1747',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_5_007_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.550',\n",
      "                  'latitude_decimal': '51.1694444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.810',\n",
      "                  'longitude_decimal': '100.975',\n",
      "                  'ph': '7.2',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '30.8',\n",
      "                  'resp_ave': '0.665',\n",
      "                  'respiration1': '0.69',\n",
      "                  'respiration2': '0.64',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s005',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '6.8',\n",
      "                  'soil_moisture2': '7.1',\n",
      "                  'soil_moisture3': '10.7',\n",
      "                  'soil_moisture_ave': '8.2',\n",
      "                  'soil_salinity1': 'not provided',\n",
      "                  'soil_salinity2': 'not provided',\n",
      "                  'soil_salinity3': '69',\n",
      "                  'soil_salinity_ave': '69',\n",
      "                  'soil_temp1': '33.1',\n",
      "                  'soil_temp2': '34.8',\n",
      "                  'soil_temp3': '32.1',\n",
      "                  'soil_temp_ave': '33.33333333',\n",
      "                  'study_sample_number': '5'},\n",
      " 'SAMEA7724112': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724112',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s006',\n",
      "                  'accession': 'SAMEA7724112',\n",
      "                  'air_temp': '30.3',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '992',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '666',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '371',\n",
      "                  'block': 'B10',\n",
      "                  'cluster': '0',\n",
      "                  'collection_timestamp': '7/9/12 14:20',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1747',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_6_002_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.550',\n",
      "                  'latitude_decimal': '51.1694444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.810',\n",
      "                  'longitude_decimal': '100.975',\n",
      "                  'ph': '7',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '34.6',\n",
      "                  'resp_ave': '0.49',\n",
      "                  'respiration1': '0.52',\n",
      "                  'respiration2': '0.46',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s006',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '5.9',\n",
      "                  'soil_moisture2': '3.8',\n",
      "                  'soil_moisture3': '8.7',\n",
      "                  'soil_moisture_ave': '6.133333333',\n",
      "                  'soil_salinity1': 'not provided',\n",
      "                  'soil_salinity2': 'not provided',\n",
      "                  'soil_salinity3': '87',\n",
      "                  'soil_salinity_ave': '87',\n",
      "                  'soil_temp1': '29.5',\n",
      "                  'soil_temp2': '31.9',\n",
      "                  'soil_temp3': '32.2',\n",
      "                  'soil_temp_ave': '31.2',\n",
      "                  'study_sample_number': '6'},\n",
      " 'SAMEA7724113': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724113',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s007',\n",
      "                  'accession': 'SAMEA7724113',\n",
      "                  'air_temp': '30.6',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '',\n",
      "                  'block': 'B10',\n",
      "                  'cluster': '0',\n",
      "                  'collection_timestamp': '7/9/12 14:38',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1747',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_7_012_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.550',\n",
      "                  'latitude_decimal': '51.1694444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.810',\n",
      "                  'longitude_decimal': '100.975',\n",
      "                  'ph': '7.1',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '35.1',\n",
      "                  'resp_ave': '0.915',\n",
      "                  'respiration1': '0.84',\n",
      "                  'respiration2': '0.99',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s007',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '5.9',\n",
      "                  'soil_moisture2': '4',\n",
      "                  'soil_moisture3': '6.6',\n",
      "                  'soil_moisture_ave': '5.5',\n",
      "                  'soil_salinity1': 'not provided',\n",
      "                  'soil_salinity2': 'not provided',\n",
      "                  'soil_salinity3': 'not provided',\n",
      "                  'soil_salinity_ave': 'not provided',\n",
      "                  'soil_temp1': '29.9',\n",
      "                  'soil_temp2': '32.6',\n",
      "                  'soil_temp3': '32.9',\n",
      "                  'soil_temp_ave': '31.8',\n",
      "                  'study_sample_number': '7'},\n",
      " 'SAMEA7724114': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724114',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s008',\n",
      "                  'accession': 'SAMEA7724114',\n",
      "                  'air_temp': '28.7',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '676',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '587',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '338',\n",
      "                  'block': 'B10',\n",
      "                  'cluster': '1',\n",
      "                  'collection_timestamp': '7/9/12 14:32',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1747',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_8_003_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.550',\n",
      "                  'latitude_decimal': '51.1694444444444',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.810',\n",
      "                  'longitude_decimal': '100.975',\n",
      "                  'ph': '7.25',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '35.3',\n",
      "                  'resp_ave': '0.435',\n",
      "                  'respiration1': '0.57',\n",
      "                  'respiration2': '0.3',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s008',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '8.6',\n",
      "                  'soil_moisture2': '9.3',\n",
      "                  'soil_moisture3': '10.6',\n",
      "                  'soil_moisture_ave': '9.5',\n",
      "                  'soil_salinity1': '73',\n",
      "                  'soil_salinity2': '69',\n",
      "                  'soil_salinity3': '63',\n",
      "                  'soil_salinity_ave': '68.33333333',\n",
      "                  'soil_temp1': '31.1',\n",
      "                  'soil_temp2': '32.8',\n",
      "                  'soil_temp3': '31.9',\n",
      "                  'soil_temp_ave': '31.93333333',\n",
      "                  'study_sample_number': '8'},\n",
      " 'SAMEA7724115': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724115',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s009',\n",
      "                  'accession': 'SAMEA7724115',\n",
      "                  'air_temp': '38.5',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '1586',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '660',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '385',\n",
      "                  'block': 'B11',\n",
      "                  'cluster': '2',\n",
      "                  'collection_timestamp': '7/9/12 13:59',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1735',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_9_010_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.524',\n",
      "                  'latitude_decimal': '51.1622222222222',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.836',\n",
      "                  'longitude_decimal': '100.982222222222',\n",
      "                  'ph': '6.95',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '24.5',\n",
      "                  'resp_ave': '0.64',\n",
      "                  'respiration1': '0.59',\n",
      "                  'respiration2': '0.69',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s009',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '5.1',\n",
      "                  'soil_moisture2': '8',\n",
      "                  'soil_moisture3': '9.3',\n",
      "                  'soil_moisture_ave': '7.466666667',\n",
      "                  'soil_salinity1': 'not provided',\n",
      "                  'soil_salinity2': 'not provided',\n",
      "                  'soil_salinity3': '64',\n",
      "                  'soil_salinity_ave': '64',\n",
      "                  'soil_temp1': '34.4',\n",
      "                  'soil_temp2': '35.3',\n",
      "                  'soil_temp3': '33.5',\n",
      "                  'soil_temp_ave': '34.4',\n",
      "                  'study_sample_number': '9'},\n",
      " 'SAMEA7724116': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724116',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:06Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s010',\n",
      "                  'accession': 'SAMEA7724116',\n",
      "                  'air_temp': '29',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '701',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '647',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '396',\n",
      "                  'block': 'B11',\n",
      "                  'cluster': '1',\n",
      "                  'collection_timestamp': '7/9/12 13:42',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1735',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_10_001_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.524',\n",
      "                  'latitude_decimal': '51.1622222222222',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.836',\n",
      "                  'longitude_decimal': '100.982222222222',\n",
      "                  'ph': '7.23',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '34.9',\n",
      "                  'resp_ave': '0.545',\n",
      "                  'respiration1': '0.71',\n",
      "                  'respiration2': '0.38',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s010',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '10.6',\n",
      "                  'soil_moisture2': '10.3',\n",
      "                  'soil_moisture3': '8.8',\n",
      "                  'soil_moisture_ave': '9.9',\n",
      "                  'soil_salinity1': '68',\n",
      "                  'soil_salinity2': '68',\n",
      "                  'soil_salinity3': '67',\n",
      "                  'soil_salinity_ave': '67.66666667',\n",
      "                  'soil_temp1': '31.2',\n",
      "                  'soil_temp2': '30.8',\n",
      "                  'soil_temp3': '29.9',\n",
      "                  'soil_temp_ave': '30.63333333',\n",
      "                  'study_sample_number': '10'},\n",
      " 'SAMEA7724117': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724117',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:05Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s011',\n",
      "                  'accession': 'SAMEA7724117',\n",
      "                  'air_temp': '30.6',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '954',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '573',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '347',\n",
      "                  'block': 'B11',\n",
      "                  'cluster': '0',\n",
      "                  'collection_timestamp': '7/9/12 14:05',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1735',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_11_011_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.524',\n",
      "                  'latitude_decimal': '51.1622222222222',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.836',\n",
      "                  'longitude_decimal': '100.982222222222',\n",
      "                  'ph': '6.77',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '41.5',\n",
      "                  'resp_ave': '0.645',\n",
      "                  'respiration1': '0.66',\n",
      "                  'respiration2': '0.63',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s011',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '3.5',\n",
      "                  'soil_moisture2': '8.3',\n",
      "                  'soil_moisture3': '5.9',\n",
      "                  'soil_moisture_ave': '5.9',\n",
      "                  'soil_salinity1': 'not provided',\n",
      "                  'soil_salinity2': 'not provided',\n",
      "                  'soil_salinity3': 'not provided',\n",
      "                  'soil_salinity_ave': 'not provided',\n",
      "                  'soil_temp1': '35.2',\n",
      "                  'soil_temp2': '36.6',\n",
      "                  'soil_temp3': '35.4',\n",
      "                  'soil_temp_ave': '35.73333333',\n",
      "                  'study_sample_number': '11'},\n",
      " 'SAMEA7724118': {'ENA first public': '2020-12-17',\n",
      "                  'ENA last update': '2020-12-16',\n",
      "                  'ENA-CHECKLIST': 'ERC000011',\n",
      "                  'External Id': 'SAMEA7724118',\n",
      "                  'INSDC center alias': 'UCSDMI',\n",
      "                  'INSDC center name': 'University of California San Diego '\n",
      "                                       'Microbiome Initiative',\n",
      "                  'INSDC first public': '2020-12-17T04:08:05Z',\n",
      "                  'INSDC last update': '2020-12-16T01:23:54Z',\n",
      "                  'INSDC status': 'public',\n",
      "                  'Submitter Id': 'qiita_sid_13114:13114.macrae.crerar.3.s012',\n",
      "                  'accession': 'SAMEA7724118',\n",
      "                  'air_temp': '27.8',\n",
      "                  'alpha_shotgun_woltka_min10k_richness': '868',\n",
      "                  'alpha_shotgun_woltka_rar10k_richness': '580',\n",
      "                  'alpha_shotgun_woltka_rar3450_richness': '355',\n",
      "                  'block': 'B11',\n",
      "                  'cluster': '2',\n",
      "                  'collection_timestamp': '7/9/12 13:34',\n",
      "                  'cur_vegetation': 'montaine steppe',\n",
      "                  'depth_sample': '0.05',\n",
      "                  'drainage_class': 'well drained',\n",
      "                  'elevation': '1735',\n",
      "                  'emp500_principal_investigator': 'MacRae-Crerar',\n",
      "                  'emp500_study_id': '3',\n",
      "                  'emp500_title': 'Mongolia soils (NSF PIRE)',\n",
      "                  'empo_2': 'Non-saline',\n",
      "                  'empo_3': 'Soil (non-saline)',\n",
      "                  'env biome': 'montane grasslands and shrubland biome',\n",
      "                  'env feature': 'steppe soil',\n",
      "                  'env package': 'soil',\n",
      "                  'env_material': 'soil',\n",
      "                  'environmental_package': 'soil',\n",
      "                  'gcms_sample_name': 'EMP_McR3_12_004_GC',\n",
      "                  'geo loc name': 'Mongolia',\n",
      "                  'inferred_curie': 'ENVO:01000194',\n",
      "                  'inferred_label': 'montane grassland biome',\n",
      "                  'latitude': '51.01.524',\n",
      "                  'latitude_decimal': '51.1622222222222',\n",
      "                  'lcms_batch': '8',\n",
      "                  'lcms_extraction_protocol': 'Bulk_Seed_PNNL',\n",
      "                  'longitude': '100.45.836',\n",
      "                  'longitude_decimal': '100.982222222222',\n",
      "                  'ph': '6.95',\n",
      "                  'physical_specimen_location': 'UCSD Knight Lab Freezer I and '\n",
      "                                                'University of Pennsylvania',\n",
      "                  'project name': 'NSF PIRE Mongolia Project',\n",
      "                  'relative humidity': '40',\n",
      "                  'resp_ave': '0.77',\n",
      "                  'respiration1': '0.83',\n",
      "                  'respiration2': '0.71',\n",
      "                  'sample name': 'qiita_sid_13114:13114.macrae.crerar.3.s012',\n",
      "                  'sample type': 'soil',\n",
      "                  'sampling method': 'bulk',\n",
      "                  'soil_moisture1': '10.4',\n",
      "                  'soil_moisture2': '9.6',\n",
      "                  'soil_moisture3': '7',\n",
      "                  'soil_moisture_ave': '9',\n",
      "                  'soil_salinity1': '69',\n",
      "                  'soil_salinity2': '76',\n",
      "                  'soil_salinity3': 'not provided',\n",
      "                  'soil_salinity_ave': '72.5',\n",
      "                  'soil_temp1': '29.2',\n",
      "                  'soil_temp2': '30',\n",
      "                  'soil_temp3': '29.4',\n",
      "                  'soil_temp_ave': '29.53333333',\n",
      "                  'study_sample_number': '12'}}\n"
     ]
    }
   ],
   "execution_count": 64
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.819339Z",
     "start_time": "2025-02-26T02:56:21.816505Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Create a lookup dictionary from term_legitimacy_data using content_value as the key\n",
    "term_lookup = {entry[\"content_value\"]: entry for entry in term_legitimacy_data}"
   ],
   "id": "797cfc298d847a9d",
   "outputs": [],
   "execution_count": 65
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.897791Z",
     "start_time": "2025-02-26T02:56:21.851891Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# List to collect logs for missing term annotations\n",
    "missing_term_logs = []\n",
    "\n",
    "# Expected subset mappings and corresponding boolean flags\n",
    "subset_validations = {\n",
    "    \"env_broad_scale\": (\"is_biome\", biome_curies),\n",
    "    \"env_local_scale\": (\"is_selected_abp\", non_biome_non_env_mat_abp_curies),\n",
    "    \"env_medium\": (\"is_env_mat\", env_mat_curies)  # Corrected order\n",
    "}\n",
    "\n",
    "# Iterate over each biosample record\n",
    "for k, sample in emp500_integrated_biosample_summaries.items():\n",
    "\n",
    "    for key in target_harmonized_names:\n",
    "        original_value = sample.get(key, \"\").strip()\n",
    "\n",
    "        # Lookup the term legitimacy data\n",
    "        term_data = term_lookup.get(original_value, {})\n",
    "\n",
    "        # Extract matching or replacement details\n",
    "        matched_label = term_data.get(\"matched_label\", \"\")\n",
    "        matched_curie = term_data.get(\"matched_curie\", \"\")\n",
    "        replacement_label = term_data.get(\"replacement_label\", \"\")\n",
    "        replacement_curie = term_data.get(\"replacement_curie\", \"\")\n",
    "        consider_replacements = term_data.get(\"consider_replacements\", \"\")\n",
    "\n",
    "        via_obsolescence = False  # Default to False\n",
    "\n",
    "        # Lookup curation in per_slot_value_curations_dict\n",
    "        curation_key = (key, original_value)\n",
    "        curation = per_slot_value_curations_dict.get(curation_key)\n",
    "\n",
    "        if curation:\n",
    "            if curation[\"done\"] and curation[\"replacement_curie\"] and curation[\"replacement_label\"]:\n",
    "                sample[f\"{key}_curation_curie\"] = curation[\"replacement_curie\"]\n",
    "                sample[f\"{key}_curation_label\"] = curation[\"replacement_label\"]\n",
    "\n",
    "        if sample[\"accession\"] in macrae_crerar_biosample_curations_dict:\n",
    "            sample[f\"{key}_curation_curie\"] = macrae_crerar_biosample_curations_dict[sample[\"accession\"]][\n",
    "                'inferred_curie']\n",
    "            sample[f\"{key}_curation_label\"] = macrae_crerar_biosample_curations_dict[sample[\"accession\"]][\n",
    "                'inferred_label']\n",
    "\n",
    "        if sample[\"accession\"] in myrold_attributes_with_land_cover_dict:\n",
    "            sample[f\"{key}_curation_curie\"] = myrold_attributes_with_land_cover_dict[sample[\"accession\"]][\n",
    "                'envo_curie']\n",
    "            sample[f\"{key}_curation_label\"] = myrold_attributes_with_land_cover_dict[sample[\"accession\"]][\n",
    "                'envo_label']\n",
    "\n",
    "        if replacement_label and replacement_curie:\n",
    "            sample[f\"{key}_via_obsolescence\"] = True\n",
    "            sample[f\"{key}_replacement_label\"] = replacement_label\n",
    "            sample[f\"{key}_replacement_curie\"] = replacement_curie\n",
    "        elif matched_label and matched_curie:\n",
    "            sample[f\"{key}_matched_label\"] = matched_label\n",
    "            sample[f\"{key}_matched_curie\"] = matched_curie\n",
    "        else:\n",
    "            # Log missing annotation details\n",
    "            missing_term_logs.append({\n",
    "                \"accession\": sample[\"accession\"],\n",
    "                \"problematic_key\": key,\n",
    "                \"problematic_value\": original_value,\n",
    "                \"consider_replacements\": consider_replacements\n",
    "            })\n",
    "\n",
    "        # Validate subset membership and set boolean flag\n",
    "        curie = replacement_curie if via_obsolescence else matched_curie\n",
    "        subset_key, expected_curies = subset_validations.get(key, (None, None))\n",
    "\n",
    "        if subset_key:\n",
    "            sample[subset_key] = curie in expected_curies if curie else False\n"
   ],
   "id": "c120b9a17f3d4168",
   "outputs": [],
   "execution_count": 66
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.935891Z",
     "start_time": "2025-02-26T02:56:21.930118Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Write missing_term_logs to a TSV file\n",
    "if missing_term_logs:\n",
    "    fieldnames = missing_term_logs[0].keys()  # Get all column headers from the first row\n",
    "\n",
    "    with open(missing_terms_tsv, mode=\"w\", newline=\"\", encoding=\"utf-8\") as file:\n",
    "        writer = csv.DictWriter(file, delimiter=\"\\t\", fieldnames=fieldnames)\n",
    "        writer.writeheader()\n",
    "        writer.writerows(missing_term_logs)\n",
    "\n",
    "    print(f\"Missing term logs saved to {missing_terms_tsv}\")"
   ],
   "id": "f00ff733f4c52e49",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing term logs saved to outputs/emp500_annotations_for_review.tsv\n"
     ]
    }
   ],
   "execution_count": 67
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:21.974765Z",
     "start_time": "2025-02-26T02:56:21.969320Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Initialize a Counter to track occurrences of (problematic_key, problematic_value)\n",
    "problematic_counts = Counter()\n",
    "\n",
    "# Count occurrences of each unique problematic_key/problematic_value combination\n",
    "for entry in missing_term_logs:\n",
    "    key_value_pair = (entry[\"problematic_key\"], entry[\"problematic_value\"])\n",
    "    problematic_counts[key_value_pair] += 1\n",
    "\n",
    "# Convert the Counter into a list of dictionaries\n",
    "problematic_summary = [\n",
    "    {\"problematic_key\": key, \"problematic_value\": value, \"count\": count}\n",
    "    for (key, value), count in problematic_counts.items()\n",
    "]\n"
   ],
   "id": "428cc1e348367c09",
   "outputs": [],
   "execution_count": 68
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:22.023586Z",
     "start_time": "2025-02-26T02:56:22.019461Z"
    }
   },
   "cell_type": "code",
   "source": "problematic_counts",
   "id": "869018ef3e02e1a3",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({('env_local_scale', 'host-associated habitat'): 133,\n",
       "         ('env_local_scale', 'marine benthic feature'): 125,\n",
       "         ('env_local_scale', 'animal-associated habitat'): 84,\n",
       "         ('env_local_scale', 'insecta-associated habitat'): 28,\n",
       "         ('env_local_scale', 'montane desert'): 26,\n",
       "         ('env_local_scale', 'forest'): 21,\n",
       "         ('env_broad_scale', 'Small lake biome'): 20,\n",
       "         ('env_local_scale', 'stratum'): 18,\n",
       "         ('env_local_scale', 'temperate coniferous forest'): 16,\n",
       "         ('env_local_scale', 'oil reservior'): 14,\n",
       "         ('env_broad_scale', 'montane grasslands and shrubland biome'): 12,\n",
       "         ('env_local_scale', 'cropland'): 10,\n",
       "         ('env_medium', 'coastal water'): 9,\n",
       "         ('env_local_scale', 'wetland'): 9,\n",
       "         ('env_local_scale', 'intermittent pond'): 7,\n",
       "         ('env_broad_scale', 'temperate shrubland'): 6,\n",
       "         ('env_medium', 'pitcher plant fluid'): 6,\n",
       "         ('env_local_scale', 'rangeland'): 5,\n",
       "         ('env_local_scale', 'bog'): 3,\n",
       "         ('env_local_scale', 'temperate woodland'): 2,\n",
       "         ('env_local_scale', 'mangrove'): 1,\n",
       "         ('env_broad_scale', 'flooded savanna'): 1,\n",
       "         ('env_broad_scale', 'neritic bathypelagic zone biome'): 1,\n",
       "         ('env_local_scale', 'temperate shrubland'): 1})"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 69
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:22.061740Z",
     "start_time": "2025-02-26T02:56:22.058159Z"
    }
   },
   "cell_type": "code",
   "source": [
    "fieldnames = problematic_summary[0].keys()  # Get all column headers from the first row\n",
    "\n",
    "with open(problematic_summary_tsv, mode=\"w\", newline=\"\", encoding=\"utf-8\") as file:\n",
    "    writer = csv.DictWriter(file, delimiter=\"\\t\", fieldnames=fieldnames)\n",
    "    writer.writeheader()\n",
    "    writer.writerows(problematic_summary)"
   ],
   "id": "37a895d5d47311e4",
   "outputs": [],
   "execution_count": 70
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:22.097734Z",
     "start_time": "2025-02-26T02:56:22.094526Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# make a LOD FOR WRITING TO TSV\n",
    "emp500_integrated_biosample_summaries_lod = list(emp500_integrated_biosample_summaries.values())"
   ],
   "id": "30171725136caee4",
   "outputs": [],
   "execution_count": 71
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:22.135273Z",
     "start_time": "2025-02-26T02:56:22.130955Z"
    }
   },
   "cell_type": "code",
   "source": [
    "\n",
    "ordered_columns = [\n",
    "    \"accession\",\n",
    "    \"biosampleGoldId\",\n",
    "    \"ncbiBioProjectAccession\",\n",
    "    \"sraExperimentIds\",\n",
    "    \"modDate\",\n",
    "    \"addDate\",\n",
    "    \"ecosystemPathId\",\n",
    "    \"ecosystem\",\n",
    "    \"ecosystemCategory\",\n",
    "    \"ecosystemType\",\n",
    "    \"ecosystemSubtype\",\n",
    "    \"specificEcosystem\",\n",
    "    \"envoBroadScale_id\",\n",
    "    \"envoBroadScale_label\",\n",
    "    \"envoBroadScale_label_lu\",\n",
    "    \"envoBroadScale_label_exact_match\",\n",
    "    \"envoBroadScale_biome\",\n",
    "    \"envoLocalScale_id\",\n",
    "    \"envoLocalScale_label\",\n",
    "    \"envoLocalScale_label_lu\",\n",
    "    \"envoLocalScale_label_exact_match\",\n",
    "    \"envoLocalScale_abp_subset\",\n",
    "    \"envoMedium_id\",\n",
    "    \"envoMedium_label\",\n",
    "    \"envoMedium_label_lu\",\n",
    "    \"envoMedium_label_exact_match\",\n",
    "    \"envoMedium_env_mat\",\n",
    "    \"envoMedium_soil\",\n",
    "    \"envoMedium_liq_water\",\n",
    "    \"envoMedium_sediment\",\n",
    "    # \"envoMedium_feces\",\n",
    "    \"envoMedium_fecal_material\",\n",
    "    \"envoMedium_animal_manure\",\n",
    "    \"envoMedium_bodily_fluid_material\",\n",
    "    \"envoMedium_sludge\",\n",
    "    \"envoMedium_petroleum\",\n",
    "    \"envoMedium_accounted_for\",\n",
    "    \"env_broad_scale\",\n",
    "    \"env_broad_scale_matched_curie\",\n",
    "    \"env_broad_scale_matched_label\",\n",
    "    \"env_broad_scale_replacement_curie\",\n",
    "    \"env_broad_scale_replacement_label\",\n",
    "    \"env_broad_scale_via_obsolescence\",\n",
    "    \"is_biome\",\n",
    "    \"env_broad_scale_curation_curie\",\n",
    "    \"env_broad_scale_curation_label\",\n",
    "    \"env_local_scale\",\n",
    "    \"env_local_scale_matched_curie\",\n",
    "    \"env_local_scale_matched_label\",\n",
    "    \"env_local_scale_replacement_curie\",\n",
    "    \"env_local_scale_replacement_label\",\n",
    "    \"env_local_scale_via_obsolescence\",\n",
    "    \"is_selected_abp\",\n",
    "    \"env_local_scale_curation_curie\",\n",
    "    \"env_local_scale_curation_label\",\n",
    "    \"env_medium\",\n",
    "    \"env_medium_matched_curie\",\n",
    "    \"env_medium_matched_label\",\n",
    "    \"env_medium_replacement_curie\",\n",
    "    \"env_medium_replacement_label\",\n",
    "    \"env_medium_via_obsolescence\",\n",
    "    \"is_env_mat\",\n",
    "    \"env_medium_curation_curie\",\n",
    "    \"env_medium_curation_label\",\n",
    "]\n"
   ],
   "id": "c838477191cfc9a2",
   "outputs": [],
   "execution_count": 72
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-26T02:56:22.197985Z",
     "start_time": "2025-02-26T02:56:22.169332Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Write merged data to a TSV file\n",
    "with open(emp500_mixs_style_triads_plus_gold_ids_tsv, mode=\"w\", newline=\"\", encoding=\"utf-8\") as file:\n",
    "    writer = csv.DictWriter(file, delimiter=\"\\t\", fieldnames=ordered_columns)\n",
    "    writer.writeheader()\n",
    "    writer.writerows(emp500_integrated_biosample_summaries_lod)\n",
    "\n",
    "print(f\"Merged MIxS-style environmental triad data saved to {emp500_mixs_style_triads_plus_gold_ids_tsv}\")"
   ],
   "id": "2616b0cb928c36d0",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged MIxS-style environmental triad data saved to outputs/emp500_matches_and_curations_plus_gold_ids.tsv\n"
     ]
    }
   ],
   "execution_count": 73
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
