{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-02-11T21:12:08.094374Z",
     "start_time": "2025-02-11T21:12:08.086058Z"
    }
   },
   "source": [
    "import pandas as pd\n",
    "import string\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import pprint"
   ],
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:01.372455Z",
     "start_time": "2025-02-11T21:07:01.355494Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Read the TSV files into pandas DataFrames\n",
    "df_citations = pd.read_csv(\"doi_metadata.tsv\", sep=\"\\t\")\n",
    "df_metadata = pd.read_csv(\"pdf_metadata.tsv\", sep=\"\\t\")\n"
   ],
   "id": "cd550b727fef33a1",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:02.862310Z",
     "start_time": "2025-02-11T21:07:02.856892Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Combine title and author information into a single field for both datasets\n",
    "df_citations['title_author'] = df_citations['Title'] + ' ' + df_citations['Authors']\n",
    "df_metadata['title_author'] = df_metadata['Title'] + ' ' + df_metadata['Author']\n"
   ],
   "id": "bd43537b2df83728",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:04.144902Z",
     "start_time": "2025-02-11T21:07:04.134430Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Preprocess the combined field\n",
    "def preprocess_text(text):\n",
    "    if pd.isnull(text):  # Handle NaN values\n",
    "        return ''\n",
    "    text = text.translate(str.maketrans('', '', string.punctuation))  # Remove punctuation\n",
    "    text = text.lower()  # Convert to lowercase\n",
    "    return text.split()  # Tokenize\n"
   ],
   "id": "dbc1bb61c353ad80",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:05.480121Z",
     "start_time": "2025-02-11T21:07:05.467876Z"
    }
   },
   "cell_type": "code",
   "source": [
    "df_citations['title_author_tokens'] = df_citations['title_author'].apply(preprocess_text)\n",
    "df_metadata['title_author_tokens'] = df_metadata['title_author'].apply(preprocess_text)\n"
   ],
   "id": "adc539d49a4db32f",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:21.340461Z",
     "start_time": "2025-02-11T21:07:21.296810Z"
    }
   },
   "cell_type": "code",
   "source": "df_citations",
   "id": "f9b7496ef86596f9",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                DOI                 Updated DOI  \\\n",
       "0         10.1128/mSystems.00045-18                         NaN   \n",
       "1         10.1101/2022.12.12.520098  10.3389/fsoil.2023.1120425   \n",
       "2        10.1038/s41564-022-01266-x                         NaN   \n",
       "3           10.1111/1462-2920.16314                         NaN   \n",
       "4                 10.1111/mec.16891                         NaN   \n",
       "5       10.1021/acs.estlett.0c00748                         NaN   \n",
       "6         10.1128/msystems.00768-19                         NaN   \n",
       "7      10.1371/journal.pone.0228165                         NaN   \n",
       "8    10.1016/j.geoderma.2021.115674                         NaN   \n",
       "9              10.1029/2022JG006889                         NaN   \n",
       "10                 10.1002/ppp.2200                         NaN   \n",
       "11       10.1038/s41467-023-36515-y                         NaN   \n",
       "12                10.1002/lno.11306                         NaN   \n",
       "13       10.1038/s41597-024-03069-7                         NaN   \n",
       "14       10.1038/s41564-020-00861-0                         NaN   \n",
       "15             10.1128/MRA.01361-19                         NaN   \n",
       "16             10.1128/mra.01080-23                         NaN   \n",
       "17        10.1038/s41597-019-0132-4                         NaN   \n",
       "18       10.1038/s41597-024-04013-5                         NaN   \n",
       "19       10.1186/s40168-020-00889-8                         NaN   \n",
       "20              10.1093/nar/gkab990                         NaN   \n",
       "21           10.1126/sciadv.adg7888                         NaN   \n",
       "22     10.1371/journal.pone.0228165                         NaN   \n",
       "23    10.3390/microorganisms9020357                         NaN   \n",
       "24  10.3897/tdwgproceedings.1.20637                         NaN   \n",
       "25        10.5194/acp-23-15783-2023                         NaN   \n",
       "\n",
       "                                                Title  \\\n",
       "0   Exploration of the Biosynthetic Potential of t...   \n",
       "1   One thousand soils for molecular understanding...   \n",
       "2   Standardized multi-omics of Earth’s microbiome...   \n",
       "3   Environmental predictors of electroactive bact...   \n",
       "4   Diversity, distribution, and expression of ops...   \n",
       "5   Long-Term Warming Decreases Redox Capacity of ...   \n",
       "6   Complementary Metagenomic Approaches Improve R...   \n",
       "7   Distinct temporal diversity profiles for nitro...   \n",
       "8   Soil pore network response to freeze-thaw cycl...   \n",
       "9   The Impact of Freeze‐Thaw History on Soil Carb...   \n",
       "10  Topography and canopy cover influence soil org...   \n",
       "11  Seasonal activities of the phyllosphere microb...   \n",
       "12  Time‐series metatranscriptomes reveal conserve...   \n",
       "13  An integrated metagenomic, metabolomic and tra...   \n",
       "14  Genomic and functional analyses of fungal and ...   \n",
       "15  Metatranscriptomic Sequencing of a Cyanobacter...   \n",
       "16  Metatranscriptomes of two biological soil crus...   \n",
       "17  Microbial metagenomes and metatranscriptomes d...   \n",
       "18  Microbial Metagenomes Across a Complete Phytop...   \n",
       "19  Influence of the polar light cycle on seasonal...   \n",
       "20  The National Microbiome Data Collaborative Dat...   \n",
       "21  Reproducible growth of<i>Brachypodium</i>in Ec...   \n",
       "22  Distinct temporal diversity profiles for nitro...   \n",
       "23  Soil Bacterial Diversity Is Positively Correla...   \n",
       "24  Using MIxS: An Implementation Report from Two ...   \n",
       "25  Active thermokarst regions contain rich  sourc...   \n",
       "\n",
       "                                            Journal  Year        Year Type  \\\n",
       "0                                          mSystems  2018  published-print   \n",
       "1                         Frontiers in Soil Science  2023        published   \n",
       "2                               Nature Microbiology  2022        published   \n",
       "3                        Environmental Microbiology  2023  published-print   \n",
       "4                                 Molecular Ecology  2023  published-print   \n",
       "5    Environmental Science &amp; Technology Letters  2021  published-print   \n",
       "6                                          mSystems  2020  published-print   \n",
       "7                                          PLOS ONE  2020        published   \n",
       "8                                          Geoderma  2022  published-print   \n",
       "9   Journal of Geophysical Research: Biogeosciences  2022  published-print   \n",
       "10             Permafrost and Periglacial Processes  2023  published-print   \n",
       "11                            Nature Communications  2023        published   \n",
       "12                       Limnology and Oceanography  2020  published-print   \n",
       "13                                  Scientific Data  2024        published   \n",
       "14                              Nature Microbiology  2021        published   \n",
       "15              Microbiology Resource Announcements  2020  published-print   \n",
       "16              Microbiology Resource Announcements  2024  published-print   \n",
       "17                                  Scientific Data  2019        published   \n",
       "18                                  Scientific Data  2024        published   \n",
       "19                                       Microbiome  2020  published-print   \n",
       "20                           Nucleic Acids Research  2022        published   \n",
       "21                                 Science Advances  2024  published-print   \n",
       "22                                         PLOS ONE  2020        published   \n",
       "23                                   Microorganisms  2021        published   \n",
       "24                              Proceedings of TDWG  2017        published   \n",
       "25                Atmospheric Chemistry and Physics  2023        published   \n",
       "\n",
       "                                            Publisher  \\\n",
       "0                   American Society for Microbiology   \n",
       "1                                  Frontiers Media SA   \n",
       "2             Springer Science and Business Media LLC   \n",
       "3                                               Wiley   \n",
       "4                                               Wiley   \n",
       "5                     American Chemical Society (ACS)   \n",
       "6                   American Society for Microbiology   \n",
       "7                    Public Library of Science (PLoS)   \n",
       "8                                         Elsevier BV   \n",
       "9                    American Geophysical Union (AGU)   \n",
       "10                                              Wiley   \n",
       "11            Springer Science and Business Media LLC   \n",
       "12                                              Wiley   \n",
       "13            Springer Science and Business Media LLC   \n",
       "14            Springer Science and Business Media LLC   \n",
       "15                  American Society for Microbiology   \n",
       "16                  American Society for Microbiology   \n",
       "17            Springer Science and Business Media LLC   \n",
       "18            Springer Science and Business Media LLC   \n",
       "19            Springer Science and Business Media LLC   \n",
       "20                      Oxford University Press (OUP)   \n",
       "21  American Association for the Advancement of Sc...   \n",
       "22                   Public Library of Science (PLoS)   \n",
       "23                                            MDPI AG   \n",
       "24                                 Pensoft Publishers   \n",
       "25                                    Copernicus GmbH   \n",
       "\n",
       "                                              Authors  Volume Issue  \\\n",
       "0   Patricia M. Blair; Miriam L. Land; Marek J. Pi...     3.0     5   \n",
       "1   Maggie M. Bowman; Alexis E. Heath; Tamas Varga...     3.0   NaN   \n",
       "2   Justin P. Shaffer; Louis-Félix Nothias; Luke R...     7.0    12   \n",
       "3   Charles N. Olmsted; Roger Ort; Patricia Q. Tra...    25.0     3   \n",
       "4   Shaomei He; Alexandra M. Linz; Sarah L. R. Ste...    32.0    11   \n",
       "5   Rachelle E. LaCroix; Nicolas Walpen; Michael S...     8.0     1   \n",
       "6   L. V. Alteio; F. Schulz; R. Seshadri; N. Vargh...     5.0     2   \n",
       "7   William C. Nelson; Emily B. Graham; Alex R. Cr...    15.0     1   \n",
       "8   Erin C. Rooney; Vanessa L. Bailey; Kaizad F. P...   411.0   NaN   \n",
       "9   Erin C. Rooney; Vanessa L. Bailey; Kaizad F. P...   127.0     5   \n",
       "10  Erin C. Rooney; Vanessa L. Bailey; Kaizad F. P...    34.0     3   \n",
       "11  Adina Howe; Nejc Stopnisek; Shane K. Dooley; F...    14.0     1   \n",
       "12  Alexandra M. Linz; Frank O. Aylward; Stefan Be...    65.0    S1   \n",
       "13  Christopher Schadt; Stanton Martin; Alyssa Car...    11.0     1   \n",
       "14  Xuefeng Peng; St. Elmo Wilken; Thomas S. Lanki...     6.0     4   \n",
       "15  Terrence H. Bell; Ryan V. Trexler; Xin Peng; M...     9.0     1   \n",
       "16  Thuy M. Nguyen; Nuttapon Pombubpa; Marcel Hunt...    13.0     2   \n",
       "17  Brent Nowinski; Christa B. Smith; Courtney M. ...     6.0     1   \n",
       "18  Brook L. Nunn; Emma Timmins-Schiffman; Miranda...    11.0     1   \n",
       "19  Pratibha Panwar; Michelle A. Allen; Timothy J....     8.0     1   \n",
       "20                                                NaN     NaN   NaN   \n",
       "21  Vlastimil Novak; Peter F. Andeer; Benjamin P. ...    10.0     1   \n",
       "22  William C. Nelson; Emily B. Graham; Alex R. Cr...    15.0     1   \n",
       "23  Akane Chiba; Yoshitaka Uchida; Susanne Kublik;...     9.0     2   \n",
       "24  Joel Sachs; Luke Thompson; Nazir El-Kayssi; Sa...     1.0   NaN   \n",
       "25  Kevin R. Barry; Thomas C. J. Hill; Marina Niet...    23.0    24   \n",
       "\n",
       "          Pages                                          DOI URL  \\\n",
       "0           NaN        https://doi.org/10.1128/msystems.00045-18   \n",
       "1           NaN       https://doi.org/10.3389/fsoil.2023.1120425   \n",
       "2     2128-2150       https://doi.org/10.1038/s41564-022-01266-x   \n",
       "3       705-720          https://doi.org/10.1111/1462-2920.16314   \n",
       "4     2798-2817                https://doi.org/10.1111/mec.16891   \n",
       "5         92-97      https://doi.org/10.1021/acs.estlett.0c00748   \n",
       "6           NaN        https://doi.org/10.1128/msystems.00768-19   \n",
       "7      e0228165     https://doi.org/10.1371/journal.pone.0228165   \n",
       "8        115674   https://doi.org/10.1016/j.geoderma.2021.115674   \n",
       "9           NaN             https://doi.org/10.1029/2022jg006889   \n",
       "10      331-358                 https://doi.org/10.1002/ppp.2200   \n",
       "11          NaN       https://doi.org/10.1038/s41467-023-36515-y   \n",
       "12          NaN                https://doi.org/10.1002/lno.11306   \n",
       "13          NaN       https://doi.org/10.1038/s41597-024-03069-7   \n",
       "14      499-511       https://doi.org/10.1038/s41564-020-00861-0   \n",
       "15          NaN             https://doi.org/10.1128/mra.01361-19   \n",
       "16          NaN             https://doi.org/10.1128/mra.01080-23   \n",
       "17          NaN        https://doi.org/10.1038/s41597-019-0132-4   \n",
       "18          NaN       https://doi.org/10.1038/s41597-024-04013-5   \n",
       "19          NaN       https://doi.org/10.1186/s40168-020-00889-8   \n",
       "20          NaN              https://doi.org/10.1093/nar/gkab990   \n",
       "21          NaN           https://doi.org/10.1126/sciadv.adg7888   \n",
       "22     e0228165     https://doi.org/10.1371/journal.pone.0228165   \n",
       "23          357    https://doi.org/10.3390/microorganisms9020357   \n",
       "24       e20637  https://doi.org/10.3897/tdwgproceedings.1.20637   \n",
       "25  15783-15793        https://doi.org/10.5194/acp-23-15783-2023   \n",
       "\n",
       "                                         title_author  \\\n",
       "0   Exploration of the Biosynthetic Potential of t...   \n",
       "1   One thousand soils for molecular understanding...   \n",
       "2   Standardized multi-omics of Earth’s microbiome...   \n",
       "3   Environmental predictors of electroactive bact...   \n",
       "4   Diversity, distribution, and expression of ops...   \n",
       "5   Long-Term Warming Decreases Redox Capacity of ...   \n",
       "6   Complementary Metagenomic Approaches Improve R...   \n",
       "7   Distinct temporal diversity profiles for nitro...   \n",
       "8   Soil pore network response to freeze-thaw cycl...   \n",
       "9   The Impact of Freeze‐Thaw History on Soil Carb...   \n",
       "10  Topography and canopy cover influence soil org...   \n",
       "11  Seasonal activities of the phyllosphere microb...   \n",
       "12  Time‐series metatranscriptomes reveal conserve...   \n",
       "13  An integrated metagenomic, metabolomic and tra...   \n",
       "14  Genomic and functional analyses of fungal and ...   \n",
       "15  Metatranscriptomic Sequencing of a Cyanobacter...   \n",
       "16  Metatranscriptomes of two biological soil crus...   \n",
       "17  Microbial metagenomes and metatranscriptomes d...   \n",
       "18  Microbial Metagenomes Across a Complete Phytop...   \n",
       "19  Influence of the polar light cycle on seasonal...   \n",
       "20                                                NaN   \n",
       "21  Reproducible growth of<i>Brachypodium</i>in Ec...   \n",
       "22  Distinct temporal diversity profiles for nitro...   \n",
       "23  Soil Bacterial Diversity Is Positively Correla...   \n",
       "24  Using MIxS: An Implementation Report from Two ...   \n",
       "25  Active thermokarst regions contain rich  sourc...   \n",
       "\n",
       "                                  title_author_tokens  \n",
       "0   [exploration, of, the, biosynthetic, potential...  \n",
       "1   [one, thousand, soils, for, molecular, underst...  \n",
       "2   [standardized, multiomics, of, earth’s, microb...  \n",
       "3   [environmental, predictors, of, electroactive,...  \n",
       "4   [diversity, distribution, and, expression, of,...  \n",
       "5   [longterm, warming, decreases, redox, capacity...  \n",
       "6   [complementary, metagenomic, approaches, impro...  \n",
       "7   [distinct, temporal, diversity, profiles, for,...  \n",
       "8   [soil, pore, network, response, to, freezethaw...  \n",
       "9   [the, impact, of, freeze‐thaw, history, on, so...  \n",
       "10  [topography, and, canopy, cover, influence, so...  \n",
       "11  [seasonal, activities, of, the, phyllosphere, ...  \n",
       "12  [time‐series, metatranscriptomes, reveal, cons...  \n",
       "13  [an, integrated, metagenomic, metabolomic, and...  \n",
       "14  [genomic, and, functional, analyses, of, funga...  \n",
       "15  [metatranscriptomic, sequencing, of, a, cyanob...  \n",
       "16  [metatranscriptomes, of, two, biological, soil...  \n",
       "17  [microbial, metagenomes, and, metatranscriptom...  \n",
       "18  [microbial, metagenomes, across, a, complete, ...  \n",
       "19  [influence, of, the, polar, light, cycle, on, ...  \n",
       "20                                                     \n",
       "21  [reproducible, growth, ofibrachypodiumiin, eco...  \n",
       "22  [distinct, temporal, diversity, profiles, for,...  \n",
       "23  [soil, bacterial, diversity, is, positively, c...  \n",
       "24  [using, mixs, an, implementation, report, from...  \n",
       "25  [active, thermokarst, regions, contain, rich, ...  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOI</th>\n",
       "      <th>Updated DOI</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "      <th>Year Type</th>\n",
       "      <th>Publisher</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Pages</th>\n",
       "      <th>DOI URL</th>\n",
       "      <th>title_author</th>\n",
       "      <th>title_author_tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10.1128/mSystems.00045-18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Exploration of the Biosynthetic Potential of t...</td>\n",
       "      <td>mSystems</td>\n",
       "      <td>2018</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Society for Microbiology</td>\n",
       "      <td>Patricia M. Blair; Miriam L. Land; Marek J. Pi...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1128/msystems.00045-18</td>\n",
       "      <td>Exploration of the Biosynthetic Potential of t...</td>\n",
       "      <td>[exploration, of, the, biosynthetic, potential...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10.1101/2022.12.12.520098</td>\n",
       "      <td>10.3389/fsoil.2023.1120425</td>\n",
       "      <td>One thousand soils for molecular understanding...</td>\n",
       "      <td>Frontiers in Soil Science</td>\n",
       "      <td>2023</td>\n",
       "      <td>published</td>\n",
       "      <td>Frontiers Media SA</td>\n",
       "      <td>Maggie M. Bowman; Alexis E. Heath; Tamas Varga...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.3389/fsoil.2023.1120425</td>\n",
       "      <td>One thousand soils for molecular understanding...</td>\n",
       "      <td>[one, thousand, soils, for, molecular, underst...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10.1038/s41564-022-01266-x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Standardized multi-omics of Earth’s microbiome...</td>\n",
       "      <td>Nature Microbiology</td>\n",
       "      <td>2022</td>\n",
       "      <td>published</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Justin P. Shaffer; Louis-Félix Nothias; Luke R...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>12</td>\n",
       "      <td>2128-2150</td>\n",
       "      <td>https://doi.org/10.1038/s41564-022-01266-x</td>\n",
       "      <td>Standardized multi-omics of Earth’s microbiome...</td>\n",
       "      <td>[standardized, multiomics, of, earth’s, microb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10.1111/1462-2920.16314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Environmental predictors of electroactive bact...</td>\n",
       "      <td>Environmental Microbiology</td>\n",
       "      <td>2023</td>\n",
       "      <td>published-print</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>Charles N. Olmsted; Roger Ort; Patricia Q. Tra...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>3</td>\n",
       "      <td>705-720</td>\n",
       "      <td>https://doi.org/10.1111/1462-2920.16314</td>\n",
       "      <td>Environmental predictors of electroactive bact...</td>\n",
       "      <td>[environmental, predictors, of, electroactive,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10.1111/mec.16891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diversity, distribution, and expression of ops...</td>\n",
       "      <td>Molecular Ecology</td>\n",
       "      <td>2023</td>\n",
       "      <td>published-print</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>Shaomei He; Alexandra M. Linz; Sarah L. R. Ste...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>11</td>\n",
       "      <td>2798-2817</td>\n",
       "      <td>https://doi.org/10.1111/mec.16891</td>\n",
       "      <td>Diversity, distribution, and expression of ops...</td>\n",
       "      <td>[diversity, distribution, and, expression, of,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10.1021/acs.estlett.0c00748</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long-Term Warming Decreases Redox Capacity of ...</td>\n",
       "      <td>Environmental Science &amp;amp; Technology Letters</td>\n",
       "      <td>2021</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Chemical Society (ACS)</td>\n",
       "      <td>Rachelle E. LaCroix; Nicolas Walpen; Michael S...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1</td>\n",
       "      <td>92-97</td>\n",
       "      <td>https://doi.org/10.1021/acs.estlett.0c00748</td>\n",
       "      <td>Long-Term Warming Decreases Redox Capacity of ...</td>\n",
       "      <td>[longterm, warming, decreases, redox, capacity...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10.1128/msystems.00768-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complementary Metagenomic Approaches Improve R...</td>\n",
       "      <td>mSystems</td>\n",
       "      <td>2020</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Society for Microbiology</td>\n",
       "      <td>L. V. Alteio; F. Schulz; R. Seshadri; N. Vargh...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1128/msystems.00768-19</td>\n",
       "      <td>Complementary Metagenomic Approaches Improve R...</td>\n",
       "      <td>[complementary, metagenomic, approaches, impro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10.1371/journal.pone.0228165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Distinct temporal diversity profiles for nitro...</td>\n",
       "      <td>PLOS ONE</td>\n",
       "      <td>2020</td>\n",
       "      <td>published</td>\n",
       "      <td>Public Library of Science (PLoS)</td>\n",
       "      <td>William C. Nelson; Emily B. Graham; Alex R. Cr...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1</td>\n",
       "      <td>e0228165</td>\n",
       "      <td>https://doi.org/10.1371/journal.pone.0228165</td>\n",
       "      <td>Distinct temporal diversity profiles for nitro...</td>\n",
       "      <td>[distinct, temporal, diversity, profiles, for,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10.1016/j.geoderma.2021.115674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soil pore network response to freeze-thaw cycl...</td>\n",
       "      <td>Geoderma</td>\n",
       "      <td>2022</td>\n",
       "      <td>published-print</td>\n",
       "      <td>Elsevier BV</td>\n",
       "      <td>Erin C. Rooney; Vanessa L. Bailey; Kaizad F. P...</td>\n",
       "      <td>411.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115674</td>\n",
       "      <td>https://doi.org/10.1016/j.geoderma.2021.115674</td>\n",
       "      <td>Soil pore network response to freeze-thaw cycl...</td>\n",
       "      <td>[soil, pore, network, response, to, freezethaw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10.1029/2022JG006889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The Impact of Freeze‐Thaw History on Soil Carb...</td>\n",
       "      <td>Journal of Geophysical Research: Biogeosciences</td>\n",
       "      <td>2022</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Geophysical Union (AGU)</td>\n",
       "      <td>Erin C. Rooney; Vanessa L. Bailey; Kaizad F. P...</td>\n",
       "      <td>127.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1029/2022jg006889</td>\n",
       "      <td>The Impact of Freeze‐Thaw History on Soil Carb...</td>\n",
       "      <td>[the, impact, of, freeze‐thaw, history, on, so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10.1002/ppp.2200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Topography and canopy cover influence soil org...</td>\n",
       "      <td>Permafrost and Periglacial Processes</td>\n",
       "      <td>2023</td>\n",
       "      <td>published-print</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>Erin C. Rooney; Vanessa L. Bailey; Kaizad F. P...</td>\n",
       "      <td>34.0</td>\n",
       "      <td>3</td>\n",
       "      <td>331-358</td>\n",
       "      <td>https://doi.org/10.1002/ppp.2200</td>\n",
       "      <td>Topography and canopy cover influence soil org...</td>\n",
       "      <td>[topography, and, canopy, cover, influence, so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10.1038/s41467-023-36515-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Seasonal activities of the phyllosphere microb...</td>\n",
       "      <td>Nature Communications</td>\n",
       "      <td>2023</td>\n",
       "      <td>published</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Adina Howe; Nejc Stopnisek; Shane K. Dooley; F...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1038/s41467-023-36515-y</td>\n",
       "      <td>Seasonal activities of the phyllosphere microb...</td>\n",
       "      <td>[seasonal, activities, of, the, phyllosphere, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>10.1002/lno.11306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Time‐series metatranscriptomes reveal conserve...</td>\n",
       "      <td>Limnology and Oceanography</td>\n",
       "      <td>2020</td>\n",
       "      <td>published-print</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>Alexandra M. Linz; Frank O. Aylward; Stefan Be...</td>\n",
       "      <td>65.0</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1002/lno.11306</td>\n",
       "      <td>Time‐series metatranscriptomes reveal conserve...</td>\n",
       "      <td>[time‐series, metatranscriptomes, reveal, cons...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10.1038/s41597-024-03069-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>An integrated metagenomic, metabolomic and tra...</td>\n",
       "      <td>Scientific Data</td>\n",
       "      <td>2024</td>\n",
       "      <td>published</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Christopher Schadt; Stanton Martin; Alyssa Car...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1038/s41597-024-03069-7</td>\n",
       "      <td>An integrated metagenomic, metabolomic and tra...</td>\n",
       "      <td>[an, integrated, metagenomic, metabolomic, and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>10.1038/s41564-020-00861-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genomic and functional analyses of fungal and ...</td>\n",
       "      <td>Nature Microbiology</td>\n",
       "      <td>2021</td>\n",
       "      <td>published</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Xuefeng Peng; St. Elmo Wilken; Thomas S. Lanki...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4</td>\n",
       "      <td>499-511</td>\n",
       "      <td>https://doi.org/10.1038/s41564-020-00861-0</td>\n",
       "      <td>Genomic and functional analyses of fungal and ...</td>\n",
       "      <td>[genomic, and, functional, analyses, of, funga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>10.1128/MRA.01361-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metatranscriptomic Sequencing of a Cyanobacter...</td>\n",
       "      <td>Microbiology Resource Announcements</td>\n",
       "      <td>2020</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Society for Microbiology</td>\n",
       "      <td>Terrence H. Bell; Ryan V. Trexler; Xin Peng; M...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1128/mra.01361-19</td>\n",
       "      <td>Metatranscriptomic Sequencing of a Cyanobacter...</td>\n",
       "      <td>[metatranscriptomic, sequencing, of, a, cyanob...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10.1128/mra.01080-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metatranscriptomes of two biological soil crus...</td>\n",
       "      <td>Microbiology Resource Announcements</td>\n",
       "      <td>2024</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Society for Microbiology</td>\n",
       "      <td>Thuy M. Nguyen; Nuttapon Pombubpa; Marcel Hunt...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1128/mra.01080-23</td>\n",
       "      <td>Metatranscriptomes of two biological soil crus...</td>\n",
       "      <td>[metatranscriptomes, of, two, biological, soil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>10.1038/s41597-019-0132-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microbial metagenomes and metatranscriptomes d...</td>\n",
       "      <td>Scientific Data</td>\n",
       "      <td>2019</td>\n",
       "      <td>published</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Brent Nowinski; Christa B. Smith; Courtney M. ...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1038/s41597-019-0132-4</td>\n",
       "      <td>Microbial metagenomes and metatranscriptomes d...</td>\n",
       "      <td>[microbial, metagenomes, and, metatranscriptom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>10.1038/s41597-024-04013-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microbial Metagenomes Across a Complete Phytop...</td>\n",
       "      <td>Scientific Data</td>\n",
       "      <td>2024</td>\n",
       "      <td>published</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Brook L. Nunn; Emma Timmins-Schiffman; Miranda...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1038/s41597-024-04013-5</td>\n",
       "      <td>Microbial Metagenomes Across a Complete Phytop...</td>\n",
       "      <td>[microbial, metagenomes, across, a, complete, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>10.1186/s40168-020-00889-8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Influence of the polar light cycle on seasonal...</td>\n",
       "      <td>Microbiome</td>\n",
       "      <td>2020</td>\n",
       "      <td>published-print</td>\n",
       "      <td>Springer Science and Business Media LLC</td>\n",
       "      <td>Pratibha Panwar; Michelle A. Allen; Timothy J....</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1186/s40168-020-00889-8</td>\n",
       "      <td>Influence of the polar light cycle on seasonal...</td>\n",
       "      <td>[influence, of, the, polar, light, cycle, on, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>10.1093/nar/gkab990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The National Microbiome Data Collaborative Dat...</td>\n",
       "      <td>Nucleic Acids Research</td>\n",
       "      <td>2022</td>\n",
       "      <td>published</td>\n",
       "      <td>Oxford University Press (OUP)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkab990</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>10.1126/sciadv.adg7888</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Reproducible growth of&lt;i&gt;Brachypodium&lt;/i&gt;in Ec...</td>\n",
       "      <td>Science Advances</td>\n",
       "      <td>2024</td>\n",
       "      <td>published-print</td>\n",
       "      <td>American Association for the Advancement of Sc...</td>\n",
       "      <td>Vlastimil Novak; Peter F. Andeer; Benjamin P. ...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1126/sciadv.adg7888</td>\n",
       "      <td>Reproducible growth of&lt;i&gt;Brachypodium&lt;/i&gt;in Ec...</td>\n",
       "      <td>[reproducible, growth, ofibrachypodiumiin, eco...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>10.1371/journal.pone.0228165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Distinct temporal diversity profiles for nitro...</td>\n",
       "      <td>PLOS ONE</td>\n",
       "      <td>2020</td>\n",
       "      <td>published</td>\n",
       "      <td>Public Library of Science (PLoS)</td>\n",
       "      <td>William C. Nelson; Emily B. Graham; Alex R. Cr...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1</td>\n",
       "      <td>e0228165</td>\n",
       "      <td>https://doi.org/10.1371/journal.pone.0228165</td>\n",
       "      <td>Distinct temporal diversity profiles for nitro...</td>\n",
       "      <td>[distinct, temporal, diversity, profiles, for,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>10.3390/microorganisms9020357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soil Bacterial Diversity Is Positively Correla...</td>\n",
       "      <td>Microorganisms</td>\n",
       "      <td>2021</td>\n",
       "      <td>published</td>\n",
       "      <td>MDPI AG</td>\n",
       "      <td>Akane Chiba; Yoshitaka Uchida; Susanne Kublik;...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2</td>\n",
       "      <td>357</td>\n",
       "      <td>https://doi.org/10.3390/microorganisms9020357</td>\n",
       "      <td>Soil Bacterial Diversity Is Positively Correla...</td>\n",
       "      <td>[soil, bacterial, diversity, is, positively, c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>10.3897/tdwgproceedings.1.20637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Using MIxS: An Implementation Report from Two ...</td>\n",
       "      <td>Proceedings of TDWG</td>\n",
       "      <td>2017</td>\n",
       "      <td>published</td>\n",
       "      <td>Pensoft Publishers</td>\n",
       "      <td>Joel Sachs; Luke Thompson; Nazir El-Kayssi; Sa...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>e20637</td>\n",
       "      <td>https://doi.org/10.3897/tdwgproceedings.1.20637</td>\n",
       "      <td>Using MIxS: An Implementation Report from Two ...</td>\n",
       "      <td>[using, mixs, an, implementation, report, from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>10.5194/acp-23-15783-2023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Active thermokarst regions contain rich  sourc...</td>\n",
       "      <td>Atmospheric Chemistry and Physics</td>\n",
       "      <td>2023</td>\n",
       "      <td>published</td>\n",
       "      <td>Copernicus GmbH</td>\n",
       "      <td>Kevin R. Barry; Thomas C. J. Hill; Marina Niet...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>24</td>\n",
       "      <td>15783-15793</td>\n",
       "      <td>https://doi.org/10.5194/acp-23-15783-2023</td>\n",
       "      <td>Active thermokarst regions contain rich  sourc...</td>\n",
       "      <td>[active, thermokarst, regions, contain, rich, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:35.216172Z",
     "start_time": "2025-02-11T21:07:35.187537Z"
    }
   },
   "cell_type": "code",
   "source": "df_metadata",
   "id": "8c060a232f2907a7",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                             Filename  \\\n",
       "0                               acp-23-15783-2023.pdf   \n",
       "1   bell-et-al-2020-metatranscriptomic-sequencing-...   \n",
       "2   Soil_Bacterial_Diversity_Is_Positively_Correla...   \n",
       "3                              BISS_article_20637.pdf   \n",
       "4                              s40168-020-00889-8.pdf   \n",
       "5                                  sciadv.adg7888.pdf   \n",
       "6                              s41597-024-04013-5.pdf   \n",
       "7                              s41564-020-00861-0.pdf   \n",
       "8                          41564_2019_Article_449.pdf   \n",
       "9                                         gkab990.pdf   \n",
       "10  blair-et-al-2018-exploration-of-the-biosynthet...   \n",
       "11                              s41597-019-0132-4.pdf   \n",
       "12  Limnology   Oceanography - 2019 - Linz - Time‐...   \n",
       "13                                           file.pdf   \n",
       "14                     microorganisms-09-00357-v2.pdf   \n",
       "15  Environmental Microbiology - 2022 - Olmsted - ...   \n",
       "16                  1-s2.0-S0016706121007540-main.pdf   \n",
       "17  Molecular Ecology - 2023 - He - Diversity  dis...   \n",
       "18  JGR Biogeosciences - 2022 - Rooney - The Impac...   \n",
       "19                             s41564-022-01266-x.pdf   \n",
       "20                             s41467-023-36515-y.pdf   \n",
       "21  nguyen-et-al-2024-metatranscriptomes-of-two-bi...   \n",
       "22                       2022.12.12.520098v2.full.pdf   \n",
       "23                             s41597-024-03069-7.pdf   \n",
       "24  alteio-et-al-2020-complementary-metagenomic-ap...   \n",
       "25  New Phytologist - 2018 - Sasse - Multilab EcoF...   \n",
       "\n",
       "                               DOI    Year    Volume          Pages  \\\n",
       "0                              NaN     NaN       NaN            NaN   \n",
       "1             10.1128/mra.01361-19  2020.0       9.0             10   \n",
       "2                              NaN     NaN       NaN            NaN   \n",
       "3                              NaN     NaN       NaN            NaN   \n",
       "4       10.1186/s40168-020-00889-8     NaN       NaN            NaN   \n",
       "5                              NaN  2024.0      10.0       eadg7888   \n",
       "6       10.1038/s41597-024-04013-5     NaN       NaN            NaN   \n",
       "7       10.1038/s41564-020-00861-0     NaN       NaN            NaN   \n",
       "8        10.1038/s41564-019-0449-y     NaN       NaN            NaN   \n",
       "9                              NaN     NaN       NaN            NaN   \n",
       "10       10.1128/msystems.00045-18  2018.0       3.0             10   \n",
       "11       10.1038/s41597-019-0132-4     NaN       NaN            NaN   \n",
       "12                             NaN  2020.0      65.0      S101-S112   \n",
       "13                             NaN     NaN       NaN            NaN   \n",
       "14                             NaN     NaN       NaN            NaN   \n",
       "15                             NaN  2023.0      25.0        705-720   \n",
       "16  10.1016/j.geoderma.2021.115674  2021.0  115674.0            NaN   \n",
       "17                             NaN  2023.0      32.0      2798-2817   \n",
       "18                             NaN  2022.0     127.0  e2022JG006889   \n",
       "19      10.1038/s41564-022-01266-x     NaN       NaN            NaN   \n",
       "20      10.1038/s41467-023-36515-y     NaN       NaN            NaN   \n",
       "21                             NaN  2024.0      13.0      e01080-23   \n",
       "22                             NaN     NaN       NaN            NaN   \n",
       "23      10.1038/s41597-024-03069-7     NaN       NaN            NaN   \n",
       "24       10.1128/msystems.00768-19  2020.0       5.0             10   \n",
       "25                             NaN  2019.0     222.0      1149-1160   \n",
       "\n",
       "                                                Title  \\\n",
       "0                                                 NaN   \n",
       "1   Metatranscriptomic Sequencing of a Cyanobacter...   \n",
       "2        Microsoft Word - microorganisms-1106083.docx   \n",
       "3                                                 NaN   \n",
       "4   Influence of the polar light cycle on seasonal...   \n",
       "5   Reproducible growth of Brachypodium in EcoFAB ...   \n",
       "6   Microbial Metagenomes Across a Complete Phytop...   \n",
       "7   Genomic and functional analyses of fungal and ...   \n",
       "8   Mediterranean grassland soil C–N compound turn...   \n",
       "9                                                 NaN   \n",
       "10  Exploration of the Biosynthetic Potential of t...   \n",
       "11  Microbial metagenomes and metatranscriptomes d...   \n",
       "12  Time‐series metatranscriptomes reveal conserve...   \n",
       "13  Distinct temporal diversity profiles for nitro...   \n",
       "14  Soil Bacterial Diversity Is Positively Correla...   \n",
       "15  Environmental predictors of electroactive bact...   \n",
       "16  Soil pore network response to freeze-thaw cycl...   \n",
       "17  Diversity, distribution, and expression of ops...   \n",
       "18  The Impact of Freeze‐Thaw History on Soil Carb...   \n",
       "19  Standardized multi-omics of Earth’s microbiome...   \n",
       "20  Seasonal activities of the phyllosphere microb...   \n",
       "21  Metatranscriptomes of two biological soil crus...   \n",
       "22                                           12861919   \n",
       "23  An integrated metagenomic, metabolomic and tra...   \n",
       "24  Complementary Metagenomic Approaches Improve R...   \n",
       "25  Multilab EcoFAB study shows highly reproducibl...   \n",
       "\n",
       "                                               Author  \\\n",
       "0                                                 NaN   \n",
       "1   Terrence H. Bell, Ryan V. Trexler, Xin Peng, M...   \n",
       "2                                               Admin   \n",
       "3                                                 NaN   \n",
       "4                                     Pratibha Panwar   \n",
       "5                                                 NaN   \n",
       "6                                       Brook L. Nunn   \n",
       "7                                        Xuefeng Peng   \n",
       "8                                     Spencer Diamond   \n",
       "9                                                 NaN   \n",
       "10  Patricia M. Blair, Miriam L. Land, Marek J. Pi...   \n",
       "11                                     Brent Nowinski   \n",
       "12                                                NaN   \n",
       "13  William C. Nelson, Emily B. Graham, Alex R. Cr...   \n",
       "14  Akane Chiba, Yoshitaka Uchida, Susanne Kublik,...   \n",
       "15                                                NaN   \n",
       "16                                     Erin C. Rooney   \n",
       "17                                                NaN   \n",
       "18                                                NaN   \n",
       "19                                  Justin P. Shaffer   \n",
       "20                                         Adina Howe   \n",
       "21  Thuy M. Nguyen, Nuttapon Pombubpa, Marcel Hunt...   \n",
       "22                                    Mans, Douglas M   \n",
       "23                                 Christopher Schadt   \n",
       "24  L. V. Alteio, F. Schulz, R. Seshadri, N. Vargh...   \n",
       "25                                                NaN   \n",
       "\n",
       "                                              Subject  \\\n",
       "0                                                 NaN   \n",
       "1   Microbiol Resour Announc 2020.9:10.1128/mra.01...   \n",
       "2                                                 NaN   \n",
       "3                                         description   \n",
       "4    Microbiome, 2020, doi:10.1186/s40168-020-00889-8   \n",
       "5                          Sci. Adv. 2024.10:eadg7888   \n",
       "6     Scientific Data, doi:10.1038/s41597-024-04013-5   \n",
       "7   Nature Microbiology, doi:10.1038/s41564-020-00...   \n",
       "8   Nature Microbiology, doi:10.1038/s41564-019-04...   \n",
       "9                                                 NaN   \n",
       "10          mSystems 2018.3:10.1128/msystems.00045-18   \n",
       "11     Scientific Data, doi:10.1038/s41597-019-0132-4   \n",
       "12         Limnology & Oceanography 2020.65:S101-S112   \n",
       "13                                                NaN   \n",
       "14  This study aimed to investigate the effects of...   \n",
       "15         Environmental Microbiology 2023.25:705-720   \n",
       "16  Geoderma, 411 (2022) 115674. doi:10.1016/j.geo...   \n",
       "17                Molecular Ecology 2023.32:2798-2817   \n",
       "18          JGR Biogeosciences 2022.127:e2022JG006889   \n",
       "19  Nature Microbiology, doi:10.1038/s41564-022-01...   \n",
       "20  Nature Communications, doi:10.1038/s41467-023-...   \n",
       "21         Microbiol Resour Announc 2024.13:e01080-23   \n",
       "22                                                NaN   \n",
       "23    Scientific Data, doi:10.1038/s41597-024-03069-7   \n",
       "24          mSystems 2020.5:10.1128/msystems.00768-19   \n",
       "25                 New Phytologist 2019.222:1149-1160   \n",
       "\n",
       "                                             Keywords Creation Date  \\\n",
       "0                                                 NaN    2023-12-22   \n",
       "1                                                 NaN    2019-12-10   \n",
       "2                                                 NaN    2021-02-11   \n",
       "3                                            keywords           NaN   \n",
       "4   Antarctic microbiology,Polar light cycle,Metag...    2020-11-14   \n",
       "5                                                 NaN    2023-12-29   \n",
       "6                                                 NaN    2025-01-15   \n",
       "7                                                 NaN    2021-03-24   \n",
       "8                                                 NaN    2019-07-17   \n",
       "9                                                 NaN    2021-12-27   \n",
       "10  biosynthetic gene clusters, natural product bi...    2018-09-24   \n",
       "11                                                NaN    2019-07-12   \n",
       "12                                                NaN    2020-01-20   \n",
       "13                                                NaN    2020-01-21   \n",
       "14  soil bacterial diversity; maize litter decompo...    2021-02-24   \n",
       "15                                                NaN    2023-02-23   \n",
       "16  Pore morphology,Pore connectivity,Freeze thaw,...    2022-01-20   \n",
       "17                                                NaN    2023-05-18   \n",
       "18                                                NaN    2022-05-19   \n",
       "19                                                NaN    2022-11-30   \n",
       "20                                                NaN    2023-02-23   \n",
       "21  biocrust, RNA, transcriptome, wetting, desert,...    2024-02-09   \n",
       "22                                                NaN    2023-03-14   \n",
       "23                                                NaN    2024-04-05   \n",
       "24  flow cytometry, metagenomics, microbial ecolog...    2020-03-05   \n",
       "25                                                NaN    2019-03-08   \n",
       "\n",
       "   Modification Date                                           Producer  \\\n",
       "0         2023-12-22                                     pdfTeX-1.40.19   \n",
       "1         2025-02-10  Adobe LiveCycle PDF Generator; modified using ...   \n",
       "2         2021-02-11                   Acrobat Distiller 21.0 (Windows)   \n",
       "3                NaN           WeasyPrint 0.31 (http://weasyprint.org/)   \n",
       "4         2020-11-15  Acrobat Distiller 10.0.0 (Windows); modified u...   \n",
       "5         2024-08-22  Adobe PDF Library 15.0; modified using iText 4...   \n",
       "6         2025-01-21  iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER S...   \n",
       "7         2021-03-24                                                NaN   \n",
       "8         2019-07-17                                                NaN   \n",
       "9         2024-08-21  Acrobat Distiller 21.0 (Windows); modified usi...   \n",
       "10        2025-02-10  Adobe LiveCycle PDF Generator; modified using ...   \n",
       "11        2019-07-15  iText® 5.3.5 ©2000-2012 1T3XT BVBA (AGPL-version)   \n",
       "12        2025-02-10  Acrobat Distiller 8.1.0 (Windows); modified us...   \n",
       "13        2020-01-21                       PDFlib+PDI 9.0.6 (C++/Win64)   \n",
       "14        2021-02-24                                     pdfTeX-1.40.18   \n",
       "15        2025-02-10  Acrobat Distiller 11.0 (Windows); modified usi...   \n",
       "16        2022-01-20                  Acrobat Distiller 8.1.0 (Windows)   \n",
       "17        2025-02-10  Adobe PDF Library 15.0; modified using iText 4...   \n",
       "18        2025-02-10  Adobe PDF Library 15.0; modified using iText 4...   \n",
       "19        2022-11-30                                                NaN   \n",
       "20        2023-02-23  iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER S...   \n",
       "21        2025-02-10  Antenna House PDF Output Library 7.3.1866; mod...   \n",
       "22        2025-02-10                  Microsoft® Word for Microsoft 365   \n",
       "23        2024-04-05  iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER S...   \n",
       "24        2025-02-10  Adobe LiveCycle PDF Generator; modified using ...   \n",
       "25        2024-07-25  Acrobat Distiller 10.0.0 (Windows); modified u...   \n",
       "\n",
       "                                              Creator  \\\n",
       "0   copernicus.cls Version 2023/11/02 10.1.11 Cope...   \n",
       "1                                                 XPP   \n",
       "2                          PScript5.dll Version 5.2.2   \n",
       "3                                                 NaN   \n",
       "4   Arbortext Advanced Print Publisher 9.1.440/W U...   \n",
       "5                       Adobe InDesign 16.2 (Windows)   \n",
       "6                                            Springer   \n",
       "7                                            Springer   \n",
       "8                                            Springer   \n",
       "9    XMLPublish v.3.0 Copyright 1991-2017 Aptara Inc.   \n",
       "10                                                XPP   \n",
       "11                                           Springer   \n",
       "12  Arbortext Advanced Print Publisher 9.1.520/W U...   \n",
       "13  Arbortext Advanced Print Publisher 11.0.3521/W...   \n",
       "14                        LaTeX with hyperref package   \n",
       "15  Arbortext Advanced Print Publisher 9.1.520/W U...   \n",
       "16                                           Elsevier   \n",
       "17                      Adobe InDesign 15.1 (Windows)   \n",
       "18                      Adobe InDesign 15.1 (Windows)   \n",
       "19                                           Springer   \n",
       "20                                           Springer   \n",
       "21  Antenna XSL House Formatter V7.3 MR2 Linux : 7...   \n",
       "22                        Appligent AppendPDF Pro 5.5   \n",
       "23                                           Springer   \n",
       "24                                                XPP   \n",
       "25  Arbortext Advanced Print Publisher 9.1.531/W U...   \n",
       "\n",
       "                                         title_author  \\\n",
       "0                                                 NaN   \n",
       "1   Metatranscriptomic Sequencing of a Cyanobacter...   \n",
       "2   Microsoft Word - microorganisms-1106083.docx A...   \n",
       "3                                                 NaN   \n",
       "4   Influence of the polar light cycle on seasonal...   \n",
       "5                                                 NaN   \n",
       "6   Microbial Metagenomes Across a Complete Phytop...   \n",
       "7   Genomic and functional analyses of fungal and ...   \n",
       "8   Mediterranean grassland soil C–N compound turn...   \n",
       "9                                                 NaN   \n",
       "10  Exploration of the Biosynthetic Potential of t...   \n",
       "11  Microbial metagenomes and metatranscriptomes d...   \n",
       "12                                                NaN   \n",
       "13  Distinct temporal diversity profiles for nitro...   \n",
       "14  Soil Bacterial Diversity Is Positively Correla...   \n",
       "15                                                NaN   \n",
       "16  Soil pore network response to freeze-thaw cycl...   \n",
       "17                                                NaN   \n",
       "18                                                NaN   \n",
       "19  Standardized multi-omics of Earth’s microbiome...   \n",
       "20  Seasonal activities of the phyllosphere microb...   \n",
       "21  Metatranscriptomes of two biological soil crus...   \n",
       "22                           12861919 Mans, Douglas M   \n",
       "23  An integrated metagenomic, metabolomic and tra...   \n",
       "24  Complementary Metagenomic Approaches Improve R...   \n",
       "25                                                NaN   \n",
       "\n",
       "                                  title_author_tokens  \n",
       "0                                                      \n",
       "1   [metatranscriptomic, sequencing, of, a, cyanob...  \n",
       "2   [microsoft, word, microorganisms1106083docx, a...  \n",
       "3                                                      \n",
       "4   [influence, of, the, polar, light, cycle, on, ...  \n",
       "5                                                      \n",
       "6   [microbial, metagenomes, across, a, complete, ...  \n",
       "7   [genomic, and, functional, analyses, of, funga...  \n",
       "8   [mediterranean, grassland, soil, c–n, compound...  \n",
       "9                                                      \n",
       "10  [exploration, of, the, biosynthetic, potential...  \n",
       "11  [microbial, metagenomes, and, metatranscriptom...  \n",
       "12                                                     \n",
       "13  [distinct, temporal, diversity, profiles, for,...  \n",
       "14  [soil, bacterial, diversity, is, positively, c...  \n",
       "15                                                     \n",
       "16  [soil, pore, network, response, to, freezethaw...  \n",
       "17                                                     \n",
       "18                                                     \n",
       "19  [standardized, multiomics, of, earth’s, microb...  \n",
       "20  [seasonal, activities, of, the, phyllosphere, ...  \n",
       "21  [metatranscriptomes, of, two, biological, soil...  \n",
       "22                       [12861919, mans, douglas, m]  \n",
       "23  [an, integrated, metagenomic, metabolomic, and...  \n",
       "24  [complementary, metagenomic, approaches, impro...  \n",
       "25                                                     "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Filename</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Year</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Pages</th>\n",
       "      <th>Title</th>\n",
       "      <th>Author</th>\n",
       "      <th>Subject</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Creation Date</th>\n",
       "      <th>Modification Date</th>\n",
       "      <th>Producer</th>\n",
       "      <th>Creator</th>\n",
       "      <th>title_author</th>\n",
       "      <th>title_author_tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acp-23-15783-2023.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-12-22</td>\n",
       "      <td>2023-12-22</td>\n",
       "      <td>pdfTeX-1.40.19</td>\n",
       "      <td>copernicus.cls Version 2023/11/02 10.1.11 Cope...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bell-et-al-2020-metatranscriptomic-sequencing-...</td>\n",
       "      <td>10.1128/mra.01361-19</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10</td>\n",
       "      <td>Metatranscriptomic Sequencing of a Cyanobacter...</td>\n",
       "      <td>Terrence H. Bell, Ryan V. Trexler, Xin Peng, M...</td>\n",
       "      <td>Microbiol Resour Announc 2020.9:10.1128/mra.01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-12-10</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Adobe LiveCycle PDF Generator; modified using ...</td>\n",
       "      <td>XPP</td>\n",
       "      <td>Metatranscriptomic Sequencing of a Cyanobacter...</td>\n",
       "      <td>[metatranscriptomic, sequencing, of, a, cyanob...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Soil_Bacterial_Diversity_Is_Positively_Correla...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microsoft Word - microorganisms-1106083.docx</td>\n",
       "      <td>Admin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-02-11</td>\n",
       "      <td>2021-02-11</td>\n",
       "      <td>Acrobat Distiller 21.0 (Windows)</td>\n",
       "      <td>PScript5.dll Version 5.2.2</td>\n",
       "      <td>Microsoft Word - microorganisms-1106083.docx A...</td>\n",
       "      <td>[microsoft, word, microorganisms1106083docx, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BISS_article_20637.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>description</td>\n",
       "      <td>keywords</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WeasyPrint 0.31 (http://weasyprint.org/)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s40168-020-00889-8.pdf</td>\n",
       "      <td>10.1186/s40168-020-00889-8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Influence of the polar light cycle on seasonal...</td>\n",
       "      <td>Pratibha Panwar</td>\n",
       "      <td>Microbiome, 2020, doi:10.1186/s40168-020-00889-8</td>\n",
       "      <td>Antarctic microbiology,Polar light cycle,Metag...</td>\n",
       "      <td>2020-11-14</td>\n",
       "      <td>2020-11-15</td>\n",
       "      <td>Acrobat Distiller 10.0.0 (Windows); modified u...</td>\n",
       "      <td>Arbortext Advanced Print Publisher 9.1.440/W U...</td>\n",
       "      <td>Influence of the polar light cycle on seasonal...</td>\n",
       "      <td>[influence, of, the, polar, light, cycle, on, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sciadv.adg7888.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>eadg7888</td>\n",
       "      <td>Reproducible growth of Brachypodium in EcoFAB ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sci. Adv. 2024.10:eadg7888</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-12-29</td>\n",
       "      <td>2024-08-22</td>\n",
       "      <td>Adobe PDF Library 15.0; modified using iText 4...</td>\n",
       "      <td>Adobe InDesign 16.2 (Windows)</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>s41597-024-04013-5.pdf</td>\n",
       "      <td>10.1038/s41597-024-04013-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microbial Metagenomes Across a Complete Phytop...</td>\n",
       "      <td>Brook L. Nunn</td>\n",
       "      <td>Scientific Data, doi:10.1038/s41597-024-04013-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-15</td>\n",
       "      <td>2025-01-21</td>\n",
       "      <td>iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER S...</td>\n",
       "      <td>Springer</td>\n",
       "      <td>Microbial Metagenomes Across a Complete Phytop...</td>\n",
       "      <td>[microbial, metagenomes, across, a, complete, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>s41564-020-00861-0.pdf</td>\n",
       "      <td>10.1038/s41564-020-00861-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genomic and functional analyses of fungal and ...</td>\n",
       "      <td>Xuefeng Peng</td>\n",
       "      <td>Nature Microbiology, doi:10.1038/s41564-020-00...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Springer</td>\n",
       "      <td>Genomic and functional analyses of fungal and ...</td>\n",
       "      <td>[genomic, and, functional, analyses, of, funga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>41564_2019_Article_449.pdf</td>\n",
       "      <td>10.1038/s41564-019-0449-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mediterranean grassland soil C–N compound turn...</td>\n",
       "      <td>Spencer Diamond</td>\n",
       "      <td>Nature Microbiology, doi:10.1038/s41564-019-04...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-07-17</td>\n",
       "      <td>2019-07-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Springer</td>\n",
       "      <td>Mediterranean grassland soil C–N compound turn...</td>\n",
       "      <td>[mediterranean, grassland, soil, c–n, compound...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gkab990.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-12-27</td>\n",
       "      <td>2024-08-21</td>\n",
       "      <td>Acrobat Distiller 21.0 (Windows); modified usi...</td>\n",
       "      <td>XMLPublish v.3.0 Copyright 1991-2017 Aptara Inc.</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>blair-et-al-2018-exploration-of-the-biosynthet...</td>\n",
       "      <td>10.1128/msystems.00045-18</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10</td>\n",
       "      <td>Exploration of the Biosynthetic Potential of t...</td>\n",
       "      <td>Patricia M. Blair, Miriam L. Land, Marek J. Pi...</td>\n",
       "      <td>mSystems 2018.3:10.1128/msystems.00045-18</td>\n",
       "      <td>biosynthetic gene clusters, natural product bi...</td>\n",
       "      <td>2018-09-24</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Adobe LiveCycle PDF Generator; modified using ...</td>\n",
       "      <td>XPP</td>\n",
       "      <td>Exploration of the Biosynthetic Potential of t...</td>\n",
       "      <td>[exploration, of, the, biosynthetic, potential...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>s41597-019-0132-4.pdf</td>\n",
       "      <td>10.1038/s41597-019-0132-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microbial metagenomes and metatranscriptomes d...</td>\n",
       "      <td>Brent Nowinski</td>\n",
       "      <td>Scientific Data, doi:10.1038/s41597-019-0132-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-07-12</td>\n",
       "      <td>2019-07-15</td>\n",
       "      <td>iText® 5.3.5 ©2000-2012 1T3XT BVBA (AGPL-version)</td>\n",
       "      <td>Springer</td>\n",
       "      <td>Microbial metagenomes and metatranscriptomes d...</td>\n",
       "      <td>[microbial, metagenomes, and, metatranscriptom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Limnology   Oceanography - 2019 - Linz - Time‐...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>S101-S112</td>\n",
       "      <td>Time‐series metatranscriptomes reveal conserve...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Limnology &amp; Oceanography 2020.65:S101-S112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-20</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Acrobat Distiller 8.1.0 (Windows); modified us...</td>\n",
       "      <td>Arbortext Advanced Print Publisher 9.1.520/W U...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>file.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Distinct temporal diversity profiles for nitro...</td>\n",
       "      <td>William C. Nelson, Emily B. Graham, Alex R. Cr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-21</td>\n",
       "      <td>2020-01-21</td>\n",
       "      <td>PDFlib+PDI 9.0.6 (C++/Win64)</td>\n",
       "      <td>Arbortext Advanced Print Publisher 11.0.3521/W...</td>\n",
       "      <td>Distinct temporal diversity profiles for nitro...</td>\n",
       "      <td>[distinct, temporal, diversity, profiles, for,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>microorganisms-09-00357-v2.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soil Bacterial Diversity Is Positively Correla...</td>\n",
       "      <td>Akane Chiba, Yoshitaka Uchida, Susanne Kublik,...</td>\n",
       "      <td>This study aimed to investigate the effects of...</td>\n",
       "      <td>soil bacterial diversity; maize litter decompo...</td>\n",
       "      <td>2021-02-24</td>\n",
       "      <td>2021-02-24</td>\n",
       "      <td>pdfTeX-1.40.18</td>\n",
       "      <td>LaTeX with hyperref package</td>\n",
       "      <td>Soil Bacterial Diversity Is Positively Correla...</td>\n",
       "      <td>[soil, bacterial, diversity, is, positively, c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Environmental Microbiology - 2022 - Olmsted - ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>705-720</td>\n",
       "      <td>Environmental predictors of electroactive bact...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Environmental Microbiology 2023.25:705-720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-23</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Acrobat Distiller 11.0 (Windows); modified usi...</td>\n",
       "      <td>Arbortext Advanced Print Publisher 9.1.520/W U...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1-s2.0-S0016706121007540-main.pdf</td>\n",
       "      <td>10.1016/j.geoderma.2021.115674</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>115674.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soil pore network response to freeze-thaw cycl...</td>\n",
       "      <td>Erin C. Rooney</td>\n",
       "      <td>Geoderma, 411 (2022) 115674. doi:10.1016/j.geo...</td>\n",
       "      <td>Pore morphology,Pore connectivity,Freeze thaw,...</td>\n",
       "      <td>2022-01-20</td>\n",
       "      <td>2022-01-20</td>\n",
       "      <td>Acrobat Distiller 8.1.0 (Windows)</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Soil pore network response to freeze-thaw cycl...</td>\n",
       "      <td>[soil, pore, network, response, to, freezethaw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Molecular Ecology - 2023 - He - Diversity  dis...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>2798-2817</td>\n",
       "      <td>Diversity, distribution, and expression of ops...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Molecular Ecology 2023.32:2798-2817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-05-18</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Adobe PDF Library 15.0; modified using iText 4...</td>\n",
       "      <td>Adobe InDesign 15.1 (Windows)</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>JGR Biogeosciences - 2022 - Rooney - The Impac...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>e2022JG006889</td>\n",
       "      <td>The Impact of Freeze‐Thaw History on Soil Carb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JGR Biogeosciences 2022.127:e2022JG006889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-05-19</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Adobe PDF Library 15.0; modified using iText 4...</td>\n",
       "      <td>Adobe InDesign 15.1 (Windows)</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>s41564-022-01266-x.pdf</td>\n",
       "      <td>10.1038/s41564-022-01266-x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Standardized multi-omics of Earth’s microbiome...</td>\n",
       "      <td>Justin P. Shaffer</td>\n",
       "      <td>Nature Microbiology, doi:10.1038/s41564-022-01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-11-30</td>\n",
       "      <td>2022-11-30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Springer</td>\n",
       "      <td>Standardized multi-omics of Earth’s microbiome...</td>\n",
       "      <td>[standardized, multiomics, of, earth’s, microb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>s41467-023-36515-y.pdf</td>\n",
       "      <td>10.1038/s41467-023-36515-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Seasonal activities of the phyllosphere microb...</td>\n",
       "      <td>Adina Howe</td>\n",
       "      <td>Nature Communications, doi:10.1038/s41467-023-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-23</td>\n",
       "      <td>2023-02-23</td>\n",
       "      <td>iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER S...</td>\n",
       "      <td>Springer</td>\n",
       "      <td>Seasonal activities of the phyllosphere microb...</td>\n",
       "      <td>[seasonal, activities, of, the, phyllosphere, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nguyen-et-al-2024-metatranscriptomes-of-two-bi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>e01080-23</td>\n",
       "      <td>Metatranscriptomes of two biological soil crus...</td>\n",
       "      <td>Thuy M. Nguyen, Nuttapon Pombubpa, Marcel Hunt...</td>\n",
       "      <td>Microbiol Resour Announc 2024.13:e01080-23</td>\n",
       "      <td>biocrust, RNA, transcriptome, wetting, desert,...</td>\n",
       "      <td>2024-02-09</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Antenna House PDF Output Library 7.3.1866; mod...</td>\n",
       "      <td>Antenna XSL House Formatter V7.3 MR2 Linux : 7...</td>\n",
       "      <td>Metatranscriptomes of two biological soil crus...</td>\n",
       "      <td>[metatranscriptomes, of, two, biological, soil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2022.12.12.520098v2.full.pdf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12861919</td>\n",
       "      <td>Mans, Douglas M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-03-14</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Microsoft® Word for Microsoft 365</td>\n",
       "      <td>Appligent AppendPDF Pro 5.5</td>\n",
       "      <td>12861919 Mans, Douglas M</td>\n",
       "      <td>[12861919, mans, douglas, m]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>s41597-024-03069-7.pdf</td>\n",
       "      <td>10.1038/s41597-024-03069-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>An integrated metagenomic, metabolomic and tra...</td>\n",
       "      <td>Christopher Schadt</td>\n",
       "      <td>Scientific Data, doi:10.1038/s41597-024-03069-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024-04-05</td>\n",
       "      <td>2024-04-05</td>\n",
       "      <td>iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER S...</td>\n",
       "      <td>Springer</td>\n",
       "      <td>An integrated metagenomic, metabolomic and tra...</td>\n",
       "      <td>[an, integrated, metagenomic, metabolomic, and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>alteio-et-al-2020-complementary-metagenomic-ap...</td>\n",
       "      <td>10.1128/msystems.00768-19</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10</td>\n",
       "      <td>Complementary Metagenomic Approaches Improve R...</td>\n",
       "      <td>L. V. Alteio, F. Schulz, R. Seshadri, N. Vargh...</td>\n",
       "      <td>mSystems 2020.5:10.1128/msystems.00768-19</td>\n",
       "      <td>flow cytometry, metagenomics, microbial ecolog...</td>\n",
       "      <td>2020-03-05</td>\n",
       "      <td>2025-02-10</td>\n",
       "      <td>Adobe LiveCycle PDF Generator; modified using ...</td>\n",
       "      <td>XPP</td>\n",
       "      <td>Complementary Metagenomic Approaches Improve R...</td>\n",
       "      <td>[complementary, metagenomic, approaches, impro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>New Phytologist - 2018 - Sasse - Multilab EcoF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>1149-1160</td>\n",
       "      <td>Multilab EcoFAB study shows highly reproducibl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New Phytologist 2019.222:1149-1160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-03-08</td>\n",
       "      <td>2024-07-25</td>\n",
       "      <td>Acrobat Distiller 10.0.0 (Windows); modified u...</td>\n",
       "      <td>Arbortext Advanced Print Publisher 9.1.531/W U...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:40.955471Z",
     "start_time": "2025-02-11T21:07:40.952074Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Create a TF-IDF vectorizer to convert the tokenized text into numerical vectors\n",
    "vectorizer = TfidfVectorizer()\n"
   ],
   "id": "e4669015f0b3da7e",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:43.568997Z",
     "start_time": "2025-02-11T21:07:43.555708Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fit the vectorizer on the combined title_author_tokens from both DataFrames\n",
    "vectorizer.fit(df_citations['title_author_tokens'].apply(lambda x: ' '.join(x)).tolist() +\n",
    "              df_metadata['title_author_tokens'].apply(lambda x: ' '.join(x)).tolist())\n"
   ],
   "id": "e1b7a666186b1420",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TfidfVectorizer()"
      ],
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>TfidfVectorizer()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>TfidfVectorizer</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html\">?<span>Documentation for TfidfVectorizer</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>TfidfVectorizer()</pre></div> </div></div></div></div>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:55.543685Z",
     "start_time": "2025-02-11T21:07:55.533661Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Transform the title_author_tokens into numerical vectors\n",
    "citations_vectors = vectorizer.transform(df_citations['title_author_tokens'].apply(lambda x: ' '.join(x)))\n",
    "metadata_vectors = vectorizer.transform(df_metadata['title_author_tokens'].apply(lambda x: ' '.join(x)))\n"
   ],
   "id": "f339bdef3d6970",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:07:59.758217Z",
     "start_time": "2025-02-11T21:07:59.751185Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Calculate the cosine similarity matrix\n",
    "cosine_sim_matrix = cosine_similarity(citations_vectors, metadata_vectors)\n"
   ],
   "id": "63b4389eaaec9651",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:08:02.818735Z",
     "start_time": "2025-02-11T21:08:02.815818Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Find the maximum cosine similarity score for each entry in doi_metadata.tsv\n",
    "max_sim_scores = cosine_sim_matrix.max(axis=1)\n"
   ],
   "id": "c376e0b2c5f452a8",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:10:39.013745Z",
     "start_time": "2025-02-11T21:10:39.009607Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Find the maximum cosine similarity score and its index for each entry in doi_metadata.tsv\n",
    "max_sim_scores = cosine_sim_matrix.max(axis=1)\n",
    "max_sim_indices = cosine_sim_matrix.argmax(axis=1)"
   ],
   "id": "81a2dc16c1ac9a0",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:12:13.925135Z",
     "start_time": "2025-02-11T21:12:13.920218Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Create a list of dictionaries to store the results\n",
    "results = []\n",
    "for i, (score, idx) in enumerate(zip(max_sim_scores, max_sim_indices)):\n",
    "    results.append({\n",
    "        'Entry': i + 1,\n",
    "        'DOI': df_citations.loc[i, 'DOI'],\n",
    "        'Filename': df_metadata.loc[idx, 'Filename'],\n",
    "        'Similarity Score': score\n",
    "    })"
   ],
   "id": "8e06fa6fd4ceb550",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:12:30.064435Z",
     "start_time": "2025-02-11T21:12:30.060117Z"
    }
   },
   "cell_type": "code",
   "source": "df = pd.DataFrame(results)",
   "id": "a230ce26f1179934",
   "outputs": [],
   "execution_count": 22
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-11T21:12:34.677928Z",
     "start_time": "2025-02-11T21:12:34.667888Z"
    }
   },
   "cell_type": "code",
   "source": "df",
   "id": "d7365f1c884a32b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    Entry                              DOI  \\\n",
       "0       1        10.1128/mSystems.00045-18   \n",
       "1       2        10.1101/2022.12.12.520098   \n",
       "2       3       10.1038/s41564-022-01266-x   \n",
       "3       4          10.1111/1462-2920.16314   \n",
       "4       5                10.1111/mec.16891   \n",
       "5       6      10.1021/acs.estlett.0c00748   \n",
       "6       7        10.1128/msystems.00768-19   \n",
       "7       8     10.1371/journal.pone.0228165   \n",
       "8       9   10.1016/j.geoderma.2021.115674   \n",
       "9      10             10.1029/2022JG006889   \n",
       "10     11                 10.1002/ppp.2200   \n",
       "11     12       10.1038/s41467-023-36515-y   \n",
       "12     13                10.1002/lno.11306   \n",
       "13     14       10.1038/s41597-024-03069-7   \n",
       "14     15       10.1038/s41564-020-00861-0   \n",
       "15     16             10.1128/MRA.01361-19   \n",
       "16     17             10.1128/mra.01080-23   \n",
       "17     18        10.1038/s41597-019-0132-4   \n",
       "18     19       10.1038/s41597-024-04013-5   \n",
       "19     20       10.1186/s40168-020-00889-8   \n",
       "20     21              10.1093/nar/gkab990   \n",
       "21     22           10.1126/sciadv.adg7888   \n",
       "22     23     10.1371/journal.pone.0228165   \n",
       "23     24    10.3390/microorganisms9020357   \n",
       "24     25  10.3897/tdwgproceedings.1.20637   \n",
       "25     26        10.5194/acp-23-15783-2023   \n",
       "\n",
       "                                             Filename  Similarity Score  \n",
       "0   blair-et-al-2018-exploration-of-the-biosynthet...          0.948377  \n",
       "1                        2022.12.12.520098v2.full.pdf          0.161309  \n",
       "2                              s41564-022-01266-x.pdf          0.187543  \n",
       "3   blair-et-al-2018-exploration-of-the-biosynthet...          0.083410  \n",
       "4                                            file.pdf          0.098154  \n",
       "5   alteio-et-al-2020-complementary-metagenomic-ap...          0.067630  \n",
       "6   alteio-et-al-2020-complementary-metagenomic-ap...          1.000000  \n",
       "7                                            file.pdf          1.000000  \n",
       "8                   1-s2.0-S0016706121007540-main.pdf          0.511187  \n",
       "9                   1-s2.0-S0016706121007540-main.pdf          0.232726  \n",
       "10                  1-s2.0-S0016706121007540-main.pdf          0.187375  \n",
       "11                             s41467-023-36515-y.pdf          0.649401  \n",
       "12                              s41597-019-0132-4.pdf          0.074472  \n",
       "13                             s41597-024-03069-7.pdf          0.435847  \n",
       "14                             s41564-020-00861-0.pdf          0.603828  \n",
       "15  bell-et-al-2020-metatranscriptomic-sequencing-...          1.000000  \n",
       "16  nguyen-et-al-2024-metatranscriptomes-of-two-bi...          1.000000  \n",
       "17  bell-et-al-2020-metatranscriptomic-sequencing-...          0.453781  \n",
       "18                             s41597-024-04013-5.pdf          0.515896  \n",
       "19                             s40168-020-00889-8.pdf          0.548277  \n",
       "20                              acp-23-15783-2023.pdf          0.000000  \n",
       "21                             s41564-022-01266-x.pdf          0.058870  \n",
       "22                                           file.pdf          1.000000  \n",
       "23                     microorganisms-09-00357-v2.pdf          0.994664  \n",
       "24                             s41597-024-03069-7.pdf          0.087636  \n",
       "25                       2022.12.12.520098v2.full.pdf          0.094915  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Filename</th>\n",
       "      <th>Similarity Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10.1128/mSystems.00045-18</td>\n",
       "      <td>blair-et-al-2018-exploration-of-the-biosynthet...</td>\n",
       "      <td>0.948377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10.1101/2022.12.12.520098</td>\n",
       "      <td>2022.12.12.520098v2.full.pdf</td>\n",
       "      <td>0.161309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>10.1038/s41564-022-01266-x</td>\n",
       "      <td>s41564-022-01266-x.pdf</td>\n",
       "      <td>0.187543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>10.1111/1462-2920.16314</td>\n",
       "      <td>blair-et-al-2018-exploration-of-the-biosynthet...</td>\n",
       "      <td>0.083410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>10.1111/mec.16891</td>\n",
       "      <td>file.pdf</td>\n",
       "      <td>0.098154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>10.1021/acs.estlett.0c00748</td>\n",
       "      <td>alteio-et-al-2020-complementary-metagenomic-ap...</td>\n",
       "      <td>0.067630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>10.1128/msystems.00768-19</td>\n",
       "      <td>alteio-et-al-2020-complementary-metagenomic-ap...</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>10.1371/journal.pone.0228165</td>\n",
       "      <td>file.pdf</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>10.1016/j.geoderma.2021.115674</td>\n",
       "      <td>1-s2.0-S0016706121007540-main.pdf</td>\n",
       "      <td>0.511187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>10.1029/2022JG006889</td>\n",
       "      <td>1-s2.0-S0016706121007540-main.pdf</td>\n",
       "      <td>0.232726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>10.1002/ppp.2200</td>\n",
       "      <td>1-s2.0-S0016706121007540-main.pdf</td>\n",
       "      <td>0.187375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>10.1038/s41467-023-36515-y</td>\n",
       "      <td>s41467-023-36515-y.pdf</td>\n",
       "      <td>0.649401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>10.1002/lno.11306</td>\n",
       "      <td>s41597-019-0132-4.pdf</td>\n",
       "      <td>0.074472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>10.1038/s41597-024-03069-7</td>\n",
       "      <td>s41597-024-03069-7.pdf</td>\n",
       "      <td>0.435847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>10.1038/s41564-020-00861-0</td>\n",
       "      <td>s41564-020-00861-0.pdf</td>\n",
       "      <td>0.603828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>10.1128/MRA.01361-19</td>\n",
       "      <td>bell-et-al-2020-metatranscriptomic-sequencing-...</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>10.1128/mra.01080-23</td>\n",
       "      <td>nguyen-et-al-2024-metatranscriptomes-of-two-bi...</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>10.1038/s41597-019-0132-4</td>\n",
       "      <td>bell-et-al-2020-metatranscriptomic-sequencing-...</td>\n",
       "      <td>0.453781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>10.1038/s41597-024-04013-5</td>\n",
       "      <td>s41597-024-04013-5.pdf</td>\n",
       "      <td>0.515896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>10.1186/s40168-020-00889-8</td>\n",
       "      <td>s40168-020-00889-8.pdf</td>\n",
       "      <td>0.548277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>10.1093/nar/gkab990</td>\n",
       "      <td>acp-23-15783-2023.pdf</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>10.1126/sciadv.adg7888</td>\n",
       "      <td>s41564-022-01266-x.pdf</td>\n",
       "      <td>0.058870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>10.1371/journal.pone.0228165</td>\n",
       "      <td>file.pdf</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>10.3390/microorganisms9020357</td>\n",
       "      <td>microorganisms-09-00357-v2.pdf</td>\n",
       "      <td>0.994664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>10.3897/tdwgproceedings.1.20637</td>\n",
       "      <td>s41597-024-03069-7.pdf</td>\n",
       "      <td>0.087636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>10.5194/acp-23-15783-2023</td>\n",
       "      <td>2022.12.12.520098v2.full.pdf</td>\n",
       "      <td>0.094915</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 23
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-12T00:46:44.288615Z",
     "start_time": "2025-02-12T00:46:44.282019Z"
    }
   },
   "cell_type": "code",
   "source": "df.to_csv(\"citations_to_metadata_mapping.tsv\", sep=\"\\t\", index=False)",
   "id": "88e222cee8656ec6",
   "outputs": [],
   "execution_count": 24
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
